// @xpla/xpla.js/dist/core/group/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgLeaveGroup = exports.MsgGroupExec = exports.MsgGroupVote = exports.MsgGroupWithdrawProposal = exports.MsgGroupSubmitProposal = exports.MsgUpdateGroupPolicyMetadata = exports.MsgUpdateGroupPolicyDecisionPolicy = exports.MsgCreateGroupWithPolicy = exports.MsgUpdateGroupPolicyAdmin = exports.MsgCreateGroupPolicy = exports.MsgUpdateGroupMetadata = exports.MsgUpdateGroupAdmin = exports.MsgUpdateGroupMembers = exports.MsgCreateGroup = void 0;
__exportStar(require("./v1/msgs"), exports);
var msgs_1 = require("./v1/msgs");
Object.defineProperty(exports, "MsgCreateGroup", { enumerable: true, get: function () { return msgs_1.MsgCreateGroupV1; } });
Object.defineProperty(exports, "MsgUpdateGroupMembers", { enumerable: true, get: function () { return msgs_1.MsgUpdateGroupMembersV1; } });
Object.defineProperty(exports, "MsgUpdateGroupAdmin", { enumerable: true, get: function () { return msgs_1.MsgUpdateGroupAdminV1; } });
Object.defineProperty(exports, "MsgUpdateGroupMetadata", { enumerable: true, get: function () { return msgs_1.MsgUpdateGroupMetadataV1; } });
Object.defineProperty(exports, "MsgCreateGroupPolicy", { enumerable: true, get: function () { return msgs_1.MsgCreateGroupPolicyV1; } });
Object.defineProperty(exports, "MsgUpdateGroupPolicyAdmin", { enumerable: true, get: function () { return msgs_1.MsgUpdateGroupPolicyAdminV1; } });
Object.defineProperty(exports, "MsgCreateGroupWithPolicy", { enumerable: true, get: function () { return msgs_1.MsgCreateGroupWithPolicyV1; } });
Object.defineProperty(exports, "MsgUpdateGroupPolicyDecisionPolicy", { enumerable: true, get: function () { return msgs_1.MsgUpdateGroupPolicyDecisionPolicyV1; } });
Object.defineProperty(exports, "MsgUpdateGroupPolicyMetadata", { enumerable: true, get: function () { return msgs_1.MsgUpdateGroupPolicyMetadataV1; } });
Object.defineProperty(exports, "MsgGroupSubmitProposal", { enumerable: true, get: function () { return msgs_1.MsgGroupSubmitProposalV1; } });
Object.defineProperty(exports, "MsgGroupWithdrawProposal", { enumerable: true, get: function () { return msgs_1.MsgGroupWithdrawProposalV1; } });
Object.defineProperty(exports, "MsgGroupVote", { enumerable: true, get: function () { return msgs_1.MsgGroupVoteV1; } });
Object.defineProperty(exports, "MsgGroupExec", { enumerable: true, get: function () { return msgs_1.MsgGroupExecV1; } });
Object.defineProperty(exports, "MsgLeaveGroup", { enumerable: true, get: function () { return msgs_1.MsgLeaveGroupV1; } });
