// @xpla/xpla.js/dist/core/auth/v1beta1/Params.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthParamsV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var numeric_1 = require("../../numeric");
var auth_1 = require("@xpla/xpla.proto/cosmos/auth/v1beta1/auth");
var AuthParamsV1B1 = /** @class */ (function (_super) {
    __extends(AuthParamsV1B1, _super);
    /** Params defines the parameters for the auth module. */
    function AuthParamsV1B1(max_memo_characters, tx_sig_limit, tx_size_cost_per_byte, sig_verify_cost_ed25519, sig_verify_cost_secp256k1) {
        var _this = _super.call(this) || this;
        _this.max_memo_characters = new numeric_1.Int(max_memo_characters);
        _this.tx_sig_limit = new numeric_1.Int(tx_sig_limit);
        _this.tx_size_cost_per_byte = new numeric_1.Int(tx_size_cost_per_byte);
        _this.sig_verify_cost_ed25519 = new numeric_1.Int(sig_verify_cost_ed25519);
        _this.sig_verify_cost_secp256k1 = new numeric_1.Int(sig_verify_cost_secp256k1);
        return _this;
    }
    AuthParamsV1B1.fromAmino = function (data, _) {
        var max_memo_characters = data.max_memo_characters, tx_sig_limit = data.tx_sig_limit, tx_size_cost_per_byte = data.tx_size_cost_per_byte, sig_verify_cost_ed25519 = data.sig_verify_cost_ed25519, sig_verify_cost_secp256k1 = data.sig_verify_cost_secp256k1;
        return new AuthParamsV1B1(max_memo_characters !== null && max_memo_characters !== void 0 ? max_memo_characters : 0, tx_sig_limit !== null && tx_sig_limit !== void 0 ? tx_sig_limit : 0, tx_size_cost_per_byte !== null && tx_size_cost_per_byte !== void 0 ? tx_size_cost_per_byte : 0, sig_verify_cost_ed25519 !== null && sig_verify_cost_ed25519 !== void 0 ? sig_verify_cost_ed25519 : 0, sig_verify_cost_secp256k1 !== null && sig_verify_cost_secp256k1 !== void 0 ? sig_verify_cost_secp256k1 : 0);
    };
    AuthParamsV1B1.prototype.toAmino = function (_) {
        var _a = this, max_memo_characters = _a.max_memo_characters, tx_sig_limit = _a.tx_sig_limit, tx_size_cost_per_byte = _a.tx_size_cost_per_byte, sig_verify_cost_ed25519 = _a.sig_verify_cost_ed25519, sig_verify_cost_secp256k1 = _a.sig_verify_cost_secp256k1;
        var res = {
            max_memo_characters: max_memo_characters.toFixed(),
            tx_sig_limit: tx_sig_limit.toFixed(),
            tx_size_cost_per_byte: tx_size_cost_per_byte.toFixed(),
            sig_verify_cost_ed25519: sig_verify_cost_ed25519.toFixed(),
            sig_verify_cost_secp256k1: sig_verify_cost_secp256k1.toFixed(),
        };
        return res;
    };
    AuthParamsV1B1.fromData = function (data, _) {
        var max_memo_characters = data.max_memo_characters, tx_sig_limit = data.tx_sig_limit, tx_size_cost_per_byte = data.tx_size_cost_per_byte, sig_verify_cost_ed25519 = data.sig_verify_cost_ed25519, sig_verify_cost_secp256k1 = data.sig_verify_cost_secp256k1;
        return new AuthParamsV1B1(max_memo_characters, tx_sig_limit, tx_size_cost_per_byte, sig_verify_cost_ed25519, sig_verify_cost_secp256k1);
    };
    AuthParamsV1B1.prototype.toData = function (_) {
        var _a = this, max_memo_characters = _a.max_memo_characters, tx_sig_limit = _a.tx_sig_limit, tx_size_cost_per_byte = _a.tx_size_cost_per_byte, sig_verify_cost_ed25519 = _a.sig_verify_cost_ed25519, sig_verify_cost_secp256k1 = _a.sig_verify_cost_secp256k1;
        var res = {
            '@type': '/cosmos.auth.v1beta1.Params',
            max_memo_characters: max_memo_characters.toFixed(),
            tx_sig_limit: tx_sig_limit.toFixed(),
            tx_size_cost_per_byte: tx_size_cost_per_byte.toFixed(),
            sig_verify_cost_ed25519: sig_verify_cost_ed25519.toFixed(),
            sig_verify_cost_secp256k1: sig_verify_cost_secp256k1.toFixed(),
        };
        return res;
    };
    AuthParamsV1B1.fromProto = function (proto, _) {
        return new AuthParamsV1B1(proto.maxMemoCharacters.toString(), proto.txSigLimit.toString(), proto.txSizeCostPerByte.toString(), proto.sigVerifyCostEd25519.toString(), proto.sigVerifyCostSecp256k1.toString());
    };
    AuthParamsV1B1.prototype.toProto = function (_) {
        var _a = this, max_memo_characters = _a.max_memo_characters, tx_sig_limit = _a.tx_sig_limit, tx_size_cost_per_byte = _a.tx_size_cost_per_byte, sig_verify_cost_ed25519 = _a.sig_verify_cost_ed25519, sig_verify_cost_secp256k1 = _a.sig_verify_cost_secp256k1;
        return auth_1.Params.fromPartial({
            maxMemoCharacters: max_memo_characters.toFixed(),
            txSigLimit: tx_sig_limit.toFixed(),
            txSizeCostPerByte: tx_size_cost_per_byte.toFixed(),
            sigVerifyCostEd25519: sig_verify_cost_ed25519.toFixed(),
            sigVerifyCostSecp256k1: sig_verify_cost_secp256k1.toFixed(),
        });
    };
    return AuthParamsV1B1;
}(json_1.JSONSerializable));
exports.AuthParamsV1B1 = AuthParamsV1B1;
