// @xpla/xpla.proto/ethermint/erc20/v1/genesis.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Params = exports.GenesisState = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
const erc20_1 = require("./erc20");
exports.protobufPackage = "ethermint.erc20.v1";
function createBaseGenesisState() {
    return { params: undefined, tokenPairs: [] };
}
exports.GenesisState = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.params !== undefined) {
            exports.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        for (const v of message.tokenPairs) {
            erc20_1.TokenPair.encode(v, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGenesisState();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.params = exports.Params.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.tokenPairs.push(erc20_1.TokenPair.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            params: isSet(object.params) ? exports.Params.fromJSON(object.params) : undefined,
            tokenPairs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.tokenPairs)
                ? object.tokenPairs.map((e) => erc20_1.TokenPair.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.params !== undefined) {
            obj.params = exports.Params.toJSON(message.params);
        }
        if ((_a = message.tokenPairs) === null || _a === void 0 ? void 0 : _a.length) {
            obj.tokenPairs = message.tokenPairs.map((e) => erc20_1.TokenPair.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.GenesisState.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGenesisState();
        message.params = (object.params !== undefined && object.params !== null)
            ? exports.Params.fromPartial(object.params)
            : undefined;
        message.tokenPairs = ((_a = object.tokenPairs) === null || _a === void 0 ? void 0 : _a.map((e) => erc20_1.TokenPair.fromPartial(e))) || [];
        return message;
    },
};
function createBaseParams() {
    return { enableErc20: false, enableEvmHook: false };
}
exports.Params = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.enableErc20 === true) {
            writer.uint32(8).bool(message.enableErc20);
        }
        if (message.enableEvmHook === true) {
            writer.uint32(16).bool(message.enableEvmHook);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseParams();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.enableErc20 = reader.bool();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.enableEvmHook = reader.bool();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            enableErc20: isSet(object.enableErc20) ? globalThis.Boolean(object.enableErc20) : false,
            enableEvmHook: isSet(object.enableEvmHook) ? globalThis.Boolean(object.enableEvmHook) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.enableErc20 === true) {
            obj.enableErc20 = message.enableErc20;
        }
        if (message.enableEvmHook === true) {
            obj.enableEvmHook = message.enableEvmHook;
        }
        return obj;
    },
    create(base) {
        return exports.Params.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseParams();
        message.enableErc20 = (_a = object.enableErc20) !== null && _a !== void 0 ? _a : false;
        message.enableEvmHook = (_b = object.enableEvmHook) !== null && _b !== void 0 ? _b : false;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
