// msg-parser/dist/MessageStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PropertiesStream_1 = require("./property/PropertiesStream");
var PropertyTag_1 = require("./property/PropertyTag");
var MessageStorage = /** @class */ (function () {
    function MessageStorage(storage, namedPropertyMappingStorage) {
        this.storage = storage;
        this.namedPropertyMappingStorage = namedPropertyMappingStorage;
    }
    MessageStorage.prototype.propertiesStream = function () {
        return this.createPropertiesStream(this.storage.streams().find(function (streamDirectoryEntry) { return PropertiesStream_1.PropertiesStream.STREAM_NAME === streamDirectoryEntry.getDirectoryEntryName(); }));
    };
    MessageStorage.prototype.streams = function () {
        return this.storage.streams();
    };
    MessageStorage.prototype.findByName = function (name) {
        return this.streams().find(function (streamDirectoryEntry) { return streamDirectoryEntry.getDirectoryEntryName().toUpperCase() === name.toUpperCase(); });
    };
    MessageStorage.prototype.underlyingDirectoryEntry = function () {
        return this.storage;
    };
    MessageStorage.prototype.getProperty = function (property) {
        var propertyType = property.propertyType;
        if (propertyType === undefined) {
            throw new Error("Property type is required");
        }
        if (property instanceof PropertyTag_1.PropertyTag) {
            return propertyType.resolveValue(this, property);
        }
        else {
            var propertyId = this.namedPropertyMappingStorage.getPropertyIdByPropertyName(property);
            return property.propertyType.resolveValue(this, new PropertyTag_1.PropertyTag(propertyId, property.propertyType));
        }
    };
    MessageStorage.VALUE_STREAM_PREFIX = "__substg1.0_";
    MessageStorage.INTERNAL_STORAGE_NAME = "__substg1.0_3701000D";
    MessageStorage.ATTACHMENT_STORAGE_PREFIX = "__attach_version1.0";
    return MessageStorage;
}());
exports.MessageStorage = MessageStorage;
