// msg-parser/dist/Msg.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var AbstractMessage_1 = require("./AbstractMessage");
var compound_binary_file_js_1 = require("compound-binary-file-js");
var PropertiesStream_1 = require("./property/PropertiesStream");
var NamedPropertyMappingStorage_1 = require("./nameid_mapping/NamedPropertyMappingStorage");
var EmbeddedMessage_1 = require("./EmbeddedMessage");
var RootStoragePropertyStream_1 = require("./property/RootStoragePropertyStream");
var MessageStorage_1 = require("./MessageStorage");
var KnownProperties_1 = require("./property/KnownProperties");
var Msg = /** @class */ (function (_super) {
    __extends(Msg, _super);
    function Msg(compoundFile) {
        var _this = _super.call(this, compoundFile.getRootStorage(), Msg.namedPropertyMappingStorage(compoundFile.getRootStorage())) || this;
        _this.compoundFile = compoundFile;
        return _this;
    }
    Msg.fromBytes = function (bytes) {
        return new Msg(compound_binary_file_js_1.CompoundFile.fromBytes(bytes));
    };
    Msg.fromUint8Array = function (bytes) {
        return new Msg(compound_binary_file_js_1.CompoundFile.fromUint8Array(new Uint8Array(bytes)));
    };
    Msg.prototype.createPropertiesStream = function (stream) {
        return new RootStoragePropertyStream_1.RootStoragePropertyStream(stream);
    };
    Msg.prototype.getNamedPropertyMappingStorage = function () {
        return Msg.namedPropertyMappingStorage(this.underlyingDirectoryEntry());
    };
    Msg.namedPropertyMappingStorage = function (storageDirectoryEntry) {
        return new NamedPropertyMappingStorage_1.NamedPropertyMappingStorage(storageDirectoryEntry.findChild(function (directoryEntry) { return directoryEntry.getDirectoryEntryName().toUpperCase() === NamedPropertyMappingStorage_1.NamedPropertyMappingStorage.STORAGE_NAME.toUpperCase(); }));
    };
    Msg.prototype.asBytes = function () {
        return this.compoundFile.asBytes();
    };
    Msg.prototype.extractEmbeddedMessage = function (embeddedMessage) {
        var nameidmapping = this.underlyingDirectoryEntry().findChild((function (directoryEntry) { return NamedPropertyMappingStorage_1.NamedPropertyMappingStorage.STORAGE_NAME.toUpperCase() === directoryEntry.getDirectoryEntryName().toUpperCase(); }));
        var copy = new compound_binary_file_js_1.CompoundFile();
        var nameidmappingCopy = copy.getRootStorage().addStorage(NamedPropertyMappingStorage_1.NamedPropertyMappingStorage.STORAGE_NAME);
        nameidmapping.eachChild(this.copyConsumer(nameidmappingCopy));
        embeddedMessage.internalStorage().underlyingDirectoryEntry().eachChild(this.copyConsumer(copy.getRootStorage()));
        var propertiesStream = copy.getRootStorage().findChild(function (directoryEntry) { return directoryEntry.getDirectoryEntryName() === PropertiesStream_1.PropertiesStream.STREAM_NAME; });
        var streamData = propertiesStream.getStreamData();
        var adjustedData = [];
        adjustedData.push.apply(adjustedData, streamData.slice(0, 24));
        adjustedData.push.apply(adjustedData, compound_binary_file_js_1.initializedWidth(8, 0));
        adjustedData.push.apply(adjustedData, streamData.slice(24, streamData.length));
        propertiesStream.setStreamData(adjustedData);
        return new Msg(copy);
    };
    Msg.prototype.copyConsumer = function (parent) {
        var me = this;
        return function (directoryEntry) {
            var copy = null;
            if (directoryEntry instanceof compound_binary_file_js_1.StorageDirectoryEntry) {
                copy = parent.addStorage(directoryEntry.getDirectoryEntryName());
                directoryEntry.eachChild(me.copyConsumer(copy));
            }
            else {
                copy = parent.addStream(directoryEntry.getDirectoryEntryName(), directoryEntry.getStreamData());
            }
            copy.setCLSID(directoryEntry.getCLSID());
            copy.setStateBits(directoryEntry.getStateBits());
            copy.setCreationTime(directoryEntry.getCreationTime());
            copy.setModifiedTime(directoryEntry.getModifiedTime());
        };
    };
    Msg.prototype.embeddedMessages = function () {
        return Msg.embeddedMessages(this, this.namedPropertyMappingStorage);
    };
    Msg.embeddedMessages = function (messageStorage, namedPropertyMappingStorage) {
        return messageStorage.underlyingDirectoryEntry().storages().filter(function (storageDirectoryEntry) { return storageDirectoryEntry.getDirectoryEntryName().startsWith(MessageStorage_1.MessageStorage.ATTACHMENT_STORAGE_PREFIX); })
            .filter(function (storageDirectoryEntry) { return storageDirectoryEntry.storages().filter(function (substorage) { return substorage.getDirectoryEntryName() === EmbeddedMessage_1.EmbeddedMessage.INTERNAL_STORAGE_NAME; }).length === 1; })
            .map(function (directoryEntry) { return new EmbeddedMessage_1.EmbeddedMessage(directoryEntry, namedPropertyMappingStorage); })
            .filter(function (embeddedMsg) { return KnownProperties_1.PidTagAttachMethod.propertyType.resolveValue(embeddedMsg, KnownProperties_1.PidTagAttachMethod) !== EmbeddedMessage_1.EmbeddedMessage.ATTACH_METHOD_CUSTOM; });
    };
    Msg.ATTACHMENT_STORAGE_PREFIX = "__attach_version1.0";
    return Msg;
}(AbstractMessage_1.AbstractMessage));
exports.Msg = Msg;
