// @xpla/xpla.js/dist/core/consensus/v1/msgs/MsgUpdateParams.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateConsensusParamsV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/consensus/v1/tx");
var Params_1 = require("../Params");
var MsgUpdateConsensusParamsV1B1 = /** @class */ (function (_super) {
    __extends(MsgUpdateConsensusParamsV1B1, _super);
    /**
     * @param authority is the address that controls the module
     */
    function MsgUpdateConsensusParamsV1B1(authority, block, evidence, validator) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.block = block;
        _this.evidence = evidence;
        _this.validator = validator;
        return _this;
    }
    MsgUpdateConsensusParamsV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, block = _a.block, evidence = _a.evidence, validator = _a.validator;
        return new MsgUpdateConsensusParamsV1B1(authority, block ? Params_1.BlockParams.fromAmino(block) : undefined, evidence ? Params_1.EvidenceParams.fromAmino(evidence) : undefined, validator ? Params_1.ValidatorParams.fromAmino(validator) : undefined);
    };
    MsgUpdateConsensusParamsV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, authority = _a.authority, block = _a.block, evidence = _a.evidence, validator = _a.validator;
        return {
            type: isClassic
                ? 'consensus/MsgUpdateParams'
                : 'cosmos-sdk/x/consensus/MsgUpdateParams',
            value: {
                authority: authority,
                block: block ? block.toAmino() : undefined,
                evidence: evidence ? evidence.toAmino() : undefined,
                validator: validator ? validator.toAmino() : undefined,
            },
        };
    };
    MsgUpdateConsensusParamsV1B1.fromData = function (data, _isClassic) {
        var authority = data.authority, block = data.block, evidence = data.evidence, validator = data.validator;
        return new MsgUpdateConsensusParamsV1B1(authority, block ? Params_1.BlockParams.fromData(block) : undefined, evidence ? Params_1.EvidenceParams.fromData(evidence) : undefined, validator ? Params_1.ValidatorParams.fromData(validator) : undefined);
    };
    MsgUpdateConsensusParamsV1B1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, block = _a.block, evidence = _a.evidence, validator = _a.validator;
        return {
            '@type': '/cosmos.consensus.v1.MsgUpdateParams',
            authority: authority,
            block: block ? block.toData() : undefined,
            evidence: evidence ? evidence.toData() : undefined,
            validator: validator ? validator.toData() : undefined,
        };
    };
    MsgUpdateConsensusParamsV1B1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateConsensusParamsV1B1(proto.authority, proto.block ? Params_1.BlockParams.fromProto(proto.block) : undefined, proto.evidence ? Params_1.EvidenceParams.fromProto(proto.evidence) : undefined, proto.validator ? Params_1.ValidatorParams.fromProto(proto.validator) : undefined);
    };
    MsgUpdateConsensusParamsV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, block = _a.block, evidence = _a.evidence, validator = _a.validator;
        return tx_1.MsgUpdateParams.fromPartial({
            authority: authority,
            block: block ? block.toProto() : undefined,
            evidence: evidence ? evidence.toProto() : undefined,
            validator: validator ? validator.toProto() : undefined,
        });
    };
    MsgUpdateConsensusParamsV1B1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.bank.v1beta1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    };
    MsgUpdateConsensusParamsV1B1.unpackAny = function (msgAny, _isClassic) {
        return MsgUpdateConsensusParamsV1B1.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    };
    return MsgUpdateConsensusParamsV1B1;
}(json_1.JSONSerializable));
exports.MsgUpdateConsensusParamsV1B1 = MsgUpdateConsensusParamsV1B1;
