// @xpla/xpla.js/dist/core/gov/v1beta1/Proposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProposalV1B1 = void 0;
var Coins_1 = require("../../Coins");
var numeric_1 = require("../../numeric");
var json_1 = require("../../../util/json");
var proposals_1 = require("../../distribution/v1beta1/proposals");
var erc20_1 = require("../../erc20");
var proposals_2 = require("../../params/proposals");
var proposals_3 = require("../../ibc/proposals");
var proposals_4 = require("./proposals");
var upgrade_1 = require("../../upgrade");
var wasm_1 = require("../../wasm");
var xpla_1 = require("../../xpla");
var gov_1 = require("@xpla/xpla.proto/cosmos/gov/v1beta1/gov");
/**
 * Stores information pertaining to a submitted proposal, such as its status and time of
 * the voting period
 */
var ProposalV1B1 = /** @class */ (function (_super) {
    __extends(ProposalV1B1, _super);
    /**
     *
     * @param id proposal's ID
     * @param content content of the proposal
     * @param status proposal's status
     * @param final_tally_result tally result
     * @param submit_time time proposal was submitted and deposit period started
     * @param deposit_end_time time deposit period will end
     * @param total_deposit amount of coins deposited by all users
     * @param voting_start_time time voting period will start
     * @param voting_end_time time voting period will end
     */
    function ProposalV1B1(id, content, status, final_tally_result, submit_time, deposit_end_time, total_deposit, voting_start_time, voting_end_time) {
        var _this = _super.call(this) || this;
        _this.id = id;
        _this.content = content;
        _this.status = status;
        _this.final_tally_result = final_tally_result;
        _this.submit_time = submit_time;
        _this.deposit_end_time = deposit_end_time;
        _this.total_deposit = total_deposit;
        _this.voting_start_time = voting_start_time;
        _this.voting_end_time = voting_end_time;
        return _this;
    }
    ProposalV1B1.fromAmino = function (data, isClassic) {
        var id = data.id, content = data.content, status = data.status, final_tally_result = data.final_tally_result, submit_time = data.submit_time, deposit_end_time = data.deposit_end_time, total_deposit = data.total_deposit, voting_start_time = data.voting_start_time, voting_end_time = data.voting_end_time;
        return new ProposalV1B1(Number.parseInt(id), ProposalV1B1.Content.fromAmino(content, isClassic), status, {
            yes: new numeric_1.Int(final_tally_result.yes || 0),
            no: new numeric_1.Int(final_tally_result.no || 0),
            abstain: new numeric_1.Int(final_tally_result.abstain || 0),
            no_with_veto: new numeric_1.Int(final_tally_result.no_with_veto || 0),
        }, new Date(submit_time), new Date(deposit_end_time), Coins_1.Coins.fromAmino(total_deposit), new Date(voting_start_time), new Date(voting_end_time));
    };
    ProposalV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, status = _a.status, final_tally_result = _a.final_tally_result;
        return {
            id: this.id.toFixed(),
            content: this.content.toAmino(isClassic),
            status: status,
            final_tally_result: {
                yes: final_tally_result.yes.toFixed(),
                no: final_tally_result.no.toFixed(),
                abstain: final_tally_result.abstain.toFixed(),
                no_with_veto: final_tally_result.no_with_veto.toFixed(),
            },
            submit_time: this.submit_time.toISOString(),
            deposit_end_time: this.deposit_end_time.toISOString(),
            total_deposit: this.total_deposit.toAmino(),
            voting_start_time: this.voting_start_time.toISOString(),
            voting_end_time: this.voting_end_time.toISOString(),
        };
    };
    ProposalV1B1.fromData = function (data, isClassic) {
        var proposal_id = data.proposal_id, content = data.content, status = data.status, final_tally_result = data.final_tally_result, submit_time = data.submit_time, deposit_end_time = data.deposit_end_time, total_deposit = data.total_deposit, voting_start_time = data.voting_start_time, voting_end_time = data.voting_end_time;
        return new ProposalV1B1(Number.parseInt(proposal_id), ProposalV1B1.Content.fromData(content, isClassic), (0, gov_1.proposalStatusFromJSON)(status), {
            yes: new numeric_1.Int((final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.yes) || 0),
            no: new numeric_1.Int((final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.no) || 0),
            abstain: new numeric_1.Int((final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.abstain) || 0),
            no_with_veto: new numeric_1.Int((final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.no_with_veto) || 0),
        }, new Date(submit_time), new Date(deposit_end_time), Coins_1.Coins.fromData(total_deposit), new Date(voting_start_time), new Date(voting_end_time));
    };
    ProposalV1B1.prototype.toData = function (isClassic) {
        var _a = this, status = _a.status, final_tally_result = _a.final_tally_result;
        return {
            proposal_id: this.id.toFixed(),
            content: this.content.toData(isClassic),
            status: (0, gov_1.proposalStatusToJSON)(status),
            final_tally_result: {
                yes: final_tally_result.yes.toString(),
                no: final_tally_result.no.toString(),
                abstain: final_tally_result.abstain.toString(),
                no_with_veto: final_tally_result.no_with_veto.toString(),
            },
            submit_time: this.submit_time.toISOString(),
            deposit_end_time: this.deposit_end_time.toISOString(),
            total_deposit: this.total_deposit.toData(),
            voting_start_time: this.voting_start_time.toISOString(),
            voting_end_time: this.voting_end_time.toISOString(),
        };
    };
    ProposalV1B1.fromProto = function (data, isClassic) {
        var _a, _b, _c, _d;
        var id = data.proposalId;
        var content = data.content;
        var status = data.status;
        var final_tally_result = data.finalTallyResult;
        var submit_time = data.submitTime;
        var deposit_end_time = data.depositEndTime;
        var total_deposit = data.totalDeposit;
        var voting_start_time = data.votingStartTime;
        var voting_end_time = data.votingEndTime;
        return new ProposalV1B1(id.toNumber(), ProposalV1B1.Content.fromProto(content, isClassic), status, {
            yes: new numeric_1.Int((_a = final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.yes) !== null && _a !== void 0 ? _a : 0),
            no: new numeric_1.Int((_b = final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.no) !== null && _b !== void 0 ? _b : 0),
            abstain: new numeric_1.Int((_c = final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.abstain) !== null && _c !== void 0 ? _c : 0),
            no_with_veto: new numeric_1.Int((_d = final_tally_result === null || final_tally_result === void 0 ? void 0 : final_tally_result.noWithVeto) !== null && _d !== void 0 ? _d : 0),
        }, submit_time, deposit_end_time, Coins_1.Coins.fromProto(total_deposit), voting_start_time, voting_end_time);
    };
    ProposalV1B1.prototype.toProto = function (isClassic) {
        var _a = this, status = _a.status, final_tally_result = _a.final_tally_result;
        var ftr;
        if (final_tally_result) {
            ftr = gov_1.TallyResult.fromPartial({
                yes: final_tally_result.yes.toString(),
                no: final_tally_result.no.toString(),
                abstain: final_tally_result.abstain.toString(),
                noWithVeto: final_tally_result.no_with_veto.toString(),
            });
        }
        return gov_1.Proposal.fromPartial({
            proposalId: this.id,
            content: this.content.packAny(isClassic),
            status: status,
            finalTallyResult: ftr,
            submitTime: this.submit_time,
            depositEndTime: this.deposit_end_time,
            totalDeposit: this.total_deposit.toProto(),
            votingEndTime: this.voting_end_time,
            votingStartTime: this.voting_start_time,
        });
    };
    return ProposalV1B1;
}(json_1.JSONSerializable));
exports.ProposalV1B1 = ProposalV1B1;
(function (ProposalV1B1) {
    ProposalV1B1.Status = gov_1.ProposalStatus;
    var Content;
    (function (Content) {
        function fromAmino(amino, isClassic) {
            switch (amino.type) {
                case 'gov/TextProposal':
                case 'cosmos-sdk/TextProposal':
                    return proposals_4.TextProposal.fromAmino(amino, isClassic);
                case 'distribution/CommunityPoolSpendProposal':
                case 'cosmos-sdk/CommunityPoolSpendProposal':
                    return proposals_1.CommunityPoolSpendProposal.fromAmino(amino, isClassic);
                case 'params/ParameterChangeProposal':
                case 'cosmos-sdk/ParameterChangeProposal':
                    return proposals_2.ParameterChangeProposal.fromAmino(amino, isClassic);
                case 'upgrade/SoftwareUpgradeProposal':
                case 'cosmos-sdk/SoftwareUpgradeProposal':
                    return upgrade_1.SoftwareUpgradeProposal.fromAmino(amino, isClassic);
                case 'upgrade/CancelSoftwareUpgradeProposal':
                case 'cosmos-sdk/CancelSoftwareUpgradeProposal':
                    return upgrade_1.CancelSoftwareUpgradeProposal.fromAmino(amino, isClassic);
                case 'ibc/ClientUpdateProposal':
                    return proposals_3.ClientUpdateProposal.fromAmino(amino, isClassic);
                case 'wasm/ClearAdminProposal':
                    return wasm_1.ClearAdminProposal.fromAmino(amino, isClassic);
                case 'wasm/ExecuteContractProposal':
                    return wasm_1.ExecuteContractProposal.fromAmino(amino, isClassic);
                case 'wasm/InstantiateContractProposal':
                    return wasm_1.InstantiateContractProposal.fromAmino(amino, isClassic);
                case 'wasm/MigrateContractProposal':
                    return wasm_1.MigrateContractProposal.fromAmino(amino, isClassic);
                case 'wasm/PinCodesProposal':
                    return wasm_1.PinCodesProposal.fromAmino(amino, isClassic);
                case 'wasm/StoreCodeProposal':
                    return wasm_1.StoreCodeProposal.fromAmino(amino, isClassic);
                case 'wasm/SudoContractProposal':
                    return wasm_1.SudoContractProposal.fromAmino(amino, isClassic);
                case 'wasm/UnpinCodesProposal':
                    return wasm_1.UnpinCodesProposal.fromAmino(amino, isClassic);
                case 'wasm/UpdateAdminProposal':
                    return wasm_1.UpdateAdminProposal.fromAmino(amino, isClassic);
                case 'wasm/UpdateInstantiateConfigProposal':
                    return wasm_1.UpdateInstantiateConfigProposal.fromAmino(amino, isClassic);
                case 'erc20/RegisterCoinProposal':
                    return erc20_1.RegisterCoinProposal.fromAmino(amino, isClassic);
                case 'erc20/RegisterERC20Proposal':
                    return erc20_1.RegisterERC20Proposal.fromAmino(amino, isClassic);
                case 'erc20/ToggleTokenConversionProposal':
                    return erc20_1.ToggleTokenConversionProposal.fromAmino(amino, isClassic);
                case 'xpla/RegisterVolunteerValidatorProposal':
                    return xpla_1.RegisterVolunteerValidatorProposal.fromAmino(amino, isClassic);
                case 'xpla/RegisterVolunteerValidatorProposalWithDeposit':
                    return xpla_1.RegisterVolunteerValidatorProposalWithDeposit.fromAmino(amino, isClassic);
                case 'xpla/UnregisterVolunteerValidatorProposal':
                    return xpla_1.UnregisterVolunteerValidatorProposal.fromAmino(amino, isClassic);
                case 'xpla/UnregisterVolunteerValidatorProposalWithDeposit':
                    return xpla_1.UnregisterVolunteerValidatorProposalWithDeposit.fromAmino(amino, isClassic);
            }
        }
        Content.fromAmino = fromAmino;
        function fromData(data, isClassic) {
            switch (data['@type']) {
                case '/cosmos.gov.v1beta1.TextProposal':
                    return proposals_4.TextProposal.fromData(data, isClassic);
                case '/cosmos.distribution.v1beta1.CommunityPoolSpendProposal':
                    return proposals_1.CommunityPoolSpendProposal.fromData(data, isClassic);
                case '/cosmos.params.v1beta1.ParameterChangeProposal':
                    return proposals_2.ParameterChangeProposal.fromData(data, isClassic);
                case '/cosmos.upgrade.v1beta1.SoftwareUpgradeProposal':
                    return upgrade_1.SoftwareUpgradeProposal.fromData(data, isClassic);
                case '/cosmos.upgrade.v1beta1.CancelSoftwareUpgradeProposal':
                    return upgrade_1.CancelSoftwareUpgradeProposal.fromData(data, isClassic);
                case '/ibc.core.client.v1.ClientUpdateProposal':
                    return proposals_3.ClientUpdateProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.ClearAdminProposal':
                    return wasm_1.ClearAdminProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.ExecuteContractProposal':
                    return wasm_1.ExecuteContractProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.InstantiateContractProposal':
                    return wasm_1.InstantiateContractProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.MigrateContractProposal':
                    return wasm_1.MigrateContractProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.PinCodesProposal':
                    return wasm_1.PinCodesProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.StoreCodeProposal':
                    return wasm_1.StoreCodeProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.SudoContractProposal':
                    return wasm_1.SudoContractProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.UnpinCodesProposal':
                    return wasm_1.UnpinCodesProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.UpdateAdminProposal':
                    return wasm_1.UpdateAdminProposal.fromData(data, isClassic);
                case '/cosmwasm.wasm.v1.UpdateInstantiateConfigProposal':
                    return wasm_1.UpdateInstantiateConfigProposal.fromData(data, isClassic);
                case '/ethermint.erc20.v1.RegisterCoinProposal':
                case '/evmos.erc20.v1.RegisterCoinProposal':
                    return erc20_1.RegisterCoinProposal.fromData(data, isClassic);
                case '/ethermint.erc20.v1.RegisterERC20Proposal':
                case '/evmos.erc20.v1.RegisterERC20Proposal':
                    return erc20_1.RegisterERC20Proposal.fromData(data, isClassic);
                case '/ethermint.erc20.v1.ToggleTokenConversionProposal':
                case '/evmos.erc20.v1.ToggleTokenConversionProposal':
                    return erc20_1.ToggleTokenConversionProposal.fromData(data, isClassic);
                case '/xpla.volunteer.v1beta1.RegisterVolunteerValidatorProposal':
                    return xpla_1.RegisterVolunteerValidatorProposal.fromData(data, isClassic);
                case '/xpla.volunteer.v1beta1.RegisterVolunteerValidatorProposalWithDeposit':
                    return xpla_1.RegisterVolunteerValidatorProposalWithDeposit.fromData(data, isClassic);
                case '/xpla.volunteer.v1beta1.UnregisterVolunteerValidatorProposal':
                    return xpla_1.UnregisterVolunteerValidatorProposal.fromData(data, isClassic);
                case '/xpla.volunteer.v1beta1.UnregisterVolunteerValidatorProposalWithDeposit':
                    return xpla_1.UnregisterVolunteerValidatorProposalWithDeposit.fromData(data, isClassic);
            }
        }
        Content.fromData = fromData;
        function fromProto(anyProto, isClassic) {
            var typeUrl = anyProto.typeUrl;
            switch (typeUrl) {
                case '/cosmos.gov.v1beta1.TextProposal':
                    return proposals_4.TextProposal.unpackAny(anyProto, isClassic);
                case '/cosmos.distribution.v1beta1.CommunityPoolSpendProposal':
                    return proposals_1.CommunityPoolSpendProposal.unpackAny(anyProto, isClassic);
                case '/cosmos.params.v1beta1.ParameterChangeProposal':
                    return proposals_2.ParameterChangeProposal.unpackAny(anyProto, isClassic);
                case '/cosmos.upgrade.v1beta1.SoftwareUpgradeProposal':
                    return upgrade_1.SoftwareUpgradeProposal.unpackAny(anyProto, isClassic);
                case '/cosmos.upgrade.v1beta1.CancelSoftwareUpgradeProposal':
                    return upgrade_1.CancelSoftwareUpgradeProposal.unpackAny(anyProto, isClassic);
                case '/ibc.core.client.v1.ClientUpdateProposal':
                    return proposals_3.ClientUpdateProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.ClearAdminProposal':
                    return wasm_1.ClearAdminProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.ExecuteContractProposal':
                    return wasm_1.ExecuteContractProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.InstantiateContractProposal':
                    return wasm_1.InstantiateContractProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.MigrateContractProposal':
                    return wasm_1.MigrateContractProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.PinCodesProposal':
                    return wasm_1.PinCodesProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.StoreCodeProposal':
                    return wasm_1.StoreCodeProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.SudoContractProposal':
                    return wasm_1.SudoContractProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.UnpinCodesProposal':
                    return wasm_1.UnpinCodesProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.UpdateAdminProposal':
                    return wasm_1.UpdateAdminProposal.unpackAny(anyProto, isClassic);
                case '/cosmwasm.wasm.v1.UpdateInstantiateConfigProposal':
                    return wasm_1.UpdateInstantiateConfigProposal.unpackAny(anyProto, isClassic);
                case '/ethermint.erc20.v1.RegisterCoinProposal':
                case '/evmos.erc20.v1.RegisterCoinProposal':
                    return erc20_1.RegisterCoinProposal.unpackAny(anyProto, isClassic);
                case '/ethermint.erc20.v1.RegisterERC20Proposal':
                case '/evmos.erc20.v1.RegisterERC20Proposal':
                    return erc20_1.RegisterERC20Proposal.unpackAny(anyProto, isClassic);
                case '/ethermint.erc20.v1.ToggleTokenConversionProposal':
                case '/evmos.erc20.v1.ToggleTokenConversionProposal':
                    return erc20_1.ToggleTokenConversionProposal.unpackAny(anyProto, isClassic);
                case '/xpla.volunteer.v1beta1.RegisterVolunteerValidatorProposal':
                    return xpla_1.RegisterVolunteerValidatorProposal.unpackAny(anyProto, isClassic);
                case '/xpla.volunteer.v1beta1.RegisterVolunteerValidatorProposalWithDeposit':
                    return xpla_1.RegisterVolunteerValidatorProposalWithDeposit.unpackAny(anyProto, isClassic);
                case '/xpla.volunteer.v1beta1.UnregisterVolunteerValidatorProposal':
                    return xpla_1.UnregisterVolunteerValidatorProposal.unpackAny(anyProto, isClassic);
                case '/xpla.volunteer.v1beta1.UnregisterVolunteerValidatorProposalWithDeposit':
                    return xpla_1.UnregisterVolunteerValidatorProposalWithDeposit.unpackAny(anyProto, isClassic);
            }
            throw "Proposal content ".concat(typeUrl, " not recognized");
        }
        Content.fromProto = fromProto;
    })(Content = ProposalV1B1.Content || (ProposalV1B1.Content = {}));
})(ProposalV1B1 = exports.ProposalV1B1 || (exports.ProposalV1B1 = {}));
exports.ProposalV1B1 = ProposalV1B1;
