// @xpla/xpla.js/dist/core/gov/v1beta1/Vote.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeightedVoteOptionV1B1 = exports.VoteV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var numeric_1 = require("../../numeric");
var gov_1 = require("@xpla/xpla.proto/cosmos/gov/v1beta1/gov");
/**
 * Stores vote information for a proposal
 */
var VoteV1B1 = /** @class */ (function (_super) {
    __extends(VoteV1B1, _super);
    /**
     * @param proposal_id ID of proposal to vote on
     * @param voter voter's account address
     * @param options voting options
     */
    function VoteV1B1(proposal_id, voter, options) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.voter = voter;
        _this.options = options;
        _this.Option = gov_1.VoteOption;
        return _this;
    }
    VoteV1B1.fromAmino = function (data, _) {
        var proposal_id = data.proposal_id, voter = data.voter, options = data.options;
        return new VoteV1B1(Number.parseInt(proposal_id), voter, options.map(function (v) { return WeightedVoteOptionV1B1.fromAmino(v); }));
    };
    VoteV1B1.prototype.toAmino = function (_) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options;
        var res = {
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            options: options.map(function (v) { return v.toAmino(); }),
        };
        return res;
    };
    VoteV1B1.fromData = function (data, _) {
        var proposal_id = data.proposal_id, voter = data.voter, options = data.options;
        return new VoteV1B1(Number.parseInt(proposal_id), voter, options.map(function (v) { return WeightedVoteOptionV1B1.fromData(v); }));
    };
    VoteV1B1.prototype.toData = function (_) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options;
        var res = {
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            options: options.map(function (v) { return v.toData(); }),
        };
        return res;
    };
    VoteV1B1.fromProto = function (proto, _) {
        return new VoteV1B1(proto.proposalId.toNumber(), proto.voter, proto.options.map(function (o) { return WeightedVoteOptionV1B1.fromProto(o); }));
    };
    VoteV1B1.prototype.toProto = function (_) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options;
        return gov_1.Vote.fromPartial({
            options: options.map(function (o) { return o.toProto(); }),
            proposalId: proposal_id,
            voter: voter,
        });
    };
    return VoteV1B1;
}(json_1.JSONSerializable));
exports.VoteV1B1 = VoteV1B1;
(function (VoteV1B1) {
    VoteV1B1.Option = gov_1.VoteOption;
})(VoteV1B1 = exports.VoteV1B1 || (exports.VoteV1B1 = {}));
exports.VoteV1B1 = VoteV1B1;
var WeightedVoteOptionV1B1 = /** @class */ (function (_super) {
    __extends(WeightedVoteOptionV1B1, _super);
    function WeightedVoteOptionV1B1(option, weight) {
        var _this = _super.call(this) || this;
        _this.option = option;
        _this.weight = new numeric_1.Dec(weight);
        return _this;
    }
    WeightedVoteOptionV1B1.fromAmino = function (data, _) {
        var option = data.option, weight = data.weight;
        return new WeightedVoteOptionV1B1(option, weight);
    };
    WeightedVoteOptionV1B1.prototype.toAmino = function (_) {
        var _a = this, option = _a.option, weight = _a.weight;
        return {
            option: option,
            weight: weight.toString(),
        };
    };
    WeightedVoteOptionV1B1.fromData = function (data, _) {
        var option = data.option, weight = data.weight;
        return new WeightedVoteOptionV1B1(option, weight);
    };
    WeightedVoteOptionV1B1.prototype.toData = function (_) {
        var _a = this, option = _a.option, weight = _a.weight;
        return {
            option: option,
            weight: weight.toString(),
        };
    };
    WeightedVoteOptionV1B1.fromProto = function (proto, _) {
        return new WeightedVoteOptionV1B1(proto.option, proto.weight);
    };
    WeightedVoteOptionV1B1.prototype.toProto = function (_) {
        var _a = this, option = _a.option, weight = _a.weight;
        return gov_1.WeightedVoteOption.fromPartial({
            option: option,
            weight: weight.toString(),
        });
    };
    return WeightedVoteOptionV1B1;
}(json_1.JSONSerializable));
exports.WeightedVoteOptionV1B1 = WeightedVoteOptionV1B1;
