// @xpla/xpla.js/dist/core/evm/v1/msgs/MsgUpdateParams.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateEvmParamsV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/ethermint/evm/v1/tx");
var Params_1 = require("../Params");
var MsgUpdateEvmParamsV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateEvmParamsV1, _super);
    /**
     * @param authority is the address that controls the module
     * @param params defines the x/bank parameters to update
     */
    function MsgUpdateEvmParamsV1(authority, params) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.params = params;
        return _this;
    }
    MsgUpdateEvmParamsV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, params = _a.params;
        return new MsgUpdateEvmParamsV1(authority, params ? Params_1.EvmParamsV1.fromAmino(params) : undefined);
    };
    MsgUpdateEvmParamsV1.prototype.toAmino = function (_isClassic) {
        var _a = this, authority = _a.authority, params = _a.params;
        return {
            type: 'ethermint/MsgUpdateParams',
            value: {
                authority: authority,
                params: params ? params.toAmino() : undefined,
            },
        };
    };
    MsgUpdateEvmParamsV1.fromData = function (data, _isClassic) {
        var authority = data.authority, params = data.params;
        return new MsgUpdateEvmParamsV1(authority, params ? Params_1.EvmParamsV1.fromData(params) : undefined);
    };
    MsgUpdateEvmParamsV1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, params = _a.params;
        return {
            '@type': '/ethermint.evm.v1.MsgUpdateParams',
            authority: authority,
            params: params ? params.toData() : undefined,
        };
    };
    MsgUpdateEvmParamsV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateEvmParamsV1(proto.authority, proto.params ? Params_1.EvmParamsV1.fromProto(proto.params) : undefined);
    };
    MsgUpdateEvmParamsV1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, params = _a.params;
        return tx_1.MsgUpdateParams.fromPartial({
            authority: authority,
            params: params ? params.toProto() : undefined,
        });
    };
    MsgUpdateEvmParamsV1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/ethermint.erc20.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    };
    MsgUpdateEvmParamsV1.unpackAny = function (msgAny, _isClassic) {
        return MsgUpdateEvmParamsV1.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    };
    return MsgUpdateEvmParamsV1;
}(json_1.JSONSerializable));
exports.MsgUpdateEvmParamsV1 = MsgUpdateEvmParamsV1;
