// @xpla/xpla.proto/google/type/date.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateMessage = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "google.type";
function createBaseDateMessage() {
    return { year: 0, month: 0, day: 0 };
}
exports.DateMessage = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.year !== 0) {
            writer.uint32(8).int32(message.year);
        }
        if (message.month !== 0) {
            writer.uint32(16).int32(message.month);
        }
        if (message.day !== 0) {
            writer.uint32(24).int32(message.day);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDateMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 8) {
                        break;
                    }
                    message.year = reader.int32();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.month = reader.int32();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.day = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            year: isSet(object.year) ? globalThis.Number(object.year) : 0,
            month: isSet(object.month) ? globalThis.Number(object.month) : 0,
            day: isSet(object.day) ? globalThis.Number(object.day) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.year !== 0) {
            obj.year = Math.round(message.year);
        }
        if (message.month !== 0) {
            obj.month = Math.round(message.month);
        }
        if (message.day !== 0) {
            obj.day = Math.round(message.day);
        }
        return obj;
    },
    create(base) {
        return exports.DateMessage.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseDateMessage();
        message.year = (_a = object.year) !== null && _a !== void 0 ? _a : 0;
        message.month = (_b = object.month) !== null && _b !== void 0 ? _b : 0;
        message.day = (_c = object.day) !== null && _c !== void 0 ? _c : 0;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
