// @xpla/xpla.proto/google/type/money.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Money = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = __importDefault(require("long"));
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "google.type";
function createBaseMoney() {
    return { currencyCode: "", units: long_1.default.ZERO, nanos: 0 };
}
exports.Money = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.currencyCode !== "") {
            writer.uint32(10).string(message.currencyCode);
        }
        if (!message.units.isZero()) {
            writer.uint32(16).int64(message.units);
        }
        if (message.nanos !== 0) {
            writer.uint32(24).int32(message.nanos);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMoney();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.currencyCode = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.units = reader.int64();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.nanos = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            currencyCode: isSet(object.currencyCode) ? globalThis.String(object.currencyCode) : "",
            units: isSet(object.units) ? long_1.default.fromValue(object.units) : long_1.default.ZERO,
            nanos: isSet(object.nanos) ? globalThis.Number(object.nanos) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.currencyCode !== "") {
            obj.currencyCode = message.currencyCode;
        }
        if (!message.units.isZero()) {
            obj.units = (message.units || long_1.default.ZERO).toString();
        }
        if (message.nanos !== 0) {
            obj.nanos = Math.round(message.nanos);
        }
        return obj;
    },
    create(base) {
        return exports.Money.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMoney();
        message.currencyCode = (_a = object.currencyCode) !== null && _a !== void 0 ? _a : "";
        message.units = (object.units !== undefined && object.units !== null) ? long_1.default.fromValue(object.units) : long_1.default.ZERO;
        message.nanos = (_b = object.nanos) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
