// @xpla/xpla.js/dist/core/xpla/v1beta1/proposals/UnregisterVolunteerValidatorProposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnregisterVolunteerValidatorProposal = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var proposal_1 = require("@xpla/xpla.proto/xpla/volunteer/v1beta1/proposal");
/**
 * Basic proposal which describes the candidate proposition that must be put into effect
 * manually if passed. Used as a general-purpose way of discovering community's
 * sentiment / interest for an arbitrary change.
 */
var UnregisterVolunteerValidatorProposal = /** @class */ (function (_super) {
    __extends(UnregisterVolunteerValidatorProposal, _super);
    /**
     * @param title proposal's title
     * @param description proposal's description
     */
    function UnregisterVolunteerValidatorProposal(title, description, validatorAddress) {
        var _this = _super.call(this) || this;
        _this.title = title;
        _this.description = description;
        _this.validatorAddress = validatorAddress;
        return _this;
    }
    UnregisterVolunteerValidatorProposal.fromAmino = function (data, _isClassic) {
        var _a = data.value, title = _a.title, description = _a.description, validatorAddress = _a.validatorAddress;
        return new UnregisterVolunteerValidatorProposal(title, description, validatorAddress);
    };
    UnregisterVolunteerValidatorProposal.prototype.toAmino = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, validatorAddress = _a.validatorAddress;
        return {
            type: 'xpla/UnregisterVolunteerValidatorProposal',
            value: {
                title: title,
                description: description,
                validatorAddress: validatorAddress,
            },
        };
    };
    UnregisterVolunteerValidatorProposal.fromData = function (proto, _isClassic) {
        var title = proto.title, description = proto.description, validatorAddress = proto.validatorAddress;
        return new UnregisterVolunteerValidatorProposal(title, description, validatorAddress);
    };
    UnregisterVolunteerValidatorProposal.prototype.toData = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, validatorAddress = _a.validatorAddress;
        return {
            '@type': '/xpla.volunteer.v1beta1.UnregisterVolunteerValidatorProposal',
            title: title,
            description: description,
            validatorAddress: validatorAddress,
        };
    };
    UnregisterVolunteerValidatorProposal.fromProto = function (proto, _isClassic) {
        return new UnregisterVolunteerValidatorProposal(proto.title, proto.description, proto.validatorAddress);
    };
    UnregisterVolunteerValidatorProposal.prototype.toProto = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, validatorAddress = _a.validatorAddress;
        return proposal_1.UnregisterVolunteerValidatorProposal.fromPartial({
            title: title,
            description: description,
            validatorAddress: validatorAddress,
        });
    };
    UnregisterVolunteerValidatorProposal.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/xpla.volunteer.v1beta1.UnregisterVolunteerValidatorProposal',
            value: proposal_1.UnregisterVolunteerValidatorProposal.encode(this.toProto(isClassic)).finish(),
        });
    };
    UnregisterVolunteerValidatorProposal.unpackAny = function (msgAny, isClassic) {
        return UnregisterVolunteerValidatorProposal.fromProto(proposal_1.UnregisterVolunteerValidatorProposal.decode(msgAny.value), isClassic);
    };
    return UnregisterVolunteerValidatorProposal;
}(json_1.JSONSerializable));
exports.UnregisterVolunteerValidatorProposal = UnregisterVolunteerValidatorProposal;
