// @xpla/xpla.js/dist/core/vesting/v1beta1/Period.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PeriodV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var Coins_1 = require("../../Coins");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var vesting_1 = require("@xpla/xpla.proto/cosmos/vesting/v1beta1/vesting");
/**
 * Period defines a length of time and amount of coins that will vest.
 */
var PeriodV1B1 = /** @class */ (function (_super) {
    __extends(PeriodV1B1, _super);
    /**
     * @param length
     * @param amount
     */
    function PeriodV1B1(length, amount) {
        var _this = _super.call(this) || this;
        _this.length = length;
        _this.amount = new Coins_1.Coins(amount);
        return _this;
    }
    PeriodV1B1.fromAmino = function (data, _isClassic) {
        var length = data.length, amount = data.amount;
        return new PeriodV1B1(Number.parseInt(length), Coins_1.Coins.fromAmino(amount));
    };
    PeriodV1B1.prototype.toAmino = function (_isClassic) {
        var _a = this, length = _a.length, amount = _a.amount;
        var res = {
            length: length.toFixed(),
            amount: amount.toAmino(),
        };
        return res;
    };
    PeriodV1B1.fromData = function (data, _isClassic) {
        var length = data.length, amount = data.amount;
        return new PeriodV1B1(Number.parseInt(length), Coins_1.Coins.fromData(amount));
    };
    PeriodV1B1.prototype.toData = function (_isClassic) {
        var _a = this, length = _a.length, amount = _a.amount;
        var res = {
            length: length.toFixed(),
            amount: amount.toData(),
        };
        return res;
    };
    PeriodV1B1.fromProto = function (proto, _isClassic) {
        return new PeriodV1B1(proto.length.toNumber(), Coins_1.Coins.fromProto(proto.amount));
    };
    PeriodV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, length = _a.length, amount = _a.amount;
        return vesting_1.Period.fromPartial({
            length: length,
            amount: amount.toProto(),
        });
    };
    PeriodV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.vesting.v1beta1.Period',
            value: vesting_1.Period.encode(this.toProto(isClassic)).finish(),
        });
    };
    PeriodV1B1.unpackAny = function (msgAny, isClassic) {
        return PeriodV1B1.fromProto(vesting_1.Period.decode(msgAny.value), isClassic);
    };
    return PeriodV1B1;
}(json_1.JSONSerializable));
exports.PeriodV1B1 = PeriodV1B1;
