// @xpla/xpla.js/dist/core/bank/v1beta1/Params.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BankParamsV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var bank_1 = require("@xpla/xpla.proto/cosmos/bank/v1beta1/bank");
var BankParamsV1B1 = /** @class */ (function (_super) {
    __extends(BankParamsV1B1, _super);
    /**
     * @param send_enabled Deprecated: Use of SendEnabled in params is deprecated.
     * @param default_send_enabled
     */
    function BankParamsV1B1(
    /** @deprecated */ send_enabled, default_send_enabled) {
        var _this = _super.call(this) || this;
        _this.send_enabled = send_enabled;
        _this.default_send_enabled = default_send_enabled;
        return _this;
    }
    BankParamsV1B1.fromAmino = function (data, _) {
        var send_enabled = data.send_enabled, default_send_enabled = data.default_send_enabled;
        return new BankParamsV1B1(send_enabled !== null && send_enabled !== void 0 ? send_enabled : [], default_send_enabled !== null && default_send_enabled !== void 0 ? default_send_enabled : false);
    };
    BankParamsV1B1.prototype.toAmino = function (_) {
        var _a = this, send_enabled = _a.send_enabled, default_send_enabled = _a.default_send_enabled;
        var res = {
            send_enabled: send_enabled,
            default_send_enabled: default_send_enabled,
        };
        return res;
    };
    BankParamsV1B1.fromData = function (data, _) {
        var send_enabled = data.send_enabled, default_send_enabled = data.default_send_enabled;
        return new BankParamsV1B1(send_enabled, default_send_enabled);
    };
    BankParamsV1B1.prototype.toData = function (_) {
        var _a = this, send_enabled = _a.send_enabled, default_send_enabled = _a.default_send_enabled;
        var res = {
            '@type': '/cosmos.bank.v1beta1.Params',
            send_enabled: send_enabled,
            default_send_enabled: default_send_enabled,
        };
        return res;
    };
    BankParamsV1B1.fromProto = function (proto, _) {
        return new BankParamsV1B1(proto.sendEnabled, proto.defaultSendEnabled);
    };
    BankParamsV1B1.prototype.toProto = function (_) {
        var _a = this, send_enabled = _a.send_enabled, default_send_enabled = _a.default_send_enabled;
        return bank_1.Params.fromPartial({
            sendEnabled: send_enabled,
            defaultSendEnabled: default_send_enabled,
        });
    };
    return BankParamsV1B1;
}(json_1.JSONSerializable));
exports.BankParamsV1B1 = BankParamsV1B1;
