// @xpla/xpla.js/dist/core/group/v1/msgs/MsgCreateGroup.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCreateGroupV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgCreateGroupV1 = /** @class */ (function (_super) {
    __extends(MsgCreateGroupV1, _super);
    /**
     * @param admin is the account address of the group admin
     * @param members defines the group members
     * @param metadata is any arbitrary metadata to attached to the group
     */
    function MsgCreateGroupV1(admin, members, metadata) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.members = members;
        _this.metadata = metadata;
        return _this;
    }
    MsgCreateGroupV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, members = _a.members, metadata = _a.metadata;
        return new MsgCreateGroupV1(admin, members, metadata);
    };
    MsgCreateGroupV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, members = _a.members, metadata = _a.metadata;
        return {
            type: isClassic ? 'group/MsgCreateGroup' : 'cosmos-sdk/MsgCreateGroup',
            value: {
                admin: admin,
                members: members,
                metadata: metadata,
            },
        };
    };
    MsgCreateGroupV1.fromData = function (data, _isClassic) {
        var admin = data.admin, members = data.members, metadata = data.metadata;
        return new MsgCreateGroupV1(admin, members, metadata);
    };
    MsgCreateGroupV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, members = _a.members, metadata = _a.metadata;
        return {
            '@type': '/cosmos.group.v1.MsgCreateGroup',
            admin: admin,
            members: members,
            metadata: metadata,
        };
    };
    MsgCreateGroupV1.fromProto = function (proto, _isClassic) {
        return new MsgCreateGroupV1(proto.admin, proto.members, proto.metadata);
    };
    MsgCreateGroupV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, members = _a.members, metadata = _a.metadata;
        return tx_1.MsgCreateGroup.fromPartial({
            admin: admin,
            members: members,
            metadata: metadata,
        });
    };
    MsgCreateGroupV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgCreateGroup',
            value: tx_1.MsgCreateGroup.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCreateGroupV1.unpackAny = function (msgAny, isClassic) {
        return MsgCreateGroupV1.fromProto(tx_1.MsgCreateGroup.decode(msgAny.value), isClassic);
    };
    return MsgCreateGroupV1;
}(json_1.JSONSerializable));
exports.MsgCreateGroupV1 = MsgCreateGroupV1;
