// @xpla/xpla.js/dist/core/group/v1/msgs/MsgCreateGroupWithPolicy.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCreateGroupWithPolicyV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgCreateGroupWithPolicyV1 = /** @class */ (function (_super) {
    __extends(MsgCreateGroupWithPolicyV1, _super);
    /**
     * @param admin is the account address of the group and group policy admin
     * @param members defines the group members
     * @param group_metadata is any arbitrary metadata attached to the group
     * @param group_policy_metadata is any arbitrary metadata attached to the group policy
     * @param group_policy_as_admin is a boolean field, if set to true, the group policy account address will be used as group and group policy admin
     * @param decision_policy specifies the group policy's decision policy
     */
    function MsgCreateGroupWithPolicyV1(admin, members, group_metadata, group_policy_metadata, group_policy_as_admin, decision_policy) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.members = members;
        _this.group_metadata = group_metadata;
        _this.group_policy_metadata = group_policy_metadata;
        _this.group_policy_as_admin = group_policy_as_admin;
        _this.decision_policy = decision_policy;
        return _this;
    }
    MsgCreateGroupWithPolicyV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, members = _a.members, group_metadata = _a.group_metadata, group_policy_metadata = _a.group_policy_metadata, group_policy_as_admin = _a.group_policy_as_admin, decision_policy = _a.decision_policy;
        return new MsgCreateGroupWithPolicyV1(admin, members, group_metadata, group_policy_metadata, group_policy_as_admin, decision_policy);
    };
    MsgCreateGroupWithPolicyV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, members = _a.members, group_metadata = _a.group_metadata, group_policy_metadata = _a.group_policy_metadata, group_policy_as_admin = _a.group_policy_as_admin, decision_policy = _a.decision_policy;
        return {
            type: isClassic
                ? 'group/MsgCreateGroupWithPolicy'
                : 'cosmos-sdk/MsgCreateGroupWithPolicy',
            value: {
                admin: admin,
                members: members,
                group_metadata: group_metadata,
                group_policy_metadata: group_policy_metadata,
                group_policy_as_admin: group_policy_as_admin,
                decision_policy: decision_policy,
            },
        };
    };
    MsgCreateGroupWithPolicyV1.fromData = function (data, _isClassic) {
        var admin = data.admin, members = data.members, group_metadata = data.group_metadata, group_policy_metadata = data.group_policy_metadata, group_policy_as_admin = data.group_policy_as_admin, decision_policy = data.decision_policy;
        return new MsgCreateGroupWithPolicyV1(admin, members, group_metadata, group_policy_metadata, group_policy_as_admin, decision_policy);
    };
    MsgCreateGroupWithPolicyV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, members = _a.members, group_metadata = _a.group_metadata, group_policy_metadata = _a.group_policy_metadata, group_policy_as_admin = _a.group_policy_as_admin, decision_policy = _a.decision_policy;
        return {
            '@type': '/cosmos.group.v1.MsgCreateGroupWithPolicy',
            admin: admin,
            members: members,
            group_metadata: group_metadata,
            group_policy_metadata: group_policy_metadata,
            group_policy_as_admin: group_policy_as_admin,
            decision_policy: decision_policy,
        };
    };
    MsgCreateGroupWithPolicyV1.fromProto = function (proto, _isClassic) {
        return new MsgCreateGroupWithPolicyV1(proto.admin, proto.members, proto.groupMetadata, proto.groupPolicyMetadata, proto.groupPolicyAsAdmin, proto.decisionPolicy);
    };
    MsgCreateGroupWithPolicyV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, members = _a.members, group_metadata = _a.group_metadata, group_policy_metadata = _a.group_policy_metadata, group_policy_as_admin = _a.group_policy_as_admin, decision_policy = _a.decision_policy;
        return tx_1.MsgCreateGroupWithPolicy.fromPartial({
            admin: admin,
            members: members,
            groupMetadata: group_metadata,
            groupPolicyMetadata: group_policy_metadata,
            groupPolicyAsAdmin: group_policy_as_admin,
            decisionPolicy: decision_policy,
        });
    };
    MsgCreateGroupWithPolicyV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgCreateGroupWithPolicy',
            value: tx_1.MsgCreateGroupWithPolicy.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCreateGroupWithPolicyV1.unpackAny = function (msgAny, isClassic) {
        return MsgCreateGroupWithPolicyV1.fromProto(tx_1.MsgCreateGroupWithPolicy.decode(msgAny.value), isClassic);
    };
    return MsgCreateGroupWithPolicyV1;
}(json_1.JSONSerializable));
exports.MsgCreateGroupWithPolicyV1 = MsgCreateGroupWithPolicyV1;
