// @xpla/xpla.js/dist/core/group/v1/msgs/MsgExec.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgGroupExecV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgGroupExecV1 = /** @class */ (function (_super) {
    __extends(MsgGroupExecV1, _super);
    /**
     * @param proposal_id is the unique ID of the proposal
     * @param executor is the account address used to execute the proposal
     */
    function MsgGroupExecV1(proposal_id, executor) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.executor = executor;
        return _this;
    }
    MsgGroupExecV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, proposal_id = _a.proposal_id, executor = _a.executor;
        return new MsgGroupExecV1(Number.parseInt(proposal_id), executor);
    };
    MsgGroupExecV1.prototype.toAmino = function (isClassic) {
        var _a = this, proposal_id = _a.proposal_id, executor = _a.executor;
        return {
            type: isClassic ? 'group/MsgExec' : 'cosmos-sdk/MsgExec',
            value: {
                proposal_id: proposal_id.toFixed(),
                executor: executor,
            },
        };
    };
    MsgGroupExecV1.fromData = function (data, _isClassic) {
        var proposal_id = data.proposal_id, executor = data.executor;
        return new MsgGroupExecV1(Number.parseInt(proposal_id), executor);
    };
    MsgGroupExecV1.prototype.toData = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, executor = _a.executor;
        return {
            '@type': '/cosmos.group.v1.MsgExec',
            proposal_id: proposal_id.toFixed(),
            executor: executor,
        };
    };
    MsgGroupExecV1.fromProto = function (proto, _isClassic) {
        return new MsgGroupExecV1(proto.proposalId.toNumber(), proto.executor);
    };
    MsgGroupExecV1.prototype.toProto = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, executor = _a.executor;
        return tx_1.MsgExec.fromPartial({
            proposalId: proposal_id,
            executor: executor,
        });
    };
    MsgGroupExecV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgExec',
            value: tx_1.MsgExec.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgGroupExecV1.unpackAny = function (msgAny, isClassic) {
        return MsgGroupExecV1.fromProto(tx_1.MsgExec.decode(msgAny.value), isClassic);
    };
    return MsgGroupExecV1;
}(json_1.JSONSerializable));
exports.MsgGroupExecV1 = MsgGroupExecV1;
