// @xpla/xpla.js/dist/core/group/v1/msgs/MsgLeaveGroup.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgLeaveGroupV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgLeaveGroupV1 = /** @class */ (function (_super) {
    __extends(MsgLeaveGroupV1, _super);
    /**
     * @param address is the account address of the group member
     * @param group_id is the unique ID of the group
     */
    function MsgLeaveGroupV1(address, group_id) {
        var _this = _super.call(this) || this;
        _this.address = address;
        _this.group_id = group_id;
        return _this;
    }
    MsgLeaveGroupV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, address = _a.address, group_id = _a.group_id;
        return new MsgLeaveGroupV1(address, Number.parseInt(group_id));
    };
    MsgLeaveGroupV1.prototype.toAmino = function (isClassic) {
        var _a = this, address = _a.address, group_id = _a.group_id;
        return {
            type: isClassic ? 'group/MsgLeaveGroup' : 'cosmos-sdk/MsgLeaveGroup',
            value: {
                address: address,
                group_id: group_id.toFixed(),
            },
        };
    };
    MsgLeaveGroupV1.fromData = function (data, _isClassic) {
        var address = data.address, group_id = data.group_id;
        return new MsgLeaveGroupV1(address, Number.parseInt(group_id));
    };
    MsgLeaveGroupV1.prototype.toData = function (_isClassic) {
        var _a = this, address = _a.address, group_id = _a.group_id;
        return {
            '@type': '/cosmos.group.v1.MsgLeaveGroup',
            address: address,
            group_id: group_id.toFixed(),
        };
    };
    MsgLeaveGroupV1.fromProto = function (proto, _isClassic) {
        return new MsgLeaveGroupV1(proto.address, proto.groupId.toNumber());
    };
    MsgLeaveGroupV1.prototype.toProto = function (_isClassic) {
        var _a = this, address = _a.address, group_id = _a.group_id;
        return tx_1.MsgLeaveGroup.fromPartial({
            address: address,
            groupId: group_id,
        });
    };
    MsgLeaveGroupV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgLeaveGroup',
            value: tx_1.MsgLeaveGroup.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgLeaveGroupV1.unpackAny = function (msgAny, isClassic) {
        return MsgLeaveGroupV1.fromProto(tx_1.MsgLeaveGroup.decode(msgAny.value), isClassic);
    };
    return MsgLeaveGroupV1;
}(json_1.JSONSerializable));
exports.MsgLeaveGroupV1 = MsgLeaveGroupV1;
