// @xpla/xpla.js/dist/core/group/v1/msgs/MsgUpdateGroupAdmin.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupAdminV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgUpdateGroupAdminV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateGroupAdminV1, _super);
    /**
     * @param admin is the current account address of the group admin
     * @param group_id is the unique ID of the group
     * @param new_admin is the group new admin account address
     */
    function MsgUpdateGroupAdminV1(admin, group_id, new_admin) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.group_id = group_id;
        _this.new_admin = new_admin;
        return _this;
    }
    MsgUpdateGroupAdminV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, group_id = _a.group_id, new_admin = _a.new_admin;
        return new MsgUpdateGroupAdminV1(admin, Number.parseInt(group_id), new_admin);
    };
    MsgUpdateGroupAdminV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, new_admin = _a.new_admin;
        return {
            type: isClassic
                ? 'group/MsgUpdateGroupAdmin'
                : 'cosmos-sdk/MsgUpdateGroupAdmin',
            value: {
                admin: admin,
                group_id: group_id.toString(),
                new_admin: new_admin,
            },
        };
    };
    MsgUpdateGroupAdminV1.fromData = function (data, _isClassic) {
        var admin = data.admin, group_id = data.group_id, new_admin = data.new_admin;
        return new MsgUpdateGroupAdminV1(admin, Number.parseInt(group_id), new_admin);
    };
    MsgUpdateGroupAdminV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, new_admin = _a.new_admin;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupAdmin',
            admin: admin,
            group_id: group_id.toString(),
            new_admin: new_admin,
        };
    };
    MsgUpdateGroupAdminV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateGroupAdminV1(proto.admin, proto.groupId.toNumber(), proto.newAdmin);
    };
    MsgUpdateGroupAdminV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, new_admin = _a.new_admin;
        return tx_1.MsgUpdateGroupAdmin.fromPartial({
            admin: admin,
            groupId: group_id,
            newAdmin: new_admin,
        });
    };
    MsgUpdateGroupAdminV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupAdmin',
            value: tx_1.MsgUpdateGroupAdmin.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateGroupAdminV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateGroupAdminV1.fromProto(tx_1.MsgUpdateGroupAdmin.decode(msgAny.value), isClassic);
    };
    return MsgUpdateGroupAdminV1;
}(json_1.JSONSerializable));
exports.MsgUpdateGroupAdminV1 = MsgUpdateGroupAdminV1;
