// @xpla/xpla.js/dist/core/group/v1/msgs/MsgUpdateGroupMembers.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupMembersV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgUpdateGroupMembersV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateGroupMembersV1, _super);
    /**
     * @param admin is the account address of the group admin
     * @param group_id is the unique ID of the group
     * @param member_updates is the list of members to update
     */
    function MsgUpdateGroupMembersV1(admin, group_id, member_updates) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.group_id = group_id;
        _this.member_updates = member_updates;
        return _this;
    }
    MsgUpdateGroupMembersV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, group_id = _a.group_id, member_updates = _a.member_updates;
        return new MsgUpdateGroupMembersV1(admin, Number.parseInt(group_id), member_updates);
    };
    MsgUpdateGroupMembersV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, member_updates = _a.member_updates;
        return {
            type: isClassic
                ? 'group/MsgUpdateGroupMembers'
                : 'cosmos-sdk/MsgUpdateGroupMembers',
            value: {
                admin: admin,
                group_id: group_id.toFixed(),
                member_updates: member_updates,
            },
        };
    };
    MsgUpdateGroupMembersV1.fromData = function (data, _isClassic) {
        var admin = data.admin, group_id = data.group_id, member_updates = data.member_updates;
        return new MsgUpdateGroupMembersV1(admin, Number.parseInt(group_id), member_updates);
    };
    MsgUpdateGroupMembersV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, member_updates = _a.member_updates;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupMembers',
            admin: admin,
            group_id: group_id.toFixed(),
            member_updates: member_updates,
        };
    };
    MsgUpdateGroupMembersV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateGroupMembersV1(proto.admin, proto.groupId.toNumber(), proto.memberUpdates);
    };
    MsgUpdateGroupMembersV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, member_updates = _a.member_updates;
        return tx_1.MsgUpdateGroupMembers.fromPartial({
            admin: admin,
            groupId: group_id,
            memberUpdates: member_updates,
        });
    };
    MsgUpdateGroupMembersV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupMembers',
            value: tx_1.MsgUpdateGroupMembers.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateGroupMembersV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateGroupMembersV1.fromProto(tx_1.MsgUpdateGroupMembers.decode(msgAny.value), isClassic);
    };
    return MsgUpdateGroupMembersV1;
}(json_1.JSONSerializable));
exports.MsgUpdateGroupMembersV1 = MsgUpdateGroupMembersV1;
