// @xpla/xpla.js/dist/core/group/v1/msgs/MsgUpdateGroupPolicyDecisionPolicy.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupPolicyDecisionPolicyV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgUpdateGroupPolicyDecisionPolicyV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateGroupPolicyDecisionPolicyV1, _super);
    /**
     * @param authority
     * @param send_enabled is the list of entries to add or update
     * @param use_default_for is a list of denoms that should use the params.default_send_enabled value
     */
    function MsgUpdateGroupPolicyDecisionPolicyV1(admin, group_policy_address, decision_policy) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.group_policy_address = group_policy_address;
        _this.decision_policy = decision_policy;
        return _this;
    }
    MsgUpdateGroupPolicyDecisionPolicyV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, group_policy_address = _a.group_policy_address, decision_policy = _a.decision_policy;
        return new MsgUpdateGroupPolicyDecisionPolicyV1(admin, group_policy_address, decision_policy);
    };
    MsgUpdateGroupPolicyDecisionPolicyV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, decision_policy = _a.decision_policy;
        return {
            type: isClassic
                ? 'group/MsgUpdateGroupPolicyDecisionPolicy'
                : 'cosmos-sdk/MsgUpdateGroupPolicyDecisionPolicy',
            value: {
                admin: admin,
                group_policy_address: group_policy_address,
                decision_policy: decision_policy,
            },
        };
    };
    MsgUpdateGroupPolicyDecisionPolicyV1.fromData = function (data, _isClassic) {
        var admin = data.admin, group_policy_address = data.group_policy_address, decision_policy = data.decision_policy;
        return new MsgUpdateGroupPolicyDecisionPolicyV1(admin, group_policy_address, decision_policy);
    };
    MsgUpdateGroupPolicyDecisionPolicyV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, decision_policy = _a.decision_policy;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupPolicyDecisionPolicy',
            admin: admin,
            group_policy_address: group_policy_address,
            decision_policy: decision_policy,
        };
    };
    MsgUpdateGroupPolicyDecisionPolicyV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateGroupPolicyDecisionPolicyV1(proto.admin, proto.groupPolicyAddress, proto.decisionPolicy);
    };
    MsgUpdateGroupPolicyDecisionPolicyV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, decision_policy = _a.decision_policy;
        return tx_1.MsgUpdateGroupPolicyDecisionPolicy.fromPartial({
            admin: admin,
            groupPolicyAddress: group_policy_address,
            decisionPolicy: decision_policy,
        });
    };
    MsgUpdateGroupPolicyDecisionPolicyV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupPolicyDecisionPolicy',
            value: tx_1.MsgUpdateGroupPolicyDecisionPolicy.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateGroupPolicyDecisionPolicyV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateGroupPolicyDecisionPolicyV1.fromProto(tx_1.MsgUpdateGroupPolicyDecisionPolicy.decode(msgAny.value), isClassic);
    };
    return MsgUpdateGroupPolicyDecisionPolicyV1;
}(json_1.JSONSerializable));
exports.MsgUpdateGroupPolicyDecisionPolicyV1 = MsgUpdateGroupPolicyDecisionPolicyV1;
