// @xpla/xpla.js/dist/core/group/v1/msgs/MsgUpdateGroupMetadata.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupMetadataV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgUpdateGroupMetadataV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateGroupMetadataV1, _super);
    /**
     * @param authority
     * @param send_enabled is the list of entries to add or update
     * @param use_default_for is a list of denoms that should use the params.default_send_enabled value
     */
    function MsgUpdateGroupMetadataV1(admin, group_id, metadata) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.group_id = group_id;
        _this.metadata = metadata;
        return _this;
    }
    MsgUpdateGroupMetadataV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, group_id = _a.group_id, metadata = _a.metadata;
        return new MsgUpdateGroupMetadataV1(admin, Number.parseInt(group_id), metadata);
    };
    MsgUpdateGroupMetadataV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, metadata = _a.metadata;
        return {
            type: isClassic
                ? 'group/MsgUpdateGroupMetadata'
                : 'cosmos-sdk/MsgUpdateGroupMetadata',
            value: {
                admin: admin,
                group_id: group_id.toFixed(),
                metadata: metadata,
            },
        };
    };
    MsgUpdateGroupMetadataV1.fromData = function (data, _isClassic) {
        var admin = data.admin, group_id = data.group_id, metadata = data.metadata;
        return new MsgUpdateGroupMetadataV1(admin, Number.parseInt(group_id), metadata);
    };
    MsgUpdateGroupMetadataV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, metadata = _a.metadata;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupMetadata',
            admin: admin,
            group_id: group_id.toFixed(),
            metadata: metadata,
        };
    };
    MsgUpdateGroupMetadataV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateGroupMetadataV1(proto.admin, proto.groupId.toNumber(), proto.metadata);
    };
    MsgUpdateGroupMetadataV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, group_id = _a.group_id, metadata = _a.metadata;
        return tx_1.MsgUpdateGroupMetadata.fromPartial({
            admin: admin,
            groupId: group_id,
            metadata: metadata,
        });
    };
    MsgUpdateGroupMetadataV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupMetadata',
            value: tx_1.MsgUpdateGroupMetadata.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateGroupMetadataV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateGroupMetadataV1.fromProto(tx_1.MsgUpdateGroupMetadata.decode(msgAny.value), isClassic);
    };
    return MsgUpdateGroupMetadataV1;
}(json_1.JSONSerializable));
exports.MsgUpdateGroupMetadataV1 = MsgUpdateGroupMetadataV1;
