// @xpla/xpla.js/dist/core/group/v1/msgs/MsgUpdateGroupPolicyAdmin.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupPolicyAdminV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgUpdateGroupPolicyAdminV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateGroupPolicyAdminV1, _super);
    /**
     * @param admin is the account address of the group admin
     * @param group_policy_address is the account address of the group policy
     * @param new_admin is the new group policy admin
     */
    function MsgUpdateGroupPolicyAdminV1(admin, group_policy_address, new_admin) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.group_policy_address = group_policy_address;
        _this.new_admin = new_admin;
        return _this;
    }
    MsgUpdateGroupPolicyAdminV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, group_policy_address = _a.group_policy_address, new_admin = _a.new_admin;
        return new MsgUpdateGroupPolicyAdminV1(admin, group_policy_address, new_admin);
    };
    MsgUpdateGroupPolicyAdminV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, new_admin = _a.new_admin;
        return {
            type: isClassic
                ? 'group/MsgUpdateGroupPolicyAdmin'
                : 'cosmos-sdk/MsgUpdateGroupPolicyAdmin',
            value: {
                admin: admin,
                group_policy_address: group_policy_address,
                new_admin: new_admin,
            },
        };
    };
    MsgUpdateGroupPolicyAdminV1.fromData = function (data, _isClassic) {
        var admin = data.admin, group_policy_address = data.group_policy_address, new_admin = data.new_admin;
        return new MsgUpdateGroupPolicyAdminV1(admin, group_policy_address, new_admin);
    };
    MsgUpdateGroupPolicyAdminV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, new_admin = _a.new_admin;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupPolicyAdmin',
            admin: admin,
            group_policy_address: group_policy_address,
            new_admin: new_admin,
        };
    };
    MsgUpdateGroupPolicyAdminV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateGroupPolicyAdminV1(proto.admin, proto.groupPolicyAddress, proto.newAdmin);
    };
    MsgUpdateGroupPolicyAdminV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, new_admin = _a.new_admin;
        return tx_1.MsgUpdateGroupPolicyAdmin.fromPartial({
            admin: admin,
            groupPolicyAddress: group_policy_address,
            newAdmin: new_admin,
        });
    };
    MsgUpdateGroupPolicyAdminV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupPolicyAdmin',
            value: tx_1.MsgUpdateGroupPolicyAdmin.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateGroupPolicyAdminV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateGroupPolicyAdminV1.fromProto(tx_1.MsgUpdateGroupPolicyAdmin.decode(msgAny.value), isClassic);
    };
    return MsgUpdateGroupPolicyAdminV1;
}(json_1.JSONSerializable));
exports.MsgUpdateGroupPolicyAdminV1 = MsgUpdateGroupPolicyAdminV1;
