// @xpla/xpla.js/dist/core/group/v1/msgs/MsgUpdateGroupPolicyMetadata.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateGroupPolicyMetadataV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgUpdateGroupPolicyMetadataV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateGroupPolicyMetadataV1, _super);
    /**
     * @param admin is the account address of the group admin
     * @param group_policy_address is the account address of group policy
     * @param metadata is the group policy metadata to be updated
     */
    function MsgUpdateGroupPolicyMetadataV1(admin, group_policy_address, metadata) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.group_policy_address = group_policy_address;
        _this.metadata = metadata;
        return _this;
    }
    MsgUpdateGroupPolicyMetadataV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, admin = _a.admin, group_policy_address = _a.group_policy_address, metadata = _a.metadata;
        return new MsgUpdateGroupPolicyMetadataV1(admin, group_policy_address, metadata);
    };
    MsgUpdateGroupPolicyMetadataV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, metadata = _a.metadata;
        return {
            type: isClassic
                ? 'group/MsgUpdateGroupPolicyMetadata'
                : 'cosmos-sdk/MsgUpdateGroupPolicyMetadata',
            value: {
                admin: admin,
                group_policy_address: group_policy_address,
                metadata: metadata,
            },
        };
    };
    MsgUpdateGroupPolicyMetadataV1.fromData = function (data, _isClassic) {
        var admin = data.admin, group_policy_address = data.group_policy_address, metadata = data.metadata;
        return new MsgUpdateGroupPolicyMetadataV1(admin, group_policy_address, metadata);
    };
    MsgUpdateGroupPolicyMetadataV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, metadata = _a.metadata;
        return {
            '@type': '/cosmos.group.v1.MsgUpdateGroupPolicyMetadata',
            admin: admin,
            group_policy_address: group_policy_address,
            metadata: metadata,
        };
    };
    MsgUpdateGroupPolicyMetadataV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateGroupPolicyMetadataV1(proto.admin, proto.groupPolicyAddress, proto.metadata);
    };
    MsgUpdateGroupPolicyMetadataV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, group_policy_address = _a.group_policy_address, metadata = _a.metadata;
        return tx_1.MsgUpdateGroupPolicyMetadata.fromPartial({
            admin: admin,
            groupPolicyAddress: group_policy_address,
            metadata: metadata,
        });
    };
    MsgUpdateGroupPolicyMetadataV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgUpdateGroupPolicyMetadata',
            value: tx_1.MsgUpdateGroupPolicyMetadata.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateGroupPolicyMetadataV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateGroupPolicyMetadataV1.fromProto(tx_1.MsgUpdateGroupPolicyMetadata.decode(msgAny.value), isClassic);
    };
    return MsgUpdateGroupPolicyMetadataV1;
}(json_1.JSONSerializable));
exports.MsgUpdateGroupPolicyMetadataV1 = MsgUpdateGroupPolicyMetadataV1;
