// @xpla/xpla.js/dist/core/group/v1/msgs/MsgVote.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgGroupVoteV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgGroupVoteV1 = /** @class */ (function (_super) {
    __extends(MsgGroupVoteV1, _super);
    /**
     * @param proposal_id is the unique ID of the proposal
     * @param voter is the voter account address
     * @param option is the voter's choice on the proposal
     * @param metadata is any arbitrary metadata attached to the vote
     * @param exec defines whether the proposal should be executed immediately after voting or not
     */
    function MsgGroupVoteV1(proposal_id, voter, option, metadata, exec) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.voter = voter;
        _this.option = option;
        _this.metadata = metadata;
        _this.exec = exec;
        return _this;
    }
    MsgGroupVoteV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata, exec = _a.exec;
        return new MsgGroupVoteV1(Number.parseInt(proposal_id), voter, option, metadata, exec);
    };
    MsgGroupVoteV1.prototype.toAmino = function (isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata, exec = _a.exec;
        return {
            type: isClassic ? 'group/MsgVote' : 'cosmos-sdk/MsgVote',
            value: {
                proposal_id: proposal_id.toFixed(),
                voter: voter,
                option: option,
                metadata: metadata,
                exec: exec,
            },
        };
    };
    MsgGroupVoteV1.fromData = function (data, _isClassic) {
        var proposal_id = data.proposal_id, voter = data.voter, option = data.option, metadata = data.metadata, exec = data.exec;
        return new MsgGroupVoteV1(Number.parseInt(proposal_id), voter, option, metadata, exec);
    };
    MsgGroupVoteV1.prototype.toData = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata, exec = _a.exec;
        return {
            '@type': '/cosmos.group.v1.MsgVote',
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            option: option,
            metadata: metadata,
            exec: exec,
        };
    };
    MsgGroupVoteV1.fromProto = function (proto, _isClassic) {
        return new MsgGroupVoteV1(proto.proposalId.toNumber(), proto.voter, proto.option, proto.metadata, proto.exec);
    };
    MsgGroupVoteV1.prototype.toProto = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata, exec = _a.exec;
        return tx_1.MsgVote.fromPartial({
            proposalId: proposal_id,
            voter: voter,
            option: option,
            metadata: metadata,
            exec: exec,
        });
    };
    MsgGroupVoteV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgVote',
            value: tx_1.MsgVote.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgGroupVoteV1.unpackAny = function (msgAny, isClassic) {
        return MsgGroupVoteV1.fromProto(tx_1.MsgVote.decode(msgAny.value), isClassic);
    };
    return MsgGroupVoteV1;
}(json_1.JSONSerializable));
exports.MsgGroupVoteV1 = MsgGroupVoteV1;
