// @xpla/xpla.js/dist/core/group/v1/msgs/MsgWithdrawProposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgGroupWithdrawProposalV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/group/v1/tx");
var MsgGroupWithdrawProposalV1 = /** @class */ (function (_super) {
    __extends(MsgGroupWithdrawProposalV1, _super);
    /**
     * @param proposal is the unique ID of the proposal
     * @param address is the admin of the group policy or one of the proposer of the proposal
     */
    function MsgGroupWithdrawProposalV1(proposal_id, address) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.address = address;
        return _this;
    }
    MsgGroupWithdrawProposalV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, proposal_id = _a.proposal_id, address = _a.address;
        return new MsgGroupWithdrawProposalV1(Number.parseInt(proposal_id), address);
    };
    MsgGroupWithdrawProposalV1.prototype.toAmino = function (isClassic) {
        var _a = this, proposal_id = _a.proposal_id, address = _a.address;
        return {
            type: isClassic
                ? 'group/MsgWithdrawProposal'
                : 'cosmos-sdk/MsgWithdrawProposal',
            value: {
                proposal_id: proposal_id.toFixed(),
                address: address,
            },
        };
    };
    MsgGroupWithdrawProposalV1.fromData = function (data, _isClassic) {
        var proposal_id = data.proposal_id, address = data.address;
        return new MsgGroupWithdrawProposalV1(Number.parseInt(proposal_id), address);
    };
    MsgGroupWithdrawProposalV1.prototype.toData = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, address = _a.address;
        return {
            '@type': '/cosmos.group.v1.MsgWithdrawProposal',
            proposal_id: proposal_id.toFixed(),
            address: address,
        };
    };
    MsgGroupWithdrawProposalV1.fromProto = function (proto, _isClassic) {
        return new MsgGroupWithdrawProposalV1(proto.proposalId.toNumber(), proto.address);
    };
    MsgGroupWithdrawProposalV1.prototype.toProto = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, address = _a.address;
        return tx_1.MsgWithdrawProposal.fromPartial({
            proposalId: proposal_id,
            address: address,
        });
    };
    MsgGroupWithdrawProposalV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.group.v1.MsgWithdrawProposal',
            value: tx_1.MsgWithdrawProposal.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgGroupWithdrawProposalV1.unpackAny = function (msgAny, isClassic) {
        return MsgGroupWithdrawProposalV1.fromProto(tx_1.MsgWithdrawProposal.decode(msgAny.value), isClassic);
    };
    return MsgGroupWithdrawProposalV1;
}(json_1.JSONSerializable));
exports.MsgGroupWithdrawProposalV1 = MsgGroupWithdrawProposalV1;
