// @xpla/xpla.js/dist/core/ibc/applications/fee/msgs/MsgRegisterPayee.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgRegisterPayee = void 0;
var json_1 = require("../../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/ibc/applications/fee/v1/tx");
/**
 * MsgRegisterCounterpartyAddress defines the request type for the RegisterCounterpartyAddress rpc
 */
var MsgRegisterPayee = /** @class */ (function (_super) {
    __extends(MsgRegisterPayee, _super);
    function MsgRegisterPayee(
    /** unique port identifier */
    portId, 
    /** unique channel identifier */
    channelId, 
    /** the relayer address */
    relayer, 
    /** the payee address */
    payee) {
        var _this = _super.call(this) || this;
        _this.portId = portId;
        _this.channelId = channelId;
        _this.relayer = relayer;
        _this.payee = payee;
        return _this;
    }
    MsgRegisterPayee.fromAmino = function (_, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        throw new Error('Amino not supported');
    };
    MsgRegisterPayee.prototype.toAmino = function (isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        throw new Error('Amino not supported');
    };
    MsgRegisterPayee.fromData = function (data, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        var portId = data.portId, channelId = data.channelId, relayer = data.relayer, payee = data.payee;
        return new MsgRegisterPayee(portId, channelId, relayer, payee);
    };
    MsgRegisterPayee.prototype.toData = function (isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        var _a = this, portId = _a.portId, channelId = _a.channelId, relayer = _a.relayer, payee = _a.payee;
        return {
            '@type': '/ibc.applications.fee.v1.MsgRegisterCounterpartyPayee',
            portId: portId,
            channelId: channelId,
            relayer: relayer,
            payee: payee,
        };
    };
    MsgRegisterPayee.fromProto = function (proto, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        return new MsgRegisterPayee(proto.portId, proto.channelId, proto.relayer, proto.payee);
    };
    MsgRegisterPayee.prototype.toProto = function (isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        var _a = this, portId = _a.portId, channelId = _a.channelId, relayer = _a.relayer, payee = _a.payee;
        return tx_1.MsgRegisterPayee.fromPartial({
            portId: portId,
            channelId: channelId,
            relayer: relayer,
            payee: payee,
        });
    };
    MsgRegisterPayee.prototype.packAny = function (isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        return any_1.Any.fromPartial({
            typeUrl: '/ibc.applications.fee.v1.MsgRegisterCounterpartyAddress',
            value: tx_1.MsgRegisterPayee.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgRegisterPayee.unpackAny = function (msgAny, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        return MsgRegisterPayee.fromProto(tx_1.MsgRegisterPayee.decode(msgAny.value));
    };
    return MsgRegisterPayee;
}(json_1.JSONSerializable));
exports.MsgRegisterPayee = MsgRegisterPayee;
