// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgAddCodeUploadParamsAddresses.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgAddCodeUploadParamsAddressesV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgAddCodeUploadParamsAddressesV1 = /** @class */ (function (_super) {
    __extends(MsgAddCodeUploadParamsAddressesV1, _super);
    /**
     * @param authority is the address of the governance account
     * @param addresses
     */
    function MsgAddCodeUploadParamsAddressesV1(authority, addresses) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.addresses = addresses;
        return _this;
    }
    MsgAddCodeUploadParamsAddressesV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, addresses = _a.addresses;
        return new MsgAddCodeUploadParamsAddressesV1(authority, addresses);
    };
    MsgAddCodeUploadParamsAddressesV1.prototype.toAmino = function (_isClassic) {
        var _a = this, authority = _a.authority, addresses = _a.addresses;
        return {
            type: 'wasm/MsgAddCodeUploadParamsAddresses',
            value: {
                authority: authority,
                addresses: addresses,
            },
        };
    };
    MsgAddCodeUploadParamsAddressesV1.fromProto = function (proto, _isClassic) {
        return new MsgAddCodeUploadParamsAddressesV1(proto.authority, proto.addresses);
    };
    MsgAddCodeUploadParamsAddressesV1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, addresses = _a.addresses;
        return tx_1.MsgAddCodeUploadParamsAddresses.fromPartial({
            authority: authority,
            addresses: addresses,
        });
    };
    MsgAddCodeUploadParamsAddressesV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgAddCodeUploadParamsAddresses',
            value: tx_1.MsgAddCodeUploadParamsAddresses.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgAddCodeUploadParamsAddressesV1.unpackAny = function (msgAny, isClassic) {
        return MsgAddCodeUploadParamsAddressesV1.fromProto(tx_1.MsgAddCodeUploadParamsAddresses.decode(msgAny.value), isClassic);
    };
    MsgAddCodeUploadParamsAddressesV1.fromData = function (data, _isClassic) {
        var authority = data.authority, addresses = data.addresses;
        return new MsgAddCodeUploadParamsAddressesV1(authority, addresses);
    };
    MsgAddCodeUploadParamsAddressesV1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, addresses = _a.addresses;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgAddCodeUploadParamsAddresses',
            authority: authority,
            addresses: addresses,
        };
    };
    return MsgAddCodeUploadParamsAddressesV1;
}(json_1.JSONSerializable));
exports.MsgAddCodeUploadParamsAddressesV1 = MsgAddCodeUploadParamsAddressesV1;
