// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgClearContractAdmin.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgClearContractAdminV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgClearContractAdminV1 = /** @class */ (function (_super) {
    __extends(MsgClearContractAdminV1, _super);
    /**
     * @param admin contract admin
     * @param contract contract address
     */
    function MsgClearContractAdminV1(admin, contract) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.contract = contract;
        return _this;
    }
    MsgClearContractAdminV1.fromAmino = function (data, isClassic) {
        if (isClassic) {
            var _a = data.value, admin = _a.admin, contract = _a.contract;
            return new MsgClearContractAdminV1(admin, contract);
        }
        else {
            var _b = data.value, sender = _b.sender, contract = _b.contract;
            return new MsgClearContractAdminV1(sender, contract);
        }
    };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    MsgClearContractAdminV1.prototype.toAmino = function (isClassic) {
        var _a = this, admin = _a.admin, contract = _a.contract;
        if (isClassic) {
            return {
                type: 'wasm/MsgClearContractAdmin',
                value: {
                    admin: admin,
                    contract: contract,
                },
            };
        }
        else {
            return {
                type: 'wasm/MsgClearAdmin',
                value: {
                    sender: admin,
                    contract: contract,
                },
            };
        }
    };
    MsgClearContractAdminV1.fromProto = function (data, _) {
        var _a = data, sender = _a.sender, contract = _a.contract;
        return new MsgClearContractAdminV1(sender, contract);
    };
    MsgClearContractAdminV1.prototype.toProto = function (_) {
        return tx_1.MsgClearAdmin.fromPartial({
            sender: this.admin,
            contract: this.contract,
        });
    };
    MsgClearContractAdminV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgClearAdmin',
            value: tx_1.MsgClearAdmin.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgClearContractAdminV1.unpackAny = function (msgAny, isClassic) {
        return MsgClearContractAdminV1.fromProto(tx_1.MsgClearAdmin.decode(msgAny.value), isClassic);
    };
    MsgClearContractAdminV1.fromData = function (data, _) {
        var sender = data.sender, contract = data.contract;
        return new MsgClearContractAdminV1(sender, contract);
    };
    MsgClearContractAdminV1.prototype.toData = function (_) {
        return {
            '@type': '/cosmwasm.wasm.v1.MsgClearAdmin',
            sender: this.admin,
            contract: this.contract,
        };
    };
    return MsgClearContractAdminV1;
}(json_1.JSONSerializable));
exports.MsgClearContractAdminV1 = MsgClearContractAdminV1;
