// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgMigrateContract.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgMigrateContractV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgMigrateContractV1 = /** @class */ (function (_super) {
    __extends(MsgMigrateContractV1, _super);
    /**
     * @param admin contract admin
     * @param contract contract address to be migrated from
     * @param new_code_id reference to the new code on the blockchain
     * @param migrate_msg JSON message to configure the migrate state of the contract
     */
    function MsgMigrateContractV1(admin, contract, new_code_id, migrate_msg // json object or string
    ) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.contract = contract;
        _this.new_code_id = new_code_id;
        _this.migrate_msg = migrate_msg;
        return _this;
    }
    MsgMigrateContractV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, sender = _a.sender, contract = _a.contract, code_id = _a.code_id, msg = _a.msg;
        return new MsgMigrateContractV1(sender, contract, Number.parseInt(code_id), msg);
    };
    MsgMigrateContractV1.prototype.toAmino = function (_isClassic) {
        var _a = this, admin = _a.admin, contract = _a.contract, new_code_id = _a.new_code_id, migrate_msg = _a.migrate_msg;
        return {
            type: 'wasm/MsgMigrateContract',
            value: {
                sender: admin,
                contract: contract,
                code_id: new_code_id.toFixed(),
                msg: (0, json_1.removeNull)(migrate_msg),
            },
        };
    };
    MsgMigrateContractV1.fromProto = function (proto, _isClassic) {
        return new MsgMigrateContractV1(proto.sender, proto.contract, proto.codeId.toNumber(), JSON.parse(Buffer.from(proto.msg).toString('utf-8')));
    };
    MsgMigrateContractV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, contract = _a.contract, new_code_id = _a.new_code_id, migrate_msg = _a.migrate_msg;
        return tx_1.MsgMigrateContract.fromPartial({
            sender: admin,
            contract: contract,
            codeId: new_code_id,
            msg: Buffer.from(JSON.stringify(migrate_msg), 'utf-8'),
        });
    };
    MsgMigrateContractV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgMigrateContract',
            value: tx_1.MsgMigrateContract.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgMigrateContractV1.unpackAny = function (msgAny, isClassic) {
        return MsgMigrateContractV1.fromProto(tx_1.MsgMigrateContract.decode(msgAny.value), isClassic);
    };
    MsgMigrateContractV1.fromData = function (data, _isClassic) {
        var sender = data.sender, contract = data.contract, code_id = data.code_id, msg = data.msg;
        return new MsgMigrateContractV1(sender, contract, Number.parseInt(code_id), msg);
    };
    MsgMigrateContractV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, contract = _a.contract, new_code_id = _a.new_code_id, migrate_msg = _a.migrate_msg;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgMigrateContract',
            sender: admin,
            contract: contract,
            code_id: new_code_id.toFixed(),
            msg: (0, json_1.removeNull)(migrate_msg),
        };
    };
    return MsgMigrateContractV1;
}(json_1.JSONSerializable));
exports.MsgMigrateContractV1 = MsgMigrateContractV1;
