// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgSudoContract.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSudoContractV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgSudoContractV1 = /** @class */ (function (_super) {
    __extends(MsgSudoContractV1, _super);
    /**
     * @param sender is the that actor that signed the messages
     * @param code_id references the stored WASM code
     * @param new_instantiate_permission is the new access control
     */
    function MsgSudoContractV1(authority, contract, msg) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.contract = contract;
        _this.msg = msg;
        return _this;
    }
    MsgSudoContractV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, contract = _a.contract, msg = _a.msg;
        return new MsgSudoContractV1(authority, contract, msg);
    };
    MsgSudoContractV1.prototype.toAmino = function (_isClassic) {
        var _a = this, authority = _a.authority, contract = _a.contract, msg = _a.msg;
        return {
            type: 'wasm/MsgSudoContract',
            value: {
                authority: authority,
                contract: contract,
                msg: (0, json_1.removeNull)(msg),
            },
        };
    };
    MsgSudoContractV1.fromProto = function (proto, _isClassic) {
        return new MsgSudoContractV1(proto.authority, proto.contract, JSON.parse(Buffer.from(proto.msg).toString('utf-8')));
    };
    MsgSudoContractV1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, contract = _a.contract, msg = _a.msg;
        return tx_1.MsgSudoContract.fromPartial({
            authority: authority,
            contract: contract,
            msg: Buffer.from(JSON.stringify((0, json_1.removeNull)(msg)), 'utf-8'),
        });
    };
    MsgSudoContractV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgSudoContract',
            value: tx_1.MsgSudoContract.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgSudoContractV1.unpackAny = function (msgAny, isClassic) {
        return MsgSudoContractV1.fromProto(tx_1.MsgSudoContract.decode(msgAny.value), isClassic);
    };
    MsgSudoContractV1.fromData = function (data, _isClassic) {
        var authority = data.authority, contract = data.contract, msg = data.msg;
        return new MsgSudoContractV1(authority, contract, msg);
    };
    MsgSudoContractV1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, contract = _a.contract, msg = _a.msg;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgSudoContract',
            authority: authority,
            contract: contract,
            msg: msg,
        };
    };
    return MsgSudoContractV1;
}(json_1.JSONSerializable));
exports.MsgSudoContractV1 = MsgSudoContractV1;
