// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgUnpinCodes.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUnpinCodesV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgUnpinCodesV1 = /** @class */ (function (_super) {
    __extends(MsgUnpinCodesV1, _super);
    /**
     * @param authority is the address of the governance account
     * @param code_ids references the new WASM codes
     */
    function MsgUnpinCodesV1(authority, code_ids) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.code_ids = code_ids;
        return _this;
    }
    MsgUnpinCodesV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, code_ids = _a.code_ids;
        return new MsgUnpinCodesV1(authority, code_ids.map(function (x) { return Number.parseInt(x); }));
    };
    MsgUnpinCodesV1.prototype.toAmino = function (_isClassic) {
        var _a = this, authority = _a.authority, code_ids = _a.code_ids;
        return {
            type: 'wasm/MsgUnpinCodes',
            value: {
                authority: authority,
                code_ids: code_ids.map(function (x) { return x.toFixed(); }),
            },
        };
    };
    MsgUnpinCodesV1.fromProto = function (proto, _isClassic) {
        return new MsgUnpinCodesV1(proto.authority, proto.codeIds.map(function (x) { return x.toNumber(); }));
    };
    MsgUnpinCodesV1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, code_ids = _a.code_ids;
        return tx_1.MsgUnpinCodes.fromPartial({
            authority: authority,
            codeIds: code_ids,
        });
    };
    MsgUnpinCodesV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgUnpinCodes',
            value: tx_1.MsgUnpinCodes.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUnpinCodesV1.unpackAny = function (msgAny, isClassic) {
        return MsgUnpinCodesV1.fromProto(tx_1.MsgUnpinCodes.decode(msgAny.value), isClassic);
    };
    MsgUnpinCodesV1.fromData = function (data, _isClassic) {
        var authority = data.authority, code_ids = data.code_ids;
        return new MsgUnpinCodesV1(authority, code_ids.map(function (x) { return Number.parseInt(x); }));
    };
    MsgUnpinCodesV1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, code_ids = _a.code_ids;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgUnpinCodes',
            authority: authority,
            code_ids: code_ids.map(function (x) { return x.toFixed(); }),
        };
    };
    return MsgUnpinCodesV1;
}(json_1.JSONSerializable));
exports.MsgUnpinCodesV1 = MsgUnpinCodesV1;
