// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgUpdateParams.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateWasmParamsV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var AccessConfig_1 = require("../AccessConfig");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var types_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/types");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgUpdateWasmParamsV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateWasmParamsV1, _super);
    function MsgUpdateWasmParamsV1(authority, codeUploadAccess, instantiateDefaultPermission) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.codeUploadAccess = codeUploadAccess;
        _this.instantiateDefaultPermission = instantiateDefaultPermission;
        return _this;
    }
    MsgUpdateWasmParamsV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, codeUploadAccess = _a.codeUploadAccess, instantiateDefaultPermission = _a.instantiateDefaultPermission;
        return new MsgUpdateWasmParamsV1(authority, codeUploadAccess, instantiateDefaultPermission);
    };
    MsgUpdateWasmParamsV1.prototype.toAmino = function (_isClassic) {
        var _a = this, authority = _a.authority, codeUploadAccess = _a.codeUploadAccess, instantiateDefaultPermission = _a.instantiateDefaultPermission;
        return {
            type: 'wasm/MsgUpdateParamsV1',
            value: {
                authority: authority,
                codeUploadAccess: codeUploadAccess,
                instantiateDefaultPermission: instantiateDefaultPermission,
            },
        };
    };
    MsgUpdateWasmParamsV1.fromData = function (data, _isClassic) {
        var authority = data.authority, codeUploadAccess = data.codeUploadAccess, instantiateDefaultPermission = data.instantiateDefaultPermission;
        return new MsgUpdateWasmParamsV1(authority, codeUploadAccess, instantiateDefaultPermission);
    };
    MsgUpdateWasmParamsV1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, codeUploadAccess = _a.codeUploadAccess, instantiateDefaultPermission = _a.instantiateDefaultPermission;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgUpdateParams',
            authority: authority,
            codeUploadAccess: codeUploadAccess,
            instantiateDefaultPermission: instantiateDefaultPermission,
        };
    };
    MsgUpdateWasmParamsV1.fromProto = function (proto, _isClassic) {
        var _a, _b, _c;
        return new MsgUpdateWasmParamsV1(proto.authority, ((_a = proto.params) === null || _a === void 0 ? void 0 : _a.codeUploadAccess)
            ? AccessConfig_1.AccessConfig.fromProto(proto.params.codeUploadAccess)
            : undefined, (_c = (_b = proto.params) === null || _b === void 0 ? void 0 : _b.instantiateDefaultPermission) !== null && _c !== void 0 ? _c : AccessConfig_1.AccessType.ACCESS_TYPE_UNSPECIFIED);
    };
    MsgUpdateWasmParamsV1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, codeUploadAccess = _a.codeUploadAccess, instantiateDefaultPermission = _a.instantiateDefaultPermission;
        return tx_1.MsgUpdateParams.fromPartial({
            authority: authority,
            params: types_1.Params.fromPartial({
                codeUploadAccess: codeUploadAccess === null || codeUploadAccess === void 0 ? void 0 : codeUploadAccess.toProto(),
                instantiateDefaultPermission: instantiateDefaultPermission,
            }),
        });
    };
    MsgUpdateWasmParamsV1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    };
    MsgUpdateWasmParamsV1.unpackAny = function (msgAny, _isClassic) {
        return MsgUpdateWasmParamsV1.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    };
    return MsgUpdateWasmParamsV1;
}(json_1.JSONSerializable));
exports.MsgUpdateWasmParamsV1 = MsgUpdateWasmParamsV1;
