// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgUpdateContractLabel.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateContractLabelV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgUpdateContractLabelV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateContractLabelV1, _super);
    /**
     * @param sender is the that actor that signed the messages
     * @param new_label string to be set
     * @param contract is the address of the smart contract
     */
    function MsgUpdateContractLabelV1(sender, new_label, contract) {
        var _this = _super.call(this) || this;
        _this.sender = sender;
        _this.new_label = new_label;
        _this.contract = contract;
        return _this;
    }
    MsgUpdateContractLabelV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, sender = _a.sender, new_label = _a.new_label, contract = _a.contract;
        return new MsgUpdateContractLabelV1(sender, new_label, contract);
    };
    MsgUpdateContractLabelV1.prototype.toAmino = function (_isClassic) {
        var _a = this, sender = _a.sender, new_label = _a.new_label, contract = _a.contract;
        return {
            type: 'wasm/MsgUpdateContractLabel',
            value: {
                sender: sender,
                new_label: new_label,
                contract: contract,
            },
        };
    };
    MsgUpdateContractLabelV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateContractLabelV1(proto.sender, proto.newLabel, proto.contract);
    };
    MsgUpdateContractLabelV1.prototype.toProto = function (_isClassic) {
        var _a = this, sender = _a.sender, new_label = _a.new_label, contract = _a.contract;
        return tx_1.MsgUpdateContractLabel.fromPartial({
            sender: sender,
            newLabel: new_label,
            contract: contract,
        });
    };
    MsgUpdateContractLabelV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgUpdateContractLabel',
            value: tx_1.MsgUpdateContractLabel.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateContractLabelV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateContractLabelV1.fromProto(tx_1.MsgUpdateContractLabel.decode(msgAny.value), isClassic);
    };
    MsgUpdateContractLabelV1.fromData = function (data, _isClassic) {
        var sender = data.sender, new_label = data.new_label, contract = data.contract;
        return new MsgUpdateContractLabelV1(sender, new_label, contract);
    };
    MsgUpdateContractLabelV1.prototype.toData = function (_isClassic) {
        var _a = this, sender = _a.sender, new_label = _a.new_label, contract = _a.contract;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgUpdateContractLabel',
            sender: sender,
            new_label: new_label,
            contract: contract,
        };
    };
    return MsgUpdateContractLabelV1;
}(json_1.JSONSerializable));
exports.MsgUpdateContractLabelV1 = MsgUpdateContractLabelV1;
