// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgUpdateInstantiateConfig.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateInstantiateConfigV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var AccessConfig_1 = require("../AccessConfig");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgUpdateInstantiateConfigV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateInstantiateConfigV1, _super);
    /**
     * @param sender is the that actor that signed the messages
     * @param code_id references the stored WASM code
     * @param new_instantiate_permission is the new access control
     */
    function MsgUpdateInstantiateConfigV1(sender, code_id, new_instantiate_permission) {
        var _this = _super.call(this) || this;
        _this.sender = sender;
        _this.code_id = code_id;
        _this.new_instantiate_permission = new_instantiate_permission;
        return _this;
    }
    MsgUpdateInstantiateConfigV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, sender = _a.sender, code_id = _a.code_id, new_instantiate_permission = _a.new_instantiate_permission;
        return new MsgUpdateInstantiateConfigV1(sender, Number.parseInt(code_id), new_instantiate_permission
            ? AccessConfig_1.AccessConfig.fromAmino(new_instantiate_permission)
            : undefined);
    };
    MsgUpdateInstantiateConfigV1.prototype.toAmino = function (_isClassic) {
        var _a = this, sender = _a.sender, code_id = _a.code_id, new_instantiate_permission = _a.new_instantiate_permission;
        return {
            type: 'wasm/MsgUpdateInstantiateConfig',
            value: {
                sender: sender,
                code_id: code_id.toFixed(),
                new_instantiate_permission: new_instantiate_permission === null || new_instantiate_permission === void 0 ? void 0 : new_instantiate_permission.toAmino(),
            },
        };
    };
    MsgUpdateInstantiateConfigV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateInstantiateConfigV1(proto.sender, proto.codeId.toNumber(), proto.newInstantiatePermission
            ? AccessConfig_1.AccessConfig.fromProto(proto.newInstantiatePermission)
            : undefined);
    };
    MsgUpdateInstantiateConfigV1.prototype.toProto = function (_isClassic) {
        var _a = this, sender = _a.sender, code_id = _a.code_id, new_instantiate_permission = _a.new_instantiate_permission;
        return tx_1.MsgUpdateInstantiateConfig.fromPartial({
            sender: sender,
            codeId: code_id,
            newInstantiatePermission: new_instantiate_permission === null || new_instantiate_permission === void 0 ? void 0 : new_instantiate_permission.toProto(),
        });
    };
    MsgUpdateInstantiateConfigV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgUpdateInstantiateConfig',
            value: tx_1.MsgUpdateInstantiateConfig.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateInstantiateConfigV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateInstantiateConfigV1.fromProto(tx_1.MsgUpdateInstantiateConfig.decode(msgAny.value), isClassic);
    };
    MsgUpdateInstantiateConfigV1.fromData = function (data, _isClassic) {
        var sender = data.sender, code_id = data.code_id, new_instantiate_permission = data.new_instantiate_permission;
        return new MsgUpdateInstantiateConfigV1(sender, Number.parseInt(code_id), new_instantiate_permission
            ? AccessConfig_1.AccessConfig.fromData(new_instantiate_permission)
            : undefined);
    };
    MsgUpdateInstantiateConfigV1.prototype.toData = function (_isClassic) {
        var _a = this, sender = _a.sender, code_id = _a.code_id, new_instantiate_permission = _a.new_instantiate_permission;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgUpdateInstantiateConfig',
            sender: sender,
            code_id: code_id.toFixed(),
            new_instantiate_permission: new_instantiate_permission === null || new_instantiate_permission === void 0 ? void 0 : new_instantiate_permission.toData(),
        };
    };
    return MsgUpdateInstantiateConfigV1;
}(json_1.JSONSerializable));
exports.MsgUpdateInstantiateConfigV1 = MsgUpdateInstantiateConfigV1;
