// @xpla/xpla.js/dist/core/wasm/v1/msgs/MsgUpdateContractAdmin.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateContractAdminV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/tx");
var MsgUpdateContractAdminV1 = /** @class */ (function (_super) {
    __extends(MsgUpdateContractAdminV1, _super);
    /**
     * @param admin contract admin
     * @param new_admin new admin
     * @param contract contract address
     */
    function MsgUpdateContractAdminV1(admin, new_admin, contract) {
        var _this = _super.call(this) || this;
        _this.admin = admin;
        _this.new_admin = new_admin;
        _this.contract = contract;
        return _this;
    }
    MsgUpdateContractAdminV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, sender = _a.sender, new_admin = _a.new_admin, contract = _a.contract;
        return new MsgUpdateContractAdminV1(sender, new_admin, contract);
    };
    MsgUpdateContractAdminV1.prototype.toAmino = function (_isClassic) {
        var _a = this, admin = _a.admin, new_admin = _a.new_admin, contract = _a.contract;
        return {
            type: 'wasm/MsgUpdateAdmin',
            value: {
                sender: admin,
                new_admin: new_admin,
                contract: contract,
            },
        };
    };
    MsgUpdateContractAdminV1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateContractAdminV1(proto.sender, proto.newAdmin, proto.contract);
    };
    MsgUpdateContractAdminV1.prototype.toProto = function (_isClassic) {
        var _a = this, admin = _a.admin, new_admin = _a.new_admin, contract = _a.contract;
        return tx_1.MsgUpdateAdmin.fromPartial({
            sender: admin,
            contract: contract,
            newAdmin: new_admin,
        });
    };
    MsgUpdateContractAdminV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmwasm.wasm.v1.MsgUpdateAdmin',
            value: tx_1.MsgUpdateAdmin.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgUpdateContractAdminV1.unpackAny = function (msgAny, isClassic) {
        return MsgUpdateContractAdminV1.fromProto(tx_1.MsgUpdateAdmin.decode(msgAny.value), isClassic);
    };
    MsgUpdateContractAdminV1.fromData = function (data, _isClassic) {
        var sender = data.sender, new_admin = data.new_admin, contract = data.contract;
        return new MsgUpdateContractAdminV1(sender, new_admin, contract);
    };
    MsgUpdateContractAdminV1.prototype.toData = function (_isClassic) {
        var _a = this, admin = _a.admin, new_admin = _a.new_admin, contract = _a.contract;
        return {
            '@type': '/cosmwasm.wasm.v1.MsgUpdateAdmin',
            sender: admin,
            new_admin: new_admin,
            contract: contract,
        };
    };
    return MsgUpdateContractAdminV1;
}(json_1.JSONSerializable));
exports.MsgUpdateContractAdminV1 = MsgUpdateContractAdminV1;
