// @xpla/xpla.js/dist/core/distribution/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateDistributionParams = exports.MsgCommunityPoolSpend = exports.MsgFundCommunityPool = exports.MsgWithdrawValidatorCommission = exports.MsgWithdrawDelegatorReward = exports.MsgSetWithdrawAddress = void 0;
__exportStar(require("./v1beta1/msgs"), exports);
__exportStar(require("./v1beta1/proposals"), exports);
__exportStar(require("./v1beta1/Params"), exports);
var msgs_1 = require("./v1beta1/msgs");
Object.defineProperty(exports, "MsgSetWithdrawAddress", { enumerable: true, get: function () { return msgs_1.MsgSetWithdrawAddressV1B1; } });
Object.defineProperty(exports, "MsgWithdrawDelegatorReward", { enumerable: true, get: function () { return msgs_1.MsgWithdrawDelegatorRewardV1B1; } });
Object.defineProperty(exports, "MsgWithdrawValidatorCommission", { enumerable: true, get: function () { return msgs_1.MsgWithdrawValidatorCommissionV1B1; } });
Object.defineProperty(exports, "MsgFundCommunityPool", { enumerable: true, get: function () { return msgs_1.MsgFundCommunityPoolV1B1; } });
Object.defineProperty(exports, "MsgCommunityPoolSpend", { enumerable: true, get: function () { return msgs_1.MsgCommunityPoolSpendV1B1; } });
Object.defineProperty(exports, "MsgUpdateDistributionParams", { enumerable: true, get: function () { return msgs_1.MsgUpdateDistributionParamsV1B1; } });
