// @xpla/xpla.js/dist/core/xpla/v1beta1/Params.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardParamsV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var numeric_1 = require("../../../core/numeric");
var reward_1 = require("@xpla/xpla.proto/xpla/reward/v1beta1/reward");
var RewardParamsV1B1 = /** @class */ (function (_super) {
    __extends(RewardParamsV1B1, _super);
    /**
     * @param fee_pool_rate
     * @param community_pool_rate
     * @param reserve_rate
     * @param reserve_account
     * @param reward_distribute_account
     */
    function RewardParamsV1B1(fee_pool_rate, community_pool_rate, reserve_rate, reserve_account, reward_distribute_account) {
        var _this = _super.call(this) || this;
        _this.reserve_account = reserve_account;
        _this.reward_distribute_account = reward_distribute_account;
        _this.fee_pool_rate = new numeric_1.Dec(fee_pool_rate);
        _this.community_pool_rate = new numeric_1.Dec(community_pool_rate);
        _this.reserve_rate = new numeric_1.Dec(reserve_rate);
        return _this;
    }
    RewardParamsV1B1.fromAmino = function (data, _) {
        var fee_pool_rate = data.fee_pool_rate, community_pool_rate = data.community_pool_rate, reserve_rate = data.reserve_rate, reserve_account = data.reserve_account, reward_distribute_account = data.reward_distribute_account;
        return new RewardParamsV1B1(fee_pool_rate !== null && fee_pool_rate !== void 0 ? fee_pool_rate : 0, community_pool_rate !== null && community_pool_rate !== void 0 ? community_pool_rate : 0, reserve_rate !== null && reserve_rate !== void 0 ? reserve_rate : 0, reserve_account !== null && reserve_account !== void 0 ? reserve_account : '', reward_distribute_account !== null && reward_distribute_account !== void 0 ? reward_distribute_account : '');
    };
    RewardParamsV1B1.prototype.toAmino = function (_) {
        var _a = this, fee_pool_rate = _a.fee_pool_rate, community_pool_rate = _a.community_pool_rate, reserve_rate = _a.reserve_rate, reserve_account = _a.reserve_account, reward_distribute_account = _a.reward_distribute_account;
        var res = {
            fee_pool_rate: fee_pool_rate.toFixed(),
            community_pool_rate: community_pool_rate.toFixed(),
            reserve_rate: reserve_rate.toFixed(),
            reserve_account: reserve_account,
            reward_distribute_account: reward_distribute_account,
        };
        return res;
    };
    RewardParamsV1B1.fromData = function (data, _) {
        var fee_pool_rate = data.fee_pool_rate, community_pool_rate = data.community_pool_rate, reserve_rate = data.reserve_rate, reserve_account = data.reserve_account, reward_distribute_account = data.reward_distribute_account;
        return new RewardParamsV1B1(fee_pool_rate, community_pool_rate, reserve_rate, reserve_account, reward_distribute_account);
    };
    RewardParamsV1B1.prototype.toData = function (_) {
        var _a = this, fee_pool_rate = _a.fee_pool_rate, community_pool_rate = _a.community_pool_rate, reserve_rate = _a.reserve_rate, reserve_account = _a.reserve_account, reward_distribute_account = _a.reward_distribute_account;
        var res = {
            '@type': '/xpla.reward.v1beta1.Params',
            fee_pool_rate: fee_pool_rate.toFixed(),
            community_pool_rate: community_pool_rate.toFixed(),
            reserve_rate: reserve_rate.toFixed(),
            reserve_account: reserve_account,
            reward_distribute_account: reward_distribute_account,
        };
        return res;
    };
    RewardParamsV1B1.fromProto = function (proto, _) {
        return new RewardParamsV1B1(proto.feePoolRate, proto.communityPoolRate, proto.reserveRate, proto.reserveAccount, proto.rewardDistributeAccount);
    };
    RewardParamsV1B1.prototype.toProto = function (_) {
        var _a = this, fee_pool_rate = _a.fee_pool_rate, community_pool_rate = _a.community_pool_rate, reserve_rate = _a.reserve_rate, reserve_account = _a.reserve_account, reward_distribute_account = _a.reward_distribute_account;
        return reward_1.Params.fromPartial({
            feePoolRate: fee_pool_rate.toFixed(),
            communityPoolRate: community_pool_rate.toFixed(),
            reserveRate: reserve_rate.toFixed(),
            reserveAccount: reserve_account,
            rewardDistributeAccount: reward_distribute_account,
        });
    };
    return RewardParamsV1B1;
}(json_1.JSONSerializable));
exports.RewardParamsV1B1 = RewardParamsV1B1;
