// @xpla/xpla.js/dist/core/vesting/v1beta1/msgs/MsgCreatePermanentLockedAccount.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCreatePermanentLockedAccountV1B1 = void 0;
var json_1 = require("../../../../util/json");
var Coins_1 = require("../../../Coins");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/vesting/v1beta1/tx");
var MsgCreatePermanentLockedAccountV1B1 = /** @class */ (function (_super) {
    __extends(MsgCreatePermanentLockedAccountV1B1, _super);
    /**
     * @param from_address sender's address
     * @param to_address recipient's address
     * @param amount value of the transaction
     */
    function MsgCreatePermanentLockedAccountV1B1(from_address, to_address, amount) {
        var _this = _super.call(this) || this;
        _this.from_address = from_address;
        _this.to_address = to_address;
        _this.amount = new Coins_1.Coins(amount);
        return _this;
    }
    MsgCreatePermanentLockedAccountV1B1.fromAmino = function (data, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        var _a = data.value, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount;
        return new MsgCreatePermanentLockedAccountV1B1(from_address, to_address, Coins_1.Coins.fromAmino(amount));
    };
    MsgCreatePermanentLockedAccountV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount;
        return {
            type: isClassic
                ? 'vesting/MsgCreatePermanentLockedAccount'
                : 'cosmos-sdk/MsgCreatePermanentLockedAccount',
            value: {
                from_address: from_address,
                to_address: to_address,
                amount: amount.toAmino(),
            },
        };
    };
    MsgCreatePermanentLockedAccountV1B1.fromData = function (data, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        var from_address = data.from_address, to_address = data.to_address, amount = data.amount;
        return new MsgCreatePermanentLockedAccountV1B1(from_address, to_address, Coins_1.Coins.fromData(amount));
    };
    MsgCreatePermanentLockedAccountV1B1.prototype.toData = function (isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount;
        return {
            '@type': '/cosmos.vesting.v1beta1.MsgCreatePermanentLockedAccount',
            from_address: from_address,
            to_address: to_address,
            amount: amount.toData(),
        };
    };
    MsgCreatePermanentLockedAccountV1B1.fromProto = function (proto, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        return new MsgCreatePermanentLockedAccountV1B1(proto.fromAddress, proto.toAddress, Coins_1.Coins.fromProto(proto.amount));
    };
    MsgCreatePermanentLockedAccountV1B1.prototype.toProto = function (isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount;
        return tx_1.MsgCreatePermanentLockedAccount.fromPartial({
            fromAddress: from_address,
            toAddress: to_address,
            amount: amount.toProto(),
        });
    };
    MsgCreatePermanentLockedAccountV1B1.prototype.packAny = function (isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.vesting.v1beta1.MsgCreatePermanentLockedAccount',
            value: tx_1.MsgCreatePermanentLockedAccount.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCreatePermanentLockedAccountV1B1.unpackAny = function (msgAny, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        return MsgCreatePermanentLockedAccountV1B1.fromProto(tx_1.MsgCreatePermanentLockedAccount.decode(msgAny.value), isClassic);
    };
    return MsgCreatePermanentLockedAccountV1B1;
}(json_1.JSONSerializable));
exports.MsgCreatePermanentLockedAccountV1B1 = MsgCreatePermanentLockedAccountV1B1;
