// @xpla/xpla.js/dist/core/vesting/v1beta1/msgs/MsgCreateVestingAccount.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCreateVestingAccountV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Coins_1 = require("../../../Coins");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/vesting/v1beta1/tx");
/**
 * MsgCreateVestingAccount defines a message that enables creating a vesting account.
 */
var MsgCreateVestingAccountV1B1 = /** @class */ (function (_super) {
    __extends(MsgCreateVestingAccountV1B1, _super);
    /**
     * @param from_address sender's address
     * @param to_address recipient's address
     * @param amount value of the transaction
     */
    function MsgCreateVestingAccountV1B1(from_address, to_address, amount, end_time, delayed) {
        var _this = _super.call(this) || this;
        _this.from_address = from_address;
        _this.to_address = to_address;
        _this.end_time = end_time;
        _this.delayed = delayed;
        _this.amount = new Coins_1.Coins(amount);
        return _this;
    }
    MsgCreateVestingAccountV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount, end_time = _a.end_time, delayed = _a.delayed;
        return new MsgCreateVestingAccountV1B1(from_address, to_address, Coins_1.Coins.fromAmino(amount), Number.parseInt(end_time), delayed);
    };
    MsgCreateVestingAccountV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount, end_time = _a.end_time, delayed = _a.delayed;
        return {
            type: isClassic
                ? 'vesting/MsgCreateVestingAccount'
                : 'cosmos-sdk/MsgCreateVestingAccount',
            value: {
                from_address: from_address,
                to_address: to_address,
                amount: amount.toAmino(),
                end_time: end_time.toFixed(),
                delayed: delayed,
            },
        };
    };
    MsgCreateVestingAccountV1B1.fromData = function (data, _isClassic) {
        var from_address = data.from_address, to_address = data.to_address, amount = data.amount, end_time = data.end_time, delayed = data.delayed;
        return new MsgCreateVestingAccountV1B1(from_address, to_address, Coins_1.Coins.fromData(amount), Number.parseInt(end_time), delayed);
    };
    MsgCreateVestingAccountV1B1.prototype.toData = function (_isClassic) {
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount, end_time = _a.end_time, delayed = _a.delayed;
        return {
            '@type': '/cosmos.vesting.v1beta1.MsgCreateVestingAccount',
            from_address: from_address,
            to_address: to_address,
            amount: amount.toData(),
            end_time: end_time.toFixed(),
            delayed: delayed,
        };
    };
    MsgCreateVestingAccountV1B1.fromProto = function (proto, _isClassic) {
        return new MsgCreateVestingAccountV1B1(proto.fromAddress, proto.toAddress, Coins_1.Coins.fromProto(proto.amount), proto.endTime.toNumber(), proto.delayed);
    };
    MsgCreateVestingAccountV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, amount = _a.amount, end_time = _a.end_time, delayed = _a.delayed;
        return tx_1.MsgCreateVestingAccount.fromPartial({
            fromAddress: from_address,
            toAddress: to_address,
            amount: amount.toProto(),
            endTime: end_time,
            delayed: delayed,
        });
    };
    MsgCreateVestingAccountV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.vesting.v1beta1.MsgCreateVestingAccount',
            value: tx_1.MsgCreateVestingAccount.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCreateVestingAccountV1B1.unpackAny = function (msgAny, isClassic) {
        if (isClassic) {
            throw new Error('Not supported for the network');
        }
        return MsgCreateVestingAccountV1B1.fromProto(tx_1.MsgCreateVestingAccount.decode(msgAny.value), isClassic);
    };
    return MsgCreateVestingAccountV1B1;
}(json_1.JSONSerializable));
exports.MsgCreateVestingAccountV1B1 = MsgCreateVestingAccountV1B1;
