// @xpla/xpla.js/dist/core/vesting/v1beta1/msgs/MsgCreatePeriodicVestingAccount.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCreatePeriodicVestingAccountV1B1 = void 0;
var json_1 = require("../../../../util/json");
var Period_1 = require("../Period");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var vesting_1 = require("@xpla/xpla.proto/cosmos/vesting/v1beta1/vesting");
var tx_1 = require("@xpla/xpla.proto/cosmos/vesting/v1beta1/tx");
var MsgCreatePeriodicVestingAccountV1B1 = /** @class */ (function (_super) {
    __extends(MsgCreatePeriodicVestingAccountV1B1, _super);
    /**
     * @param from_address sender's address
     * @param to_address recipient's address
     * @param start of vesting as unix time (in seconds)
     */
    function MsgCreatePeriodicVestingAccountV1B1(from_address, to_address, start_time, vesting_periods) {
        var _this = _super.call(this) || this;
        _this.from_address = from_address;
        _this.to_address = to_address;
        _this.start_time = start_time;
        _this.vesting_periods = vesting_periods;
        return _this;
    }
    MsgCreatePeriodicVestingAccountV1B1.fromAmino = function (data, isClassic) {
        var _a = data.value, from_address = _a.from_address, to_address = _a.to_address, start_time = _a.start_time, vesting_periods = _a.vesting_periods;
        return new MsgCreatePeriodicVestingAccountV1B1(from_address, to_address, Number.parseInt(start_time), vesting_periods.map(function (x) { return Period_1.PeriodV1B1.fromAmino(x, isClassic); }));
    };
    MsgCreatePeriodicVestingAccountV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, start_time = _a.start_time, vesting_periods = _a.vesting_periods;
        return {
            type: isClassic
                ? 'vesting/MsgCreatePeriodicVestingAccount'
                : 'cosmos-sdk/MsgCreatePeriodicVestingAccount',
            value: {
                from_address: from_address,
                to_address: to_address,
                start_time: start_time.toFixed(),
                vesting_periods: vesting_periods.map(function (msg) {
                    return msg.toAmino(isClassic);
                }),
            },
        };
    };
    MsgCreatePeriodicVestingAccountV1B1.fromData = function (data, isClassic) {
        var from_address = data.from_address, to_address = data.to_address, start_time = data.start_time, vesting_periods = data.vesting_periods;
        return new MsgCreatePeriodicVestingAccountV1B1(from_address, to_address, Number.parseInt(start_time), vesting_periods.map(function (x) { return Period_1.PeriodV1B1.fromData(x, isClassic); }));
    };
    MsgCreatePeriodicVestingAccountV1B1.prototype.toData = function (isClassic) {
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, start_time = _a.start_time, vesting_periods = _a.vesting_periods;
        return {
            '@type': '/cosmos.vesting.v1beta1.MsgCreatePeriodicVestingAccount',
            from_address: from_address,
            to_address: to_address,
            start_time: start_time.toFixed(),
            vesting_periods: vesting_periods.map(function (msg) { return msg.toData(isClassic); }),
        };
    };
    MsgCreatePeriodicVestingAccountV1B1.fromProto = function (proto, isClassic) {
        return new MsgCreatePeriodicVestingAccountV1B1(proto.fromAddress, proto.toAddress, proto.startTime.toNumber(), proto.vestingPeriods.map(function (x) { return Period_1.PeriodV1B1.fromProto(x, isClassic); }));
    };
    MsgCreatePeriodicVestingAccountV1B1.prototype.toProto = function (isClassic) {
        var _a = this, from_address = _a.from_address, to_address = _a.to_address, start_time = _a.start_time, vesting_periods = _a.vesting_periods;
        return tx_1.MsgCreatePeriodicVestingAccount.fromPartial({
            fromAddress: from_address,
            toAddress: to_address,
            startTime: start_time,
            vestingPeriods: vesting_periods.map(function (x) {
                return vesting_1.Period.decode(x.packAny(isClassic).value);
            }),
        });
    };
    MsgCreatePeriodicVestingAccountV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.vesting.v1beta1.MsgCreatePeriodicVestingAccount',
            value: tx_1.MsgCreatePeriodicVestingAccount.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCreatePeriodicVestingAccountV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgCreatePeriodicVestingAccountV1B1.fromProto(tx_1.MsgCreatePeriodicVestingAccount.decode(msgAny.value), isClassic);
    };
    return MsgCreatePeriodicVestingAccountV1B1;
}(json_1.JSONSerializable));
exports.MsgCreatePeriodicVestingAccountV1B1 = MsgCreatePeriodicVestingAccountV1B1;
