// @xpla/xpla.js/dist/client/ecd/msgs/EvmTx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvmMessage = exports.EvmTxInfo = void 0;
var numeric_1 = require("../../../core/numeric");
var EvmTxInfo = /** @class */ (function () {
    function EvmTxInfo(blockHash, blockNumber, from, to, contractAddress, cumulativeGasUsed, effectiveGasPrice, gasUsed, transactionHash, transactionIndex, type, logs, logsBloom, root, status) {
        this.blockHash = blockHash;
        this.blockNumber = blockNumber;
        this.from = from;
        this.to = to;
        this.contractAddress = contractAddress;
        this.cumulativeGasUsed = cumulativeGasUsed;
        this.effectiveGasPrice = effectiveGasPrice;
        this.gasUsed = gasUsed;
        this.transactionHash = transactionHash;
        this.transactionIndex = transactionIndex;
        this.type = type;
        this.logs = logs;
        this.logsBloom = logsBloom;
        this.root = root;
        this.status = status;
    }
    EvmTxInfo.fromData = function (data) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        return new EvmTxInfo(data.blockHash, numeric_1.Numeric.parse((_a = data.blockNumber) !== null && _a !== void 0 ? _a : '0x0').toNumber(), data.from, data.to, data.contractAddress, numeric_1.Numeric.parse((_b = data.cumulativeGasUsed) !== null && _b !== void 0 ? _b : '0x0').toNumber(), numeric_1.Numeric.parse((_c = data.effectiveGasPrice) !== null && _c !== void 0 ? _c : '0x0').toNumber(), numeric_1.Numeric.parse((_d = data.gasUsed) !== null && _d !== void 0 ? _d : '0x0').toNumber(), (_e = data.transactionHash) !== null && _e !== void 0 ? _e : '', numeric_1.Numeric.parse((_f = data.transactionIndex) !== null && _f !== void 0 ? _f : '0x0').toNumber(), numeric_1.Numeric.parse((_g = data.type) !== null && _g !== void 0 ? _g : '0x0').toNumber(), (_h = data.logs) !== null && _h !== void 0 ? _h : [], (_j = data.logsBloom) !== null && _j !== void 0 ? _j : '', (_k = data.root) !== null && _k !== void 0 ? _k : '', numeric_1.Numeric.parse((_l = data.status) !== null && _l !== void 0 ? _l : '0x0').toNumber());
    };
    return EvmTxInfo;
}());
exports.EvmTxInfo = EvmTxInfo;
var EvmMessage = /** @class */ (function () {
    function EvmMessage() {
    }
    return EvmMessage;
}());
exports.EvmMessage = EvmMessage;
