// @xpla/xpla.js/dist/core/erc20/v1/proposals/RegisterERC20Proposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisterERC20Proposal = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var erc20_1 = require("@xpla/xpla.proto/ethermint/erc20/v1/erc20");
/**
 * erc20 RegisterERC20Proposal
 */
var RegisterERC20Proposal = /** @class */ (function (_super) {
    __extends(RegisterERC20Proposal, _super);
    /**
     * @param title title of the proposal
     * @param description proposal description
     * @param erc20addresses contract addresses of ERC20 token
     */
    function RegisterERC20Proposal(title, description, erc20addresses) {
        var _this = _super.call(this) || this;
        _this.title = title;
        _this.description = description;
        _this.erc20addresses = erc20addresses;
        return _this;
    }
    RegisterERC20Proposal.fromAmino = function (data, _isClassic) {
        var _a = data.value, title = _a.title, description = _a.description, erc20addresses = _a.erc20addresses;
        return new RegisterERC20Proposal(title, description, erc20addresses);
    };
    RegisterERC20Proposal.prototype.toAmino = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, erc20addresses = _a.erc20addresses;
        return {
            type: 'erc20/RegisterERC20Proposal',
            value: {
                title: title,
                description: description,
                erc20addresses: erc20addresses,
            },
        };
    };
    RegisterERC20Proposal.fromData = function (data, _isClassic) {
        var title = data.title, description = data.description, erc20addresses = data.erc20addresses;
        return new RegisterERC20Proposal(title, description, erc20addresses);
    };
    RegisterERC20Proposal.prototype.toData = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, erc20addresses = _a.erc20addresses;
        return {
            '@type': '/ethermint.erc20.v1.RegisterERC20Proposal',
            title: title,
            description: description,
            erc20addresses: erc20addresses,
        };
    };
    RegisterERC20Proposal.fromProto = function (proto, _isClassic) {
        return new RegisterERC20Proposal(proto.title, proto.description, proto.erc20addresses);
    };
    RegisterERC20Proposal.prototype.toProto = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, erc20addresses = _a.erc20addresses;
        return erc20_1.RegisterERC20Proposal.fromPartial({
            title: title,
            description: description,
            erc20addresses: erc20addresses,
        });
    };
    RegisterERC20Proposal.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/ethermint.erc20.v1.RegisterERC20Proposal',
            value: erc20_1.RegisterERC20Proposal.encode(this.toProto(isClassic)).finish(),
        });
    };
    RegisterERC20Proposal.unpackAny = function (msgAny, isClassic) {
        return RegisterERC20Proposal.fromProto(erc20_1.RegisterERC20Proposal.decode(msgAny.value), isClassic);
    };
    return RegisterERC20Proposal;
}(json_1.JSONSerializable));
exports.RegisterERC20Proposal = RegisterERC20Proposal;
