// @xpla/xpla.js/dist/core/erc20/v1/proposals/RegisterCoinProposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisterCoinProposal = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var erc20_1 = require("@xpla/xpla.proto/ethermint/erc20/v1/erc20");
/**
 * erc20 RegisterCoinProposal
 */
var RegisterCoinProposal = /** @class */ (function (_super) {
    __extends(RegisterCoinProposal, _super);
    /**
     * @param title title of the proposal
     * @param description proposal description
     * @param metadata metadata of the native Cosmos coin
     */
    function RegisterCoinProposal(title, description, metadata) {
        var _this = _super.call(this) || this;
        _this.title = title;
        _this.description = description;
        _this.metadata = metadata;
        return _this;
    }
    RegisterCoinProposal.fromAmino = function (data, _isClassic) {
        var _a = data.value, title = _a.title, description = _a.description, metadata = _a.metadata;
        return new RegisterCoinProposal(title, description, metadata);
    };
    RegisterCoinProposal.prototype.toAmino = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, metadata = _a.metadata;
        return {
            type: 'erc20/RegisterCoinProposal',
            value: {
                title: title,
                description: description,
                metadata: metadata,
            },
        };
    };
    RegisterCoinProposal.fromData = function (data, _isClassic) {
        var title = data.title, description = data.description, metadata = data.metadata;
        return new RegisterCoinProposal(title, description, metadata);
    };
    RegisterCoinProposal.prototype.toData = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, metadata = _a.metadata;
        return {
            '@type': '/ethermint.erc20.v1.RegisterCoinProposal',
            title: title,
            description: description,
            metadata: metadata,
        };
    };
    RegisterCoinProposal.fromProto = function (proto, _isClassic) {
        return new RegisterCoinProposal(proto.title, proto.description, proto.metadata);
    };
    RegisterCoinProposal.prototype.toProto = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, metadata = _a.metadata;
        return erc20_1.RegisterCoinProposal.fromPartial({
            title: title,
            description: description,
            metadata: metadata,
        });
    };
    RegisterCoinProposal.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/ethermint.erc20.v1.RegisterCoinProposal',
            value: erc20_1.RegisterCoinProposal.encode(this.toProto(isClassic)).finish(),
        });
    };
    RegisterCoinProposal.unpackAny = function (msgAny, isClassic) {
        return RegisterCoinProposal.fromProto(erc20_1.RegisterCoinProposal.decode(msgAny.value), isClassic);
    };
    return RegisterCoinProposal;
}(json_1.JSONSerializable));
exports.RegisterCoinProposal = RegisterCoinProposal;
