// @xpla/xpla.js/dist/core/upgrade/v1beta1/msgs/MsgCancelUpgrade.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCancelUpgradeV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/upgrade/v1beta1/tx");
/**
 * Submit a proposal alongside an initial deposit.
 */
var MsgCancelUpgradeV1B1 = /** @class */ (function (_super) {
    __extends(MsgCancelUpgradeV1B1, _super);
    /**
     * @param authority is the address that controls the module
     */
    function MsgCancelUpgradeV1B1(authority) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        return _this;
    }
    MsgCancelUpgradeV1B1.fromAmino = function (data, _isClassic) {
        var authority = data.value.authority;
        return new MsgCancelUpgradeV1B1(authority);
    };
    MsgCancelUpgradeV1B1.prototype.toAmino = function (isClassic) {
        var authority = this.authority;
        return {
            type: isClassic
                ? 'upgrade/MsgCancelUpgrade'
                : 'cosmos-sdk/MsgCancelUpgrade',
            value: {
                authority: authority,
            },
        };
    };
    MsgCancelUpgradeV1B1.fromData = function (data, _isClassic) {
        var authority = data.authority;
        return new MsgCancelUpgradeV1B1(authority);
    };
    MsgCancelUpgradeV1B1.prototype.toData = function (_isClassic) {
        var authority = this.authority;
        return {
            '@type': '/cosmos.upgrade.v1beta1.MsgCancelUpgrade',
            authority: authority,
        };
    };
    MsgCancelUpgradeV1B1.fromProto = function (proto, _isClassic) {
        return new MsgCancelUpgradeV1B1(proto.authority);
    };
    MsgCancelUpgradeV1B1.prototype.toProto = function (_isClassic) {
        var authority = this.authority;
        return tx_1.MsgCancelUpgrade.fromPartial({
            authority: authority,
        });
    };
    MsgCancelUpgradeV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.upgrade.v1beta1.MsgCancelUpgrade',
            value: tx_1.MsgCancelUpgrade.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCancelUpgradeV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgCancelUpgradeV1B1.fromProto(tx_1.MsgCancelUpgrade.decode(msgAny.value), isClassic);
    };
    return MsgCancelUpgradeV1B1;
}(json_1.JSONSerializable));
exports.MsgCancelUpgradeV1B1 = MsgCancelUpgradeV1B1;
