// @xpla/xpla.js/dist/core/upgrade/v1beta1/msgs/MsgSoftwareUpgrade.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSoftwareUpgradeV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Plan_1 = require("../Plan");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/upgrade/v1beta1/tx");
var MsgSoftwareUpgradeV1B1 = /** @class */ (function (_super) {
    __extends(MsgSoftwareUpgradeV1B1, _super);
    /**
     * @param authority is the address that controls the module
     * @param plan is the upgrade plan
     */
    function MsgSoftwareUpgradeV1B1(authority, plan) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.plan = plan;
        return _this;
    }
    MsgSoftwareUpgradeV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, plan = _a.plan;
        return new MsgSoftwareUpgradeV1B1(authority, plan ? Plan_1.PlanV1B1.fromAmino(plan) : undefined);
    };
    MsgSoftwareUpgradeV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, authority = _a.authority, plan = _a.plan;
        return {
            type: isClassic
                ? 'upgrade/MsgSoftwareUpgrade'
                : 'cosmos-sdk/MsgSoftwareUpgrade',
            value: {
                authority: authority,
                plan: plan ? plan.toAmino() : undefined,
            },
        };
    };
    MsgSoftwareUpgradeV1B1.fromData = function (data, _isClassic) {
        var authority = data.authority, plan = data.plan;
        return new MsgSoftwareUpgradeV1B1(authority, plan ? Plan_1.PlanV1B1.fromData(plan) : undefined);
    };
    MsgSoftwareUpgradeV1B1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, plan = _a.plan;
        return {
            '@type': '/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade',
            authority: authority,
            plan: plan ? plan.toData() : undefined,
        };
    };
    MsgSoftwareUpgradeV1B1.fromProto = function (proto, _isClassic) {
        return new MsgSoftwareUpgradeV1B1(proto.authority, proto.plan ? Plan_1.PlanV1B1.fromProto(proto.plan) : undefined);
    };
    MsgSoftwareUpgradeV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, plan = _a.plan;
        return tx_1.MsgSoftwareUpgrade.fromPartial({
            authority: authority,
            plan: plan,
        });
    };
    MsgSoftwareUpgradeV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade',
            value: tx_1.MsgSoftwareUpgrade.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgSoftwareUpgradeV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgSoftwareUpgradeV1B1.fromProto(tx_1.MsgSoftwareUpgrade.decode(msgAny.value), isClassic);
    };
    return MsgSoftwareUpgradeV1B1;
}(json_1.JSONSerializable));
exports.MsgSoftwareUpgradeV1B1 = MsgSoftwareUpgradeV1B1;
