// @xpla/xpla.js/dist/core/gov/v1/msgs/MsgExecLegacyContent.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgExecLegacyContentV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/gov/v1/tx");
var Proposal_1 = require("../../v1beta1/Proposal");
/**
 * MsgExecLegacyContent is used to wrap the legacy content field into a message.
 * This ensures backwards compatibility with v1beta1.MsgSubmitProposal.
 */
var MsgExecLegacyContentV1 = /** @class */ (function (_super) {
    __extends(MsgExecLegacyContentV1, _super);
    /**
     * @param proposal_id ID of proposal to vote on
     * @param voter voter's account address
     * @param option one of voting options
     */
    function MsgExecLegacyContentV1(authority, content) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.content = content;
        return _this;
    }
    MsgExecLegacyContentV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, content = _a.content, authority = _a.authority;
        return new MsgExecLegacyContentV1(authority, content ? Proposal_1.ProposalV1B1.Content.fromAmino(content) : undefined);
    };
    MsgExecLegacyContentV1.prototype.toAmino = function (_isClassic) {
        var _a = this, content = _a.content, authority = _a.authority;
        return {
            type: 'cosmos-sdk/v1/MsgExecLegacyContent',
            value: {
                content: content === null || content === void 0 ? void 0 : content.toAmino(),
                authority: authority,
            },
        };
    };
    MsgExecLegacyContentV1.fromData = function (data, _isClassic) {
        var content = data.content, authority = data.authority;
        return new MsgExecLegacyContentV1(authority, content ? Proposal_1.ProposalV1B1.Content.fromData(content) : undefined);
    };
    MsgExecLegacyContentV1.prototype.toData = function (_isClassic) {
        var _a = this, content = _a.content, authority = _a.authority;
        return {
            '@type': '/cosmos.gov.v1.MsgExecLegacyContent',
            content: content === null || content === void 0 ? void 0 : content.toData(),
            authority: authority,
        };
    };
    MsgExecLegacyContentV1.fromProto = function (proto, _isClassic) {
        return new MsgExecLegacyContentV1(proto.authority, proto.content ? Proposal_1.ProposalV1B1.Content.fromProto(proto.content) : undefined);
    };
    MsgExecLegacyContentV1.prototype.toProto = function (_isClassic) {
        var _a = this, content = _a.content, authority = _a.authority;
        return tx_1.MsgExecLegacyContent.fromPartial({
            content: content === null || content === void 0 ? void 0 : content.packAny(),
            authority: authority,
        });
    };
    MsgExecLegacyContentV1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1.MsgExecLegacyContent',
            value: tx_1.MsgExecLegacyContent.encode(this.toProto()).finish(),
        });
    };
    MsgExecLegacyContentV1.unpackAny = function (msgAny, _isClassic) {
        return MsgExecLegacyContentV1.fromProto(tx_1.MsgExecLegacyContent.decode(msgAny.value));
    };
    return MsgExecLegacyContentV1;
}(json_1.JSONSerializable));
exports.MsgExecLegacyContentV1 = MsgExecLegacyContentV1;
