// @xpla/xpla.js/dist/core/gov/v1/msgs/MsgSubmitProposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgSubmitProposalV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Coins_1 = require("../../../Coins");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/gov/v1/tx");
var Msg_1 = require("../../../Msg");
/**
 * Submit a proposal alongside an initial deposit.
 */
var MsgSubmitProposalV1 = /** @class */ (function (_super) {
    __extends(MsgSubmitProposalV1, _super);
    /**
     * @param messages are the arbitrary messages to be executed if proposal passes
     * @param initial_deposit is the deposit value that must be paid at proposal submission
     * @param proposer is the account address of the proposer
     * @param metadata is any arbitrary metadata attached to the proposal
     * @param title is the title of the proposal
     * @param summary is the summary of the proposal
     */
    function MsgSubmitProposalV1(messages, initial_deposit, proposer, metadata, title, summary) {
        var _this = _super.call(this) || this;
        _this.messages = messages;
        _this.proposer = proposer;
        _this.metadata = metadata;
        _this.title = title;
        _this.summary = summary;
        _this.initial_deposit = new Coins_1.Coins(initial_deposit);
        return _this;
    }
    MsgSubmitProposalV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, messages = _a.messages, initial_deposit = _a.initial_deposit, proposer = _a.proposer, metadata = _a.metadata, title = _a.title, summary = _a.summary;
        return new MsgSubmitProposalV1(messages.map(function (a) { return Msg_1.Msg.fromAmino(a); }), Coins_1.Coins.fromAmino(initial_deposit), proposer, metadata !== null && metadata !== void 0 ? metadata : '', title, summary);
    };
    MsgSubmitProposalV1.prototype.toAmino = function (_isClassic) {
        var _a = this, messages = _a.messages, initial_deposit = _a.initial_deposit, proposer = _a.proposer, metadata = _a.metadata, title = _a.title, summary = _a.summary;
        return {
            type: 'cosmos-sdk/v1/MsgSubmitProposal',
            value: {
                messages: messages.map(function (a) { return a.toAmino(); }),
                initial_deposit: initial_deposit.toAmino(),
                proposer: proposer,
                metadata: metadata ? metadata : undefined,
                title: title,
                summary: summary,
            },
        };
    };
    MsgSubmitProposalV1.fromData = function (data, _isClassic) {
        var messages = data.messages, initial_deposit = data.initial_deposit, proposer = data.proposer, metadata = data.metadata, title = data.title, summary = data.summary;
        return new MsgSubmitProposalV1(messages.map(function (a) { return Msg_1.Msg.fromData(a); }), Coins_1.Coins.fromData(initial_deposit), proposer, metadata !== null && metadata !== void 0 ? metadata : '', title, summary);
    };
    MsgSubmitProposalV1.prototype.toData = function (_isClassic) {
        var _a = this, messages = _a.messages, initial_deposit = _a.initial_deposit, proposer = _a.proposer, metadata = _a.metadata, title = _a.title, summary = _a.summary;
        return {
            '@type': '/cosmos.gov.v1.MsgSubmitProposal',
            messages: messages.map(function (a) { return a.toData(); }),
            initial_deposit: initial_deposit.toData(),
            proposer: proposer,
            metadata: metadata ? metadata : undefined,
            title: title,
            summary: summary,
        };
    };
    MsgSubmitProposalV1.fromProto = function (proto, _isClassic) {
        return new MsgSubmitProposalV1(proto.messages.map(function (m) { return Msg_1.Msg.fromProto(m); }), Coins_1.Coins.fromProto(proto.initialDeposit), proto.proposer, proto.metadata, proto.title, proto.summary);
    };
    MsgSubmitProposalV1.prototype.toProto = function (_isClassic) {
        var _a = this, messages = _a.messages, initial_deposit = _a.initial_deposit, proposer = _a.proposer, metadata = _a.metadata, title = _a.title, summary = _a.summary;
        return tx_1.MsgSubmitProposal.fromPartial({
            messages: messages.map(function (m) { return m.packAny(); }),
            initialDeposit: initial_deposit.toProto(),
            proposer: proposer,
            metadata: metadata,
            title: title,
            summary: summary,
        });
    };
    MsgSubmitProposalV1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1.MsgSubmitProposal',
            value: tx_1.MsgSubmitProposal.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgSubmitProposalV1.unpackAny = function (msgAny, isClassic) {
        return MsgSubmitProposalV1.fromProto(tx_1.MsgSubmitProposal.decode(msgAny.value), isClassic);
    };
    return MsgSubmitProposalV1;
}(json_1.JSONSerializable));
exports.MsgSubmitProposalV1 = MsgSubmitProposalV1;
