// @xpla/xpla.js/dist/core/gov/v1/msgs/MsgVote.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgVoteV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/gov/v1/tx");
/**
 * Vote for a proposal
 */
var MsgVoteV1 = /** @class */ (function (_super) {
    __extends(MsgVoteV1, _super);
    /**
     * @param proposal_id defines the unique id of the proposal
     * @param voter is the voter address for the proposal
     * @param option defines the vote option
     * @param metadata is any arbitrary metadata attached to the Vote
     */
    function MsgVoteV1(proposal_id, voter, option, metadata) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.voter = voter;
        _this.option = option;
        _this.metadata = metadata;
        return _this;
    }
    MsgVoteV1.fromAmino = function (data, _isClassic) {
        var _a = data.value, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata;
        return new MsgVoteV1(Number.parseInt(proposal_id), voter, option, metadata);
    };
    MsgVoteV1.prototype.toAmino = function (isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata;
        return {
            type: isClassic ? 'gov/MsgVote' : 'cosmos-sdk/MsgVote',
            value: {
                proposal_id: proposal_id.toFixed(),
                voter: voter,
                option: option,
                metadata: metadata,
            },
        };
    };
    MsgVoteV1.fromData = function (data, _isClassic) {
        var proposal_id = data.proposal_id, voter = data.voter, option = data.option, metadata = data.metadata;
        return new MsgVoteV1(Number.parseInt(proposal_id), voter, option, metadata);
    };
    MsgVoteV1.prototype.toData = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata;
        return {
            '@type': '/cosmos.gov.v1.MsgVote',
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            option: option,
            metadata: metadata,
        };
    };
    MsgVoteV1.fromProto = function (proto, _isClassic) {
        return new MsgVoteV1(proto.proposalId.toNumber(), proto.voter, proto.option, proto.metadata);
    };
    MsgVoteV1.prototype.toProto = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option, metadata = _a.metadata;
        return tx_1.MsgVote.fromPartial({
            proposalId: proposal_id,
            voter: voter,
            option: option,
            metadata: metadata,
        });
    };
    MsgVoteV1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1.MsgVote',
            value: tx_1.MsgVote.encode(this.toProto()).finish(),
        });
    };
    MsgVoteV1.unpackAny = function (msgAny, _isClassic) {
        return MsgVoteV1.fromProto(tx_1.MsgVote.decode(msgAny.value));
    };
    return MsgVoteV1;
}(json_1.JSONSerializable));
exports.MsgVoteV1 = MsgVoteV1;
