// @xpla/xpla.js/dist/core/distribution/v1beta1/Params.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DistributionParamsV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var numeric_1 = require("../../numeric");
var distribution_1 = require("@xpla/xpla.proto/cosmos/distribution/v1beta1/distribution");
var DistributionParamsV1B1 = /** @class */ (function (_super) {
    __extends(DistributionParamsV1B1, _super);
    /**
     * @param community_tax
     * @param base_proposer_reward Deprecated: This field is deprecated and is no longer used in the x/distribution module's reward mechanism
     * @param bonus_proposer_reward Deprecated: This field is deprecated and is no longer used in the x/distribution module's reward mechanism
     * @param withdraw_addr_enabled
     */
    function DistributionParamsV1B1(community_tax, 
    /** @deprecated */ base_proposer_reward, 
    /** @deprecated */ bonus_proposer_reward, withdraw_addr_enabled) {
        var _this = _super.call(this) || this;
        _this.withdraw_addr_enabled = withdraw_addr_enabled;
        _this.community_tax = new numeric_1.Dec(community_tax);
        _this.base_proposer_reward = new numeric_1.Dec(base_proposer_reward);
        _this.bonus_proposer_reward = new numeric_1.Dec(bonus_proposer_reward);
        return _this;
    }
    DistributionParamsV1B1.fromAmino = function (data, _) {
        var community_tax = data.community_tax, base_proposer_reward = data.base_proposer_reward, bonus_proposer_reward = data.bonus_proposer_reward, withdraw_addr_enabled = data.withdraw_addr_enabled;
        return new DistributionParamsV1B1(community_tax !== null && community_tax !== void 0 ? community_tax : 0, base_proposer_reward !== null && base_proposer_reward !== void 0 ? base_proposer_reward : 0, bonus_proposer_reward !== null && bonus_proposer_reward !== void 0 ? bonus_proposer_reward : 0, withdraw_addr_enabled !== null && withdraw_addr_enabled !== void 0 ? withdraw_addr_enabled : false);
    };
    DistributionParamsV1B1.prototype.toAmino = function (_) {
        var _a = this, community_tax = _a.community_tax, base_proposer_reward = _a.base_proposer_reward, bonus_proposer_reward = _a.bonus_proposer_reward, withdraw_addr_enabled = _a.withdraw_addr_enabled;
        var res = {
            community_tax: community_tax.toFixed(),
            base_proposer_reward: base_proposer_reward.toFixed(),
            bonus_proposer_reward: bonus_proposer_reward.toFixed(),
            withdraw_addr_enabled: withdraw_addr_enabled,
        };
        return res;
    };
    DistributionParamsV1B1.fromData = function (data, _) {
        var community_tax = data.community_tax, base_proposer_reward = data.base_proposer_reward, bonus_proposer_reward = data.bonus_proposer_reward, withdraw_addr_enabled = data.withdraw_addr_enabled;
        return new DistributionParamsV1B1(community_tax, base_proposer_reward, bonus_proposer_reward, withdraw_addr_enabled);
    };
    DistributionParamsV1B1.prototype.toData = function (_) {
        var _a = this, community_tax = _a.community_tax, base_proposer_reward = _a.base_proposer_reward, bonus_proposer_reward = _a.bonus_proposer_reward, withdraw_addr_enabled = _a.withdraw_addr_enabled;
        var res = {
            '@type': '/cosmos.distribution.v1beta1.Params',
            community_tax: community_tax.toFixed(),
            base_proposer_reward: base_proposer_reward.toFixed(),
            bonus_proposer_reward: bonus_proposer_reward.toFixed(),
            withdraw_addr_enabled: withdraw_addr_enabled,
        };
        return res;
    };
    DistributionParamsV1B1.fromProto = function (proto, _) {
        return new DistributionParamsV1B1(proto.communityTax, proto.baseProposerReward, proto.bonusProposerReward, proto.withdrawAddrEnabled);
    };
    DistributionParamsV1B1.prototype.toProto = function (_) {
        var _a = this, community_tax = _a.community_tax, base_proposer_reward = _a.base_proposer_reward, bonus_proposer_reward = _a.bonus_proposer_reward, withdraw_addr_enabled = _a.withdraw_addr_enabled;
        return distribution_1.Params.fromPartial({
            communityTax: community_tax.toFixed(),
            baseProposerReward: base_proposer_reward.toFixed(),
            bonusProposerReward: bonus_proposer_reward.toFixed(),
            withdrawAddrEnabled: withdraw_addr_enabled,
        });
    };
    return DistributionParamsV1B1;
}(json_1.JSONSerializable));
exports.DistributionParamsV1B1 = DistributionParamsV1B1;
