// @xpla/xpla.js/dist/core/wasm/v1/Params.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodesParamsV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var AccessConfig_1 = require("./AccessConfig");
var types_1 = require("@xpla/xpla.proto/cosmwasm/wasm/v1/types");
var CodesParamsV1 = /** @class */ (function (_super) {
    __extends(CodesParamsV1, _super);
    /**
     * @param mint_denom type of coin to mint
     * @param inflation_rate_change maximum annual change in inflation rate
     * @param inflation_max maximum inflation rate
     * @param inflation_min minimum inflation rate
     * @param goal_bonded goal of percent bonded atoms
     * @param blocks_per_year expected blocks per year
     */
    function CodesParamsV1(code_upload_access, instantiate_default_permission) {
        var _this = _super.call(this) || this;
        _this.code_upload_access = code_upload_access;
        _this.instantiate_default_permission = instantiate_default_permission;
        return _this;
    }
    CodesParamsV1.fromAmino = function (data, _) {
        var code_upload_access = data.code_upload_access, instantiate_default_permission = data.instantiate_default_permission;
        return new CodesParamsV1(code_upload_access
            ? AccessConfig_1.AccessConfig.fromAmino(code_upload_access)
            : undefined, (0, types_1.accessTypeFromJSON)(instantiate_default_permission));
    };
    CodesParamsV1.prototype.toAmino = function (_) {
        var _a = this, code_upload_access = _a.code_upload_access, instantiate_default_permission = _a.instantiate_default_permission;
        var res = {
            code_upload_access: code_upload_access
                ? code_upload_access.toAmino()
                : undefined,
            instantiate_default_permission: (0, types_1.accessTypeToJSON)(instantiate_default_permission),
        };
        return res;
    };
    CodesParamsV1.fromData = function (data, _) {
        var code_upload_access = data.code_upload_access, instantiate_default_permission = data.instantiate_default_permission;
        return new CodesParamsV1(code_upload_access
            ? AccessConfig_1.AccessConfig.fromData(code_upload_access)
            : undefined, (0, types_1.accessTypeFromJSON)(instantiate_default_permission));
    };
    CodesParamsV1.prototype.toData = function (_) {
        var _a = this, code_upload_access = _a.code_upload_access, instantiate_default_permission = _a.instantiate_default_permission;
        var res = {
            '@type': '/cosmwasm.wasm.v1.Params',
            code_upload_access: code_upload_access
                ? code_upload_access.toData()
                : undefined,
            instantiate_default_permission: (0, types_1.accessTypeToJSON)(instantiate_default_permission),
        };
        return res;
    };
    CodesParamsV1.fromProto = function (proto, _) {
        return new CodesParamsV1(proto.codeUploadAccess
            ? AccessConfig_1.AccessConfig.fromProto(proto.codeUploadAccess)
            : undefined, proto.instantiateDefaultPermission);
    };
    CodesParamsV1.prototype.toProto = function (_) {
        var _a = this, code_upload_access = _a.code_upload_access, instantiate_default_permission = _a.instantiate_default_permission;
        return types_1.Params.fromPartial({
            codeUploadAccess: code_upload_access
                ? code_upload_access.toProto()
                : undefined,
            instantiateDefaultPermission: instantiate_default_permission,
        });
    };
    return CodesParamsV1;
}(json_1.JSONSerializable));
exports.CodesParamsV1 = CodesParamsV1;
