// @xpla/xpla.js/dist/core/distribution/v1beta1/msgs/MsgCommunityPoolSpend.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCommunityPoolSpendV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Coins_1 = require("../../../Coins");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/distribution/v1beta1/tx");
var MsgCommunityPoolSpendV1B1 = /** @class */ (function (_super) {
    __extends(MsgCommunityPoolSpendV1B1, _super);
    /**
     * @param validator_address validator's operator address
     */
    function MsgCommunityPoolSpendV1B1(authority, recipient, amount) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.recipient = recipient;
        _this.amount = new Coins_1.Coins(amount);
        return _this;
    }
    MsgCommunityPoolSpendV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, authority = _a.authority, recipient = _a.recipient, amount = _a.amount;
        return new MsgCommunityPoolSpendV1B1(authority, recipient, Coins_1.Coins.fromAmino(amount));
    };
    MsgCommunityPoolSpendV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, authority = _a.authority, recipient = _a.recipient, amount = _a.amount;
        return {
            type: isClassic
                ? 'distribution/MsgCommunityPoolSpend'
                : 'cosmos-sdk/distr/MsgCommunityPoolSpend',
            value: {
                authority: authority,
                recipient: recipient,
                amount: amount.toAmino(),
            },
        };
    };
    MsgCommunityPoolSpendV1B1.fromData = function (proto, _isClassic) {
        var authority = proto.authority, recipient = proto.recipient, amount = proto.amount;
        return new MsgCommunityPoolSpendV1B1(authority, recipient, Coins_1.Coins.fromAmino(amount));
    };
    MsgCommunityPoolSpendV1B1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, recipient = _a.recipient, amount = _a.amount;
        return {
            '@type': '/cosmos.distribution.v1beta1.MsgCommunityPoolSpend',
            authority: authority,
            recipient: recipient,
            amount: amount.toData(),
        };
    };
    MsgCommunityPoolSpendV1B1.fromProto = function (proto, _isClassic) {
        return new MsgCommunityPoolSpendV1B1(proto.authority, proto.recipient, Coins_1.Coins.fromProto(proto.amount));
    };
    MsgCommunityPoolSpendV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, recipient = _a.recipient, amount = _a.amount;
        return tx_1.MsgCommunityPoolSpend.fromPartial({
            authority: authority,
            recipient: recipient,
            amount: amount.toProto(),
        });
    };
    MsgCommunityPoolSpendV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.distribution.v1beta1.MsgCommunityPoolSpend',
            value: tx_1.MsgCommunityPoolSpend.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCommunityPoolSpendV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgCommunityPoolSpendV1B1.fromProto(tx_1.MsgCommunityPoolSpend.decode(msgAny.value), isClassic);
    };
    return MsgCommunityPoolSpendV1B1;
}(json_1.JSONSerializable));
exports.MsgCommunityPoolSpendV1B1 = MsgCommunityPoolSpendV1B1;
