// @xpla/xpla.js/dist/core/distribution/v1beta1/msgs/MsgFundCommunityPool.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgFundCommunityPoolV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Coins_1 = require("../../../Coins");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/distribution/v1beta1/tx");
var MsgFundCommunityPoolV1B1 = /** @class */ (function (_super) {
    __extends(MsgFundCommunityPoolV1B1, _super);
    /**
     * @param depositor depositor's account address
     * @param amount coins to fund the community pool
     */
    function MsgFundCommunityPoolV1B1(depositor, amount) {
        var _this = _super.call(this) || this;
        _this.depositor = depositor;
        _this.amount = new Coins_1.Coins(amount);
        return _this;
    }
    MsgFundCommunityPoolV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, depositor = _a.depositor, amount = _a.amount;
        return new MsgFundCommunityPoolV1B1(depositor, Coins_1.Coins.fromAmino(amount));
    };
    MsgFundCommunityPoolV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, depositor = _a.depositor, amount = _a.amount;
        return {
            type: isClassic
                ? 'distribution/MsgFundCommunityPool'
                : 'cosmos-sdk/MsgFundCommunityPool',
            value: {
                depositor: depositor,
                amount: amount.toAmino(),
            },
        };
    };
    MsgFundCommunityPoolV1B1.fromData = function (proto, _isClassic) {
        var depositor = proto.depositor, amount = proto.amount;
        return new MsgFundCommunityPoolV1B1(depositor, Coins_1.Coins.fromData(amount));
    };
    MsgFundCommunityPoolV1B1.prototype.toData = function (_isClassic) {
        var _a = this, depositor = _a.depositor, amount = _a.amount;
        return {
            '@type': '/cosmos.distribution.v1beta1.MsgFundCommunityPool',
            depositor: depositor,
            amount: amount.toData(),
        };
    };
    MsgFundCommunityPoolV1B1.fromProto = function (proto, _isClassic) {
        return new MsgFundCommunityPoolV1B1(proto.depositor, Coins_1.Coins.fromProto(proto.amount));
    };
    MsgFundCommunityPoolV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, depositor = _a.depositor, amount = _a.amount;
        return tx_1.MsgFundCommunityPool.fromPartial({
            amount: amount.toProto(),
            depositor: depositor,
        });
    };
    MsgFundCommunityPoolV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.distribution.v1beta1.MsgFundCommunityPool',
            value: tx_1.MsgFundCommunityPool.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgFundCommunityPoolV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgFundCommunityPoolV1B1.fromProto(tx_1.MsgFundCommunityPool.decode(msgAny.value), isClassic);
    };
    return MsgFundCommunityPoolV1B1;
}(json_1.JSONSerializable));
exports.MsgFundCommunityPoolV1B1 = MsgFundCommunityPoolV1B1;
