// @xpla/xpla.js/dist/core/distribution/v1beta1/msgs/MsgWithdrawValidatorCommission.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgWithdrawValidatorCommissionV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/distribution/v1beta1/tx");
/**
 * A validator can withdraw their outstanding commission rewards accrued from all
 * delegations (not including its self-delegation) into their associated account's
 * withdraw address.
 */
var MsgWithdrawValidatorCommissionV1B1 = /** @class */ (function (_super) {
    __extends(MsgWithdrawValidatorCommissionV1B1, _super);
    /**
     * @param validator_address validator's operator address
     */
    function MsgWithdrawValidatorCommissionV1B1(validator_address) {
        var _this = _super.call(this) || this;
        _this.validator_address = validator_address;
        return _this;
    }
    MsgWithdrawValidatorCommissionV1B1.fromAmino = function (data, _isClassic) {
        var validator_address = data.value.validator_address;
        return new MsgWithdrawValidatorCommissionV1B1(validator_address);
    };
    MsgWithdrawValidatorCommissionV1B1.prototype.toAmino = function (isClassic) {
        var validator_address = this.validator_address;
        return {
            type: isClassic
                ? 'distribution/MsgWithdrawValidatorCommission'
                : 'cosmos-sdk/MsgWithdrawValCommission',
            value: {
                validator_address: validator_address,
            },
        };
    };
    MsgWithdrawValidatorCommissionV1B1.fromData = function (proto, _isClassic) {
        var validator_address = proto.validator_address;
        return new MsgWithdrawValidatorCommissionV1B1(validator_address);
    };
    MsgWithdrawValidatorCommissionV1B1.prototype.toData = function (_isClassic) {
        var validator_address = this.validator_address;
        return {
            '@type': '/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission',
            validator_address: validator_address,
        };
    };
    MsgWithdrawValidatorCommissionV1B1.fromProto = function (proto, _isClassic) {
        return new MsgWithdrawValidatorCommissionV1B1(proto.validatorAddress);
    };
    MsgWithdrawValidatorCommissionV1B1.prototype.toProto = function (_isClassic) {
        var validator_address = this.validator_address;
        return tx_1.MsgWithdrawValidatorCommission.fromPartial({
            validatorAddress: validator_address,
        });
    };
    MsgWithdrawValidatorCommissionV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission',
            value: tx_1.MsgWithdrawValidatorCommission.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgWithdrawValidatorCommissionV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgWithdrawValidatorCommissionV1B1.fromProto(tx_1.MsgWithdrawValidatorCommission.decode(msgAny.value), isClassic);
    };
    return MsgWithdrawValidatorCommissionV1B1;
}(json_1.JSONSerializable));
exports.MsgWithdrawValidatorCommissionV1B1 = MsgWithdrawValidatorCommissionV1B1;
