// @xpla/xpla.js/dist/core/wasm/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateInstantiateConfig = exports.MsgUpdateWasmParams = exports.MsgUpdateContractLabel = exports.MsgStoreAndMigrateContract = exports.MsgRemoveCodeUploadParamsAddresses = exports.MsgAddCodeUploadParamsAddresses = exports.MsgStoreAndInstantiateContract = exports.MsgUnpinCodes = exports.MsgPinCodes = exports.MsgSudoContract = exports.MsgClearContractAdmin = exports.MsgUpdateContractAdmin = exports.MsgMigrateContract = exports.MsgExecuteContract = exports.MsgInstantiateContract2 = exports.MsgInstantiateContract = exports.MsgStoreCode = void 0;
__exportStar(require("./v1/msgs"), exports);
__exportStar(require("./v1/proposals"), exports);
__exportStar(require("./v1/AbsoluteTxPosition"), exports);
__exportStar(require("./v1/AccessConfig"), exports);
__exportStar(require("./v1/AccessConfigUpdate"), exports);
__exportStar(require("./v1/AccessTypeParam"), exports);
__exportStar(require("./v1/HistoryEntry"), exports);
__exportStar(require("./v1/Params"), exports);
__exportStar(require("./util"), exports);
var msgs_1 = require("./v1/msgs");
Object.defineProperty(exports, "MsgStoreCode", { enumerable: true, get: function () { return msgs_1.MsgStoreCodeV1; } });
Object.defineProperty(exports, "MsgInstantiateContract", { enumerable: true, get: function () { return msgs_1.MsgInstantiateContractV1; } });
Object.defineProperty(exports, "MsgInstantiateContract2", { enumerable: true, get: function () { return msgs_1.MsgInstantiateContract2V1; } });
Object.defineProperty(exports, "MsgExecuteContract", { enumerable: true, get: function () { return msgs_1.MsgExecuteContractV1; } });
Object.defineProperty(exports, "MsgMigrateContract", { enumerable: true, get: function () { return msgs_1.MsgMigrateContractV1; } });
Object.defineProperty(exports, "MsgUpdateContractAdmin", { enumerable: true, get: function () { return msgs_1.MsgUpdateContractAdminV1; } });
Object.defineProperty(exports, "MsgClearContractAdmin", { enumerable: true, get: function () { return msgs_1.MsgClearContractAdminV1; } });
Object.defineProperty(exports, "MsgSudoContract", { enumerable: true, get: function () { return msgs_1.MsgSudoContractV1; } });
Object.defineProperty(exports, "MsgPinCodes", { enumerable: true, get: function () { return msgs_1.MsgPinCodesV1; } });
Object.defineProperty(exports, "MsgUnpinCodes", { enumerable: true, get: function () { return msgs_1.MsgUnpinCodesV1; } });
Object.defineProperty(exports, "MsgStoreAndInstantiateContract", { enumerable: true, get: function () { return msgs_1.MsgStoreAndInstantiateContractV1; } });
Object.defineProperty(exports, "MsgAddCodeUploadParamsAddresses", { enumerable: true, get: function () { return msgs_1.MsgAddCodeUploadParamsAddressesV1; } });
Object.defineProperty(exports, "MsgRemoveCodeUploadParamsAddresses", { enumerable: true, get: function () { return msgs_1.MsgRemoveCodeUploadParamsAddressesV1; } });
Object.defineProperty(exports, "MsgStoreAndMigrateContract", { enumerable: true, get: function () { return msgs_1.MsgStoreAndMigrateContractV1; } });
Object.defineProperty(exports, "MsgUpdateContractLabel", { enumerable: true, get: function () { return msgs_1.MsgUpdateContractLabelV1; } });
Object.defineProperty(exports, "MsgUpdateWasmParams", { enumerable: true, get: function () { return msgs_1.MsgUpdateWasmParamsV1; } });
Object.defineProperty(exports, "MsgUpdateInstantiateConfig", { enumerable: true, get: function () { return msgs_1.MsgUpdateInstantiateConfigV1; } });
