// @xpla/xpla.js/dist/core/gov/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeightedVoteOption = exports.Vote = exports.Proposal = exports.MsgVoteWeighted = exports.MsgVote = exports.MsgSubmitProposal = exports.MsgDeposit = void 0;
__exportStar(require("./v1/msgs"), exports);
__exportStar(require("./v1/Proposal"), exports);
__exportStar(require("./v1/Vote"), exports);
__exportStar(require("./v1/Params"), exports);
__exportStar(require("./v1beta1/msgs"), exports);
__exportStar(require("./v1beta1/proposals"), exports);
__exportStar(require("./v1beta1/Proposal"), exports);
__exportStar(require("./v1beta1/Vote"), exports);
var msgs_1 = require("./v1beta1/msgs");
Object.defineProperty(exports, "MsgDeposit", { enumerable: true, get: function () { return msgs_1.MsgDepositV1B1; } });
Object.defineProperty(exports, "MsgSubmitProposal", { enumerable: true, get: function () { return msgs_1.MsgSubmitProposalV1B1; } });
Object.defineProperty(exports, "MsgVote", { enumerable: true, get: function () { return msgs_1.MsgVoteV1B1; } });
Object.defineProperty(exports, "MsgVoteWeighted", { enumerable: true, get: function () { return msgs_1.MsgVoteWeightedV1B1; } });
var Proposal_1 = require("./v1beta1/Proposal");
Object.defineProperty(exports, "Proposal", { enumerable: true, get: function () { return Proposal_1.ProposalV1B1; } });
var Vote_1 = require("./v1beta1/Vote");
Object.defineProperty(exports, "Vote", { enumerable: true, get: function () { return Vote_1.VoteV1B1; } });
Object.defineProperty(exports, "WeightedVoteOption", { enumerable: true, get: function () { return Vote_1.WeightedVoteOptionV1B1; } });
