// @xpla/xpla.js/dist/core/staking/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validator = exports.UnbondingDelegation = exports.Redelegation = exports.Delegation = exports.MsgUpdateStakingParams = exports.MsgCancelUnbondingDelegation = exports.MsgEditValidator = exports.MsgCreateValidator = exports.MsgBeginRedelegate = exports.MsgUndelegate = exports.MsgDelegate = void 0;
__exportStar(require("./v1beta1/msgs"), exports);
__exportStar(require("./v1beta1/Delegation"), exports);
__exportStar(require("./v1beta1/Redelegation"), exports);
__exportStar(require("./v1beta1/UnbondingDelegation"), exports);
__exportStar(require("./v1beta1/Validator"), exports);
__exportStar(require("./v1beta1/Params"), exports);
var msgs_1 = require("./v1beta1/msgs");
Object.defineProperty(exports, "MsgDelegate", { enumerable: true, get: function () { return msgs_1.MsgDelegateV1B1; } });
Object.defineProperty(exports, "MsgUndelegate", { enumerable: true, get: function () { return msgs_1.MsgUndelegateV1B1; } });
Object.defineProperty(exports, "MsgBeginRedelegate", { enumerable: true, get: function () { return msgs_1.MsgBeginRedelegateV1B1; } });
Object.defineProperty(exports, "MsgCreateValidator", { enumerable: true, get: function () { return msgs_1.MsgCreateValidatorV1B1; } });
Object.defineProperty(exports, "MsgEditValidator", { enumerable: true, get: function () { return msgs_1.MsgEditValidatorV1B1; } });
Object.defineProperty(exports, "MsgCancelUnbondingDelegation", { enumerable: true, get: function () { return msgs_1.MsgCancelUnbondingDelegationV1B1; } });
Object.defineProperty(exports, "MsgUpdateStakingParams", { enumerable: true, get: function () { return msgs_1.MsgUpdateStakingParamsV1B1; } });
var Delegation_1 = require("./v1beta1/Delegation");
Object.defineProperty(exports, "Delegation", { enumerable: true, get: function () { return Delegation_1.DelegationV1B1; } });
var Redelegation_1 = require("./v1beta1/Redelegation");
Object.defineProperty(exports, "Redelegation", { enumerable: true, get: function () { return Redelegation_1.RedelegationV1B1; } });
var UnbondingDelegation_1 = require("./v1beta1/UnbondingDelegation");
Object.defineProperty(exports, "UnbondingDelegation", { enumerable: true, get: function () { return UnbondingDelegation_1.UnbondingDelegationV1B1; } });
var Validator_1 = require("./v1beta1/Validator");
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return Validator_1.ValidatorV1B1; } });
