// @xpla/xpla.js/dist/core/gov/v1beta1/msgs/MsgVoteWeighted.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgVoteWeightedV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Vote_1 = require("../Vote");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/gov/v1beta1/tx");
/**
 * Weighted vote for a proposal
 */
var MsgVoteWeightedV1B1 = /** @class */ (function (_super) {
    __extends(MsgVoteWeightedV1B1, _super);
    /**
     * @param proposal_id ID of proposal to vote on
     * @param voter voter's account address
     * @param option one of voting options
     */
    function MsgVoteWeightedV1B1(proposal_id, voter, options) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.voter = voter;
        _this.options = options;
        return _this;
    }
    MsgVoteWeightedV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options;
        return new MsgVoteWeightedV1B1(Number.parseInt(proposal_id), voter, options.map(function (o) { return Vote_1.WeightedVoteOptionV1B1.fromAmino(o); }));
    };
    MsgVoteWeightedV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options;
        return {
            type: isClassic ? 'gov/MsgVoteWeighted' : 'cosmos-sdk/MsgVoteWeighted',
            value: {
                proposal_id: proposal_id.toFixed(),
                voter: voter,
                options: options.map(function (o) { return o.toAmino(); }),
            },
        };
    };
    MsgVoteWeightedV1B1.fromData = function (data, _isClassic) {
        var proposal_id = data.proposal_id, voter = data.voter, options = data.options;
        return new MsgVoteWeightedV1B1(Number.parseInt(proposal_id), voter, options.map(function (o) { return Vote_1.WeightedVoteOptionV1B1.fromData(o); }));
    };
    MsgVoteWeightedV1B1.prototype.toData = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options;
        return {
            '@type': '/cosmos.gov.v1beta1.MsgVoteWeighted',
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            options: options.map(function (o) { return o.toData(); }),
        };
    };
    MsgVoteWeightedV1B1.fromProto = function (proto, _isClassic) {
        return new MsgVoteWeightedV1B1(proto.proposalId.toNumber(), proto.voter, proto.options.map(function (o) { return Vote_1.WeightedVoteOptionV1B1.fromProto(o); }));
    };
    MsgVoteWeightedV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options;
        return tx_1.MsgVoteWeighted.fromPartial({
            options: options.map(function (o) { return o.toProto(); }),
            proposalId: proposal_id,
            voter: voter,
        });
    };
    MsgVoteWeightedV1B1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1beta1.MsgVoteWeighted',
            value: tx_1.MsgVoteWeighted.encode(this.toProto()).finish(),
        });
    };
    MsgVoteWeightedV1B1.unpackAny = function (msgAny, _isClassic) {
        return MsgVoteWeightedV1B1.fromProto(tx_1.MsgVoteWeighted.decode(msgAny.value));
    };
    return MsgVoteWeightedV1B1;
}(json_1.JSONSerializable));
exports.MsgVoteWeightedV1B1 = MsgVoteWeightedV1B1;
