// @xpla/xpla.js/dist/core/gov/v1beta1/msgs/MsgVote.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgVoteV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var gov_1 = require("@xpla/xpla.proto/cosmos/gov/v1beta1/gov");
var tx_1 = require("@xpla/xpla.proto/cosmos/gov/v1beta1/tx");
/**
 * Vote for a proposal
 */
var MsgVoteV1B1 = /** @class */ (function (_super) {
    __extends(MsgVoteV1B1, _super);
    /**
     * @param proposal_id ID of proposal to vote on
     * @param voter voter's account address
     * @param option one of voting options
     */
    function MsgVoteV1B1(proposal_id, voter, option) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.voter = voter;
        _this.option = option;
        return _this;
    }
    MsgVoteV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option;
        return new MsgVoteV1B1(Number.parseInt(proposal_id), voter, option);
    };
    MsgVoteV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option;
        return {
            type: isClassic ? 'gov/MsgVote' : 'cosmos-sdk/MsgVote',
            value: {
                proposal_id: proposal_id.toFixed(),
                voter: voter,
                option: option,
            },
        };
    };
    MsgVoteV1B1.fromData = function (data, _isClassic) {
        var proposal_id = data.proposal_id, voter = data.voter, option = data.option;
        return new MsgVoteV1B1(Number.parseInt(proposal_id), voter, option);
    };
    MsgVoteV1B1.prototype.toData = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option;
        return {
            '@type': '/cosmos.gov.v1beta1.MsgVote',
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            option: option,
        };
    };
    MsgVoteV1B1.fromProto = function (proto, _isClassic) {
        return new MsgVoteV1B1(proto.proposalId.toNumber(), proto.voter, proto.option);
    };
    MsgVoteV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, option = _a.option;
        return tx_1.MsgVote.fromPartial({
            option: option,
            proposalId: proposal_id,
            voter: voter,
        });
    };
    MsgVoteV1B1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.gov.v1beta1.MsgVote',
            value: tx_1.MsgVote.encode(this.toProto()).finish(),
        });
    };
    MsgVoteV1B1.unpackAny = function (msgAny, _isClassic) {
        return MsgVoteV1B1.fromProto(tx_1.MsgVote.decode(msgAny.value));
    };
    return MsgVoteV1B1;
}(json_1.JSONSerializable));
exports.MsgVoteV1B1 = MsgVoteV1B1;
(function (MsgVoteV1B1) {
    MsgVoteV1B1.Option = gov_1.VoteOption;
})(MsgVoteV1B1 = exports.MsgVoteV1B1 || (exports.MsgVoteV1B1 = {}));
exports.MsgVoteV1B1 = MsgVoteV1B1;
