// @xpla/xpla.js/dist/core/gov/v1/Vote.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeightedVoteOptionV1 = exports.VoteV1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../util/json");
var numeric_1 = require("../../numeric");
var gov_1 = require("@xpla/xpla.proto/cosmos/gov/v1/gov");
/**
 * Stores vote information for a proposal
 */
var VoteV1 = /** @class */ (function (_super) {
    __extends(VoteV1, _super);
    /**
     * @param proposal_id defines the unique id of the proposal
     * @param voter is the voter address of the proposal
     * @param options is the weighted vote options
     * @param metadata is any  arbitrary metadata to attached to the vote
     */
    function VoteV1(proposal_id, voter, options, metadata) {
        var _this = _super.call(this) || this;
        _this.proposal_id = proposal_id;
        _this.voter = voter;
        _this.options = options;
        _this.metadata = metadata;
        _this.Option = gov_1.VoteOption;
        return _this;
    }
    VoteV1.fromAmino = function (data, _) {
        var proposal_id = data.proposal_id, voter = data.voter, options = data.options, metadata = data.metadata;
        return new VoteV1(Number.parseInt(proposal_id), voter, options.map(function (v) { return WeightedVoteOptionV1.fromAmino(v); }), metadata);
    };
    VoteV1.prototype.toAmino = function (_) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options, metadata = _a.metadata;
        var res = {
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            options: options.map(function (v) { return v.toAmino(); }),
            metadata: metadata,
        };
        return res;
    };
    VoteV1.fromData = function (data, _) {
        var proposal_id = data.proposal_id, voter = data.voter, options = data.options, metadata = data.metadata;
        return new VoteV1(Number.parseInt(proposal_id), voter, options.map(function (v) { return WeightedVoteOptionV1.fromData(v); }), metadata);
    };
    VoteV1.prototype.toData = function (_) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options, metadata = _a.metadata;
        var res = {
            proposal_id: proposal_id.toFixed(),
            voter: voter,
            options: options.map(function (v) { return v.toData(); }),
            metadata: metadata,
        };
        return res;
    };
    VoteV1.fromProto = function (proto, _) {
        return new VoteV1(proto.proposalId.toNumber(), proto.voter, proto.options.map(function (o) { return WeightedVoteOptionV1.fromProto(o); }), proto.metadata);
    };
    VoteV1.prototype.toProto = function (_) {
        var _a = this, proposal_id = _a.proposal_id, voter = _a.voter, options = _a.options, metadata = _a.metadata;
        return gov_1.Vote.fromPartial({
            proposalId: proposal_id,
            voter: voter,
            options: options.map(function (o) { return o.toProto(); }),
            metadata: metadata,
        });
    };
    return VoteV1;
}(json_1.JSONSerializable));
exports.VoteV1 = VoteV1;
var WeightedVoteOptionV1 = /** @class */ (function (_super) {
    __extends(WeightedVoteOptionV1, _super);
    function WeightedVoteOptionV1(option, weight) {
        var _this = _super.call(this) || this;
        _this.option = option;
        _this.weight = new numeric_1.Dec(weight);
        return _this;
    }
    WeightedVoteOptionV1.fromAmino = function (data, _) {
        var option = data.option, weight = data.weight;
        return new WeightedVoteOptionV1(option, weight);
    };
    WeightedVoteOptionV1.prototype.toAmino = function (_) {
        var _a = this, option = _a.option, weight = _a.weight;
        return {
            option: option,
            weight: weight.toString(),
        };
    };
    WeightedVoteOptionV1.fromData = function (data, _) {
        var option = data.option, weight = data.weight;
        return new WeightedVoteOptionV1(option, weight);
    };
    WeightedVoteOptionV1.prototype.toData = function (_) {
        var _a = this, option = _a.option, weight = _a.weight;
        return {
            option: option,
            weight: weight.toString(),
        };
    };
    WeightedVoteOptionV1.fromProto = function (proto, _) {
        return new WeightedVoteOptionV1(proto.option, proto.weight);
    };
    WeightedVoteOptionV1.prototype.toProto = function (_) {
        var _a = this, option = _a.option, weight = _a.weight;
        return gov_1.WeightedVoteOption.fromPartial({
            option: option,
            weight: weight.toString(),
        });
    };
    return WeightedVoteOptionV1;
}(json_1.JSONSerializable));
exports.WeightedVoteOptionV1 = WeightedVoteOptionV1;
