// @xpla/xpla.js/dist/core/xpla/v1beta1/msgs/MsgFundFeeCollector.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgFundFeeCollectorV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Coins_1 = require("../../../Coins");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/xpla/reward/v1beta1/tx");
/**
 * A basic message for sending [[Coins]] between Xpla accounts.
 */
var MsgFundFeeCollectorV1B1 = /** @class */ (function (_super) {
    __extends(MsgFundFeeCollectorV1B1, _super);
    /**
     * @param amount value of the transaction
     * @param depositor depositor's address
     */
    function MsgFundFeeCollectorV1B1(amount, depositor) {
        var _this = _super.call(this) || this;
        _this.depositor = depositor;
        _this.amount = new Coins_1.Coins(amount);
        return _this;
    }
    MsgFundFeeCollectorV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, amount = _a.amount, depositor = _a.depositor;
        return new MsgFundFeeCollectorV1B1(Coins_1.Coins.fromAmino(amount), depositor);
    };
    MsgFundFeeCollectorV1B1.prototype.toAmino = function (_isClassic) {
        var _a = this, amount = _a.amount, depositor = _a.depositor;
        return {
            type: 'xpladev/MsgFundFeeCollector',
            value: {
                amount: amount.toAmino(),
                depositor: depositor,
            },
        };
    };
    MsgFundFeeCollectorV1B1.fromData = function (data, _isClassic) {
        var amount = data.amount, depositor = data.depositor;
        return new MsgFundFeeCollectorV1B1(Coins_1.Coins.fromData(amount), depositor);
    };
    MsgFundFeeCollectorV1B1.prototype.toData = function (_isClassic) {
        var _a = this, amount = _a.amount, depositor = _a.depositor;
        return {
            '@type': '/xpla.reward.v1beta1.MsgFundFeeCollector',
            amount: amount.toData(),
            depositor: depositor,
        };
    };
    MsgFundFeeCollectorV1B1.fromProto = function (proto, _isClassic) {
        return new MsgFundFeeCollectorV1B1(Coins_1.Coins.fromProto(proto.amount), proto.depositor);
    };
    MsgFundFeeCollectorV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, amount = _a.amount, depositor = _a.depositor;
        return tx_1.MsgFundFeeCollector.fromPartial({
            amount: amount.toProto(),
            depositor: depositor,
        });
    };
    MsgFundFeeCollectorV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/xpla.reward.v1beta1.MsgFundFeeCollector',
            value: tx_1.MsgFundFeeCollector.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgFundFeeCollectorV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgFundFeeCollectorV1B1.fromProto(tx_1.MsgFundFeeCollector.decode(msgAny.value), isClassic);
    };
    return MsgFundFeeCollectorV1B1;
}(json_1.JSONSerializable));
exports.MsgFundFeeCollectorV1B1 = MsgFundFeeCollectorV1B1;
