// msg-parser/dist/nameid_mapping/IndexAndKindInformation.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var Long = __importStar(require("long"));
var PropertyKind;
(function (PropertyKind) {
    PropertyKind[PropertyKind["NUMERIC"] = 0] = "NUMERIC";
    PropertyKind[PropertyKind["STRING"] = 1] = "STRING";
})(PropertyKind = exports.PropertyKind || (exports.PropertyKind = {}));
var IndexAndKindInformation = /** @class */ (function () {
    function IndexAndKindInformation(bytes) {
        this.propertyIndex = Long.fromBytesLE(bytes.slice(2, 4)).toNumber();
        this.bytes = bytes;
        var twoTrailingBytesArray = bytes.slice(0, 2);
        var twoTrailingBytes = Long.fromBytesLE(twoTrailingBytesArray).toNumber();
        this.guidIndex = twoTrailingBytes >> 1;
        if ((twoTrailingBytes & 1) === 1) {
            this.propertyKind = PropertyKind.STRING;
        }
        else {
            this.propertyKind = PropertyKind.NUMERIC;
        }
    }
    IndexAndKindInformation.prototype.getPropertyIndex = function () {
        return this.propertyIndex;
    };
    IndexAndKindInformation.prototype.getGuidIndex = function () {
        return this.guidIndex;
    };
    IndexAndKindInformation.prototype.getPropertyKind = function () {
        return this.propertyKind;
    };
    return IndexAndKindInformation;
}());
exports.IndexAndKindInformation = IndexAndKindInformation;
