// @xpla/xpla.js/dist/core/upgrade/v1beta1/proposals/CancelSoftwareUpgradeProposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CancelSoftwareUpgradeProposalV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var upgrade_1 = require("@xpla/xpla.proto/cosmos/upgrade/v1beta1/upgrade");
/**
 *  CancelSoftwareUpgradeProposal is a gov Content type for cancelling a software upgrade
 */
var CancelSoftwareUpgradeProposalV1B1 = /** @class */ (function (_super) {
    __extends(CancelSoftwareUpgradeProposalV1B1, _super);
    /**
     *
     * @param title
     * @param description
     */
    function CancelSoftwareUpgradeProposalV1B1(title, description) {
        var _this = _super.call(this) || this;
        _this.title = title;
        _this.description = description;
        return _this;
    }
    CancelSoftwareUpgradeProposalV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, title = _a.title, description = _a.description;
        return new CancelSoftwareUpgradeProposalV1B1(title, description);
    };
    CancelSoftwareUpgradeProposalV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, title = _a.title, description = _a.description;
        return {
            type: isClassic
                ? 'upgrade/CancelSoftwareUpgradeProposal'
                : 'cosmos-sdk/CancelSoftwareUpgradeProposal',
            value: {
                title: title,
                description: description,
            },
        };
    };
    CancelSoftwareUpgradeProposalV1B1.fromData = function (data, _isClassic) {
        var title = data.title, description = data.description;
        return new CancelSoftwareUpgradeProposalV1B1(title, description);
    };
    CancelSoftwareUpgradeProposalV1B1.prototype.toData = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description;
        return {
            '@type': '/cosmos.upgrade.v1beta1.CancelSoftwareUpgradeProposal',
            title: title,
            description: description,
        };
    };
    CancelSoftwareUpgradeProposalV1B1.fromProto = function (proto, _isClassic) {
        return new CancelSoftwareUpgradeProposalV1B1(proto.title, proto.description);
    };
    CancelSoftwareUpgradeProposalV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description;
        return upgrade_1.CancelSoftwareUpgradeProposal.fromPartial({
            title: title,
            description: description,
        });
    };
    CancelSoftwareUpgradeProposalV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.upgrade.v1beta1.CancelSoftwareUpgradeProposal',
            value: upgrade_1.CancelSoftwareUpgradeProposal.encode(this.toProto(isClassic)).finish(),
        });
    };
    CancelSoftwareUpgradeProposalV1B1.unpackAny = function (msgAny, isClassic) {
        return CancelSoftwareUpgradeProposalV1B1.fromProto(upgrade_1.CancelSoftwareUpgradeProposal.decode(msgAny.value), isClassic);
    };
    return CancelSoftwareUpgradeProposalV1B1;
}(json_1.JSONSerializable));
exports.CancelSoftwareUpgradeProposalV1B1 = CancelSoftwareUpgradeProposalV1B1;
