// @xpla/xpla.js/dist/core/upgrade/v1beta1/proposals/SoftwareUpgradeProposal.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoftwareUpgradeProposalV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Plan_1 = require("../Plan");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var upgrade_1 = require("@xpla/xpla.proto/cosmos/upgrade/v1beta1/upgrade");
/**
 * Softwareupgradeproposal is a gov Content type for initiating a software upgrade.
 */
var SoftwareUpgradeProposalV1B1 = /** @class */ (function (_super) {
    __extends(SoftwareUpgradeProposalV1B1, _super);
    /**
     *
     * @param title
     * @param description
     * @param plan
     */
    function SoftwareUpgradeProposalV1B1(title, description, plan) {
        var _this = _super.call(this) || this;
        _this.title = title;
        _this.description = description;
        _this.plan = plan;
        return _this;
    }
    SoftwareUpgradeProposalV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, title = _a.title, description = _a.description, plan = _a.plan;
        return new SoftwareUpgradeProposalV1B1(title, description, plan ? Plan_1.PlanV1B1.fromAmino(plan) : undefined);
    };
    SoftwareUpgradeProposalV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, title = _a.title, description = _a.description, plan = _a.plan;
        return {
            type: isClassic
                ? 'upgrade/SoftwareUpgradeProposal'
                : 'cosmos-sdk/SoftwareUpgradeProposal',
            value: {
                title: title,
                description: description,
                plan: plan ? plan.toAmino() : undefined,
            },
        };
    };
    SoftwareUpgradeProposalV1B1.fromData = function (data, _isClassic) {
        var title = data.title, description = data.description, plan = data.plan;
        return new SoftwareUpgradeProposalV1B1(title, description, plan ? Plan_1.PlanV1B1.fromData(plan) : undefined);
    };
    SoftwareUpgradeProposalV1B1.prototype.toData = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, plan = _a.plan;
        return {
            '@type': '/cosmos.upgrade.v1beta1.SoftwareUpgradeProposal',
            title: title,
            description: description,
            plan: plan ? plan.toData() : undefined,
        };
    };
    SoftwareUpgradeProposalV1B1.fromProto = function (proto, _isClassic) {
        return new SoftwareUpgradeProposalV1B1(proto.title, proto.description, proto.plan ? Plan_1.PlanV1B1.fromProto(proto.plan) : undefined);
    };
    SoftwareUpgradeProposalV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, title = _a.title, description = _a.description, plan = _a.plan;
        return upgrade_1.SoftwareUpgradeProposal.fromPartial({
            title: title,
            description: description,
            plan: plan ? plan.toProto() : undefined,
        });
    };
    SoftwareUpgradeProposalV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.upgrade.v1beta1.SoftwareUpgradeProposal',
            value: upgrade_1.SoftwareUpgradeProposal.encode(this.toProto(isClassic)).finish(),
        });
    };
    SoftwareUpgradeProposalV1B1.unpackAny = function (msgAny, isClassic) {
        return SoftwareUpgradeProposalV1B1.fromProto(upgrade_1.SoftwareUpgradeProposal.decode(msgAny.value), isClassic);
    };
    return SoftwareUpgradeProposalV1B1;
}(json_1.JSONSerializable));
exports.SoftwareUpgradeProposalV1B1 = SoftwareUpgradeProposalV1B1;
