// @xpla/xpla.js/dist/core/staking/v1beta1/msgs/MsgCancelUnbondingDelegation.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgCancelUnbondingDelegationV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Coin_1 = require("../../../Coin");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
var tx_1 = require("@xpla/xpla.proto/cosmos/staking/v1beta1/tx");
/**
 * A delegator can undelegate an amount of bonded Luna, and will begin the unbonding
 * process for those funds. The unbonding process takes 21 days to complete, during
 * which the Luna cannot be transacted or swapped.
 */
var MsgCancelUnbondingDelegationV1B1 = /** @class */ (function (_super) {
    __extends(MsgCancelUnbondingDelegationV1B1, _super);
    /**
     * @param delegator_address delegator's account address
     * @param validator_address validator's operator address
     * @param amount Luna to be undelegated
     */
    function MsgCancelUnbondingDelegationV1B1(delegator_address, validator_address, amount, creation_height) {
        var _this = _super.call(this) || this;
        _this.delegator_address = delegator_address;
        _this.validator_address = validator_address;
        _this.amount = amount;
        _this.creation_height = creation_height;
        return _this;
    }
    MsgCancelUnbondingDelegationV1B1.fromAmino = function (data, _isClassic) {
        var _a = data.value, delegator_address = _a.delegator_address, validator_address = _a.validator_address, amount = _a.amount, creation_height = _a.creation_height;
        return new MsgCancelUnbondingDelegationV1B1(delegator_address, validator_address, amount ? Coin_1.Coin.fromAmino(amount) : undefined, Number.parseInt(creation_height));
    };
    MsgCancelUnbondingDelegationV1B1.prototype.toAmino = function (isClassic) {
        var _a = this, delegator_address = _a.delegator_address, validator_address = _a.validator_address, amount = _a.amount, creation_height = _a.creation_height;
        return {
            type: isClassic
                ? 'staking/MsgCancelUnbondingDelegation'
                : 'cosmos-sdk/MsgCancelUnbondingDelegation',
            value: {
                delegator_address: delegator_address,
                validator_address: validator_address,
                amount: amount ? amount.toAmino() : undefined,
                creation_height: creation_height.toFixed(),
            },
        };
    };
    MsgCancelUnbondingDelegationV1B1.fromProto = function (proto, _isClassic) {
        return new MsgCancelUnbondingDelegationV1B1(proto.delegatorAddress, proto.validatorAddress, proto.amount ? Coin_1.Coin.fromProto(proto.amount) : undefined, proto.creationHeight.toNumber());
    };
    MsgCancelUnbondingDelegationV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, delegator_address = _a.delegator_address, validator_address = _a.validator_address, amount = _a.amount, creation_height = _a.creation_height;
        return tx_1.MsgCancelUnbondingDelegation.fromPartial({
            delegatorAddress: delegator_address,
            validatorAddress: validator_address,
            amount: amount ? amount.toProto() : undefined,
            creationHeight: creation_height,
        });
    };
    MsgCancelUnbondingDelegationV1B1.prototype.packAny = function (isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.staking.v1beta1.MsgCancelUnbondingDelegation',
            value: tx_1.MsgCancelUnbondingDelegation.encode(this.toProto(isClassic)).finish(),
        });
    };
    MsgCancelUnbondingDelegationV1B1.unpackAny = function (msgAny, isClassic) {
        return MsgCancelUnbondingDelegationV1B1.fromProto(tx_1.MsgCancelUnbondingDelegation.decode(msgAny.value), isClassic);
    };
    MsgCancelUnbondingDelegationV1B1.fromData = function (data, _isClassic) {
        var delegator_address = data.delegator_address, validator_address = data.validator_address, amount = data.amount, creation_height = data.creation_height;
        return new MsgCancelUnbondingDelegationV1B1(delegator_address, validator_address, amount ? Coin_1.Coin.fromData(amount) : undefined, Number.parseInt(creation_height));
    };
    MsgCancelUnbondingDelegationV1B1.prototype.toData = function (_isClassic) {
        var _a = this, delegator_address = _a.delegator_address, validator_address = _a.validator_address, amount = _a.amount, creation_height = _a.creation_height;
        return {
            '@type': '/cosmos.staking.v1beta1.MsgCancelUnbondingDelegation',
            delegator_address: delegator_address,
            validator_address: validator_address,
            amount: amount ? amount.toData() : undefined,
            creation_height: creation_height.toFixed(),
        };
    };
    return MsgCancelUnbondingDelegationV1B1;
}(json_1.JSONSerializable));
exports.MsgCancelUnbondingDelegationV1B1 = MsgCancelUnbondingDelegationV1B1;
