// @xpla/xpla.js/dist/core/staking/v1beta1/msgs/MsgUpdateParams.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateStakingParamsV1B1 = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
var json_1 = require("../../../../util/json");
var Params_1 = require("../Params");
var tx_1 = require("@xpla/xpla.proto/cosmos/staking/v1beta1/tx");
var any_1 = require("@xpla/xpla.proto/google/protobuf/any");
/**
 * A delegator can undelegate an amount of bonded Luna, and will begin the unbonding
 * process for those funds. The unbonding process takes 21 days to complete, during
 * which the Luna cannot be transacted or swapped.
 */
var MsgUpdateStakingParamsV1B1 = /** @class */ (function (_super) {
    __extends(MsgUpdateStakingParamsV1B1, _super);
    /**
     * @param authority address that controls the module
     * @param params params defines the x/staking parameters to update
     */
    function MsgUpdateStakingParamsV1B1(authority, params) {
        var _this = _super.call(this) || this;
        _this.authority = authority;
        _this.params = params;
        return _this;
    }
    MsgUpdateStakingParamsV1B1.fromAmino = function (data, _isClassic) {
        var authority = data.authority, params = data.params;
        return new MsgUpdateStakingParamsV1B1(authority, params ? Params_1.StakingParamsV1B1.fromAmino(params) : undefined);
    };
    MsgUpdateStakingParamsV1B1.prototype.toAmino = function (_isClassic) {
        var _a = this, authority = _a.authority, params = _a.params;
        return {
            type: 'cosmos-sdk/MsgUpdateStakingParams',
            authority: authority,
            params: params ? params.toAmino() : undefined,
        };
    };
    MsgUpdateStakingParamsV1B1.fromProto = function (proto, _isClassic) {
        return new MsgUpdateStakingParamsV1B1(proto.authority, proto.params ? Params_1.StakingParamsV1B1.fromProto(proto.params) : undefined);
    };
    MsgUpdateStakingParamsV1B1.prototype.toProto = function (_isClassic) {
        var _a = this, authority = _a.authority, params = _a.params;
        return tx_1.MsgUpdateParams.fromPartial({
            authority: authority,
            params: params ? params.toProto() : undefined,
        });
    };
    MsgUpdateStakingParamsV1B1.prototype.packAny = function (_isClassic) {
        return any_1.Any.fromPartial({
            typeUrl: '/cosmos.staking.v1beta1.MsgUpdateParams',
            value: tx_1.MsgUpdateParams.encode(this.toProto()).finish(),
        });
    };
    MsgUpdateStakingParamsV1B1.unpackAny = function (msgAny, _isClassic) {
        return MsgUpdateStakingParamsV1B1.fromProto(tx_1.MsgUpdateParams.decode(msgAny.value));
    };
    MsgUpdateStakingParamsV1B1.fromData = function (data, _isClassic) {
        var authority = data.authority, params = data.params;
        return new MsgUpdateStakingParamsV1B1(authority, params ? Params_1.StakingParamsV1B1.fromData(params) : undefined);
    };
    MsgUpdateStakingParamsV1B1.prototype.toData = function (_isClassic) {
        var _a = this, authority = _a.authority, params = _a.params;
        return {
            '@type': '/cosmos.staking.v1beta1.MsgUpdateParams',
            authority: authority,
            params: params ? params.toData() : undefined,
        };
    };
    return MsgUpdateStakingParamsV1B1;
}(json_1.JSONSerializable));
exports.MsgUpdateStakingParamsV1B1 = MsgUpdateStakingParamsV1B1;
