// msg-parser/dist/property/PropertyTag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NamedPropertyMappingStorage_1 = require("../nameid_mapping/NamedPropertyMappingStorage");
var utils_1 = require("../utils");
var PropertyTag = /** @class */ (function () {
    function PropertyTag(propertyId, propertyType) {
        this.propertyId = propertyId;
        this.propertyType = propertyType;
    }
    PropertyTag.prototype.isNamed = function () {
        return (this.propertyId & NamedPropertyMappingStorage_1.NamedPropertyMappingStorage.NAMED_PROPERTY_ID_BASE) === NamedPropertyMappingStorage_1.NamedPropertyMappingStorage.NAMED_PROPERTY_ID_BASE;
    };
    PropertyTag.prototype.isMultiValued = function () {
        return (this.propertyType.id & PropertyTag.MV_PROPERTY_TYPE_BASE) === PropertyTag.MV_PROPERTY_TYPE_BASE;
    };
    PropertyTag.equal = function (tag1, tag2) {
        if (tag1 == null || tag2 == null) {
            return false;
        }
        else {
            return tag1.propertyId === tag2.propertyId &&
                tag1.propertyType.id === tag2.propertyType.id;
        }
    };
    PropertyTag.prototype.toString = function () {
        return utils_1.toHex(this.propertyId).toUpperCase() + utils_1.toHex(this.propertyType.id).toUpperCase();
    };
    PropertyTag.MV_PROPERTY_TYPE_BASE = 0x1000;
    return PropertyTag;
}());
exports.PropertyTag = PropertyTag;
