// msg-parser/dist/property/PropertyInfo.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var Long = __importStar(require("long"));
var PropertyTag_1 = require("./PropertyTag");
var PropertyTypes_1 = require("./property_type/PropertyTypes");
var PropertyInfo = /** @class */ (function () {
    function PropertyInfo(bytes) {
        this.bytes = bytes;
    }
    PropertyInfo.prototype.propertyId = function () {
        return Long.fromBytesLE(this.bytes.slice(2, 4)).toNumber();
    };
    PropertyInfo.prototype.propertyTypeId = function () {
        return Long.fromBytesLE(this.bytes.slice(0, 2)).toNumber();
    };
    PropertyInfo.prototype.propertyTag = function () {
        return new PropertyTag_1.PropertyTag(this.propertyId(), PropertyTypes_1.propertyTypeForId(this.propertyTypeId()));
    };
    PropertyInfo.prototype.flags = function () {
        return Long.fromBytesLE(this.bytes.slice(4, 8)).toNumber();
    };
    PropertyInfo.prototype.data = function () {
        return this.bytes.slice(8, 16);
    };
    return PropertyInfo;
}());
exports.PropertyInfo = PropertyInfo;
