// msg-parser/dist/property/PropertiesStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PropertyInfo_1 = require("./PropertyInfo");
var PropertyTag_1 = require("./PropertyTag");
var PropertiesStream = /** @class */ (function () {
    function PropertiesStream(stream, headerLength) {
        this.stream = stream;
        this.headerLength = headerLength;
    }
    PropertiesStream.prototype.properties = function () {
        var data = this.stream.read(this.headerLength, this.stream.getStreamSize());
        var result = [];
        for (var i = 0; i < data.length / 16; i++) {
            result.push(new PropertyInfo_1.PropertyInfo(data.slice(i * 16, (i + 1) * 16)));
        }
        return result;
    };
    PropertiesStream.prototype.getProperty = function (tag) {
        return this.properties().find(function (propertyInfo) { return PropertyTag_1.PropertyTag.equal(propertyInfo.propertyTag(), tag); });
    };
    PropertiesStream.prototype.findProperty = function (tag) {
        return this.properties().find(function (propertyInfo) { return PropertyTag_1.PropertyTag.equal(propertyInfo.propertyTag(), tag); });
    };
    PropertiesStream.STREAM_NAME = "__properties_version1.0";
    return PropertiesStream;
}());
exports.PropertiesStream = PropertiesStream;
